
# ClusterProfiler's GSEA (with gene set database GO)


In this script, we will do the following two things: 

- 1. Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through 
    all steps required to run clusterProfiler's GSEA with gene set database GO.
    
- 2. We will go through all (meaningful) researchers' degrees of freedom.

Note that we provide a separate file for the gene set databases GO and KEGG since clusterProfiler differs in the accepted gene ID formats for them. While for GO, a variety of formats, including Ensembl ID, is accepted, the gene IDs must be converted to NCBI (Entrez) ID when choosing KEGG.  

### Libraries 

All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.

```{r, eval = FALSE}
install.packages("BiocManager")
BiocManager::install("clusterProfiler")
BiocManager::install("org.Hs.eg.db")
BiocManager::install("DESeq2")
```

**Load Libraries **
```{r, message = FALSE, warning = FALSE}
library(clusterProfiler)
library(org.Hs.eg.db)
library(DESeq2)
```

Description of the packages 

- **clusterProfiler**: We here use clusterProfiler's implementation of GSEA.

- **org.Hs.eg.db**: Provides genome-wide annotation for humans. When working with a different organism,  the user has to provide a different package (see chapter 'About'). Note that this library is required when running clusterProfiler's GSEA with gene set database GO. 

- **DESeq2**: In contrast detecting the differential expressed genes, which is done solely based on adjusted $p$-vaules, we additionally use the estimated log fold changes when generating the gene ranking. In DESeq2, the estimated log fold change values are shrunken using an additional function, hence loading the library.  



## Load data 
Note that, as described above, clusterProfiler accepts Ensembl IDs when working with gene set database GO. When generating the ranking of genes with DESeq2, a shrinkage of the log fold change estimates is performed which requires the output of function *DESeq()*, here named **'dds'** (see Chapter 'Differential expression analysis'). 

```{r}
load("./data/Results_Differential_Expression_Analysis/DE_results_limma_Ensembl.Rdata")
load("./data/Results_Differential_Expression_Analysis/DE_results_DESeq2_Ensembl.Rdata")
load("./data/Results_Differential_Expression_Analysis/DE_results_edgeR_Ensembl.Rdata")

# load dds object 
load("./data/Results_Differential_Expression_Analysis/dds_Ensembl.Rdata")
```

## Run GSEA 


### Step 1: Generation of required input object

In the following, we illustrate the creation of the required input object using the three different methods for differential expression analysis, voom/limma, DESeq2, and edgeR, which shall serve as three options. The required input object for clusterProfiler is an "order ranked geneList", i.e. a vector with a ranking value for each gene, named with the respective gene IDs and ordered in a descending manner. \
Here we choose the following gene-level statistic based on which we can transform the results of differential expression analysis to the gene ranking: 
$$(-1) * \log_{10}(p\text{-value}) * \text{sign}(\text{log fold change})$$. \
Note that "$p$-value" denotes the non-adjusted p-value. 
    

#### Option 1: Generate required input using voom/limma

Before starting, we want to have a short look at the results table of differential expression analysis generated with limma so that we can identify the relevant columns. 

```{r}
head(DE_results_limma_Ensembl, n = 10)

```
Generate gene ranking: 
```{r}
# 1. Subset the gene expression data set to those genes that have a p-value (i.e. which have been NOT been excluded from differential expression analysis)

# indicate those genes with an existing p-value
ind_nonNA_pvalue_limma_Ensembl <- !is.na(DE_results_limma_Ensembl$P.Value)

# subset gene expression data set to those genes with an existing p-value
DE_results_noNA_Ensembl <- DE_results_limma_Ensembl[ind_nonNA_pvalue_limma_Ensembl, ]

# 2. create vector that contains the value of the gene-level ranking metric for each gene
rankvec_limma_Ensembl <- sign(DE_results_noNA_Ensembl$logFC)*(-1)*log10(DE_results_noNA_Ensembl$P.Value)

# 3. assign respective gene ID to each value in the vector
names(rankvec_limma_Ensembl) <- rownames(DE_results_noNA_Ensembl)

# 4. sort the vector in descending order
rankvec_limma_Ensembl <- sort(rankvec_limma_Ensembl, decreasing=TRUE)
```

Inspect the first entries of the ranking vector
```{r}
head(rankvec_limma_Ensembl , n = 10)
```


#### Option 2: Generate required input using DESeq2

**Step 1: Perform differential expression analysis**

This step is performed in Chapter 'Differential expression analysis'. Note that we must exit the DESeq2 workflow from the script a few steps early since there is an additional step required in for the creation of the required input object. We proceed with the object "dds_Ensembl" we've obtained in step 2 of DESeq2's workflow for differential expression analysis.

```{r, }
dds_Ensembl
```



**Step 2: Shrink estimated log fold change values**

The intuition behind shrinkage of log fold change values is that, as mentioned in the paper, RNA-Seq data consists (in its raw form) of count data which is inherently heteroscedastic, i.e. the variance of the count data depends on the mean count of the count data. It is observable that ratios between counts are considerably noisier in low magnitudes of counts compared to higher magnitudes, i.e. the log fold changes between both conditions are higher if the overall magnitude of counts is lower. \
DESeq2 addresses this issue by shrinking the estimated log fold changes in the direction of 0 the magnitude of shrinkage is higher if the available information for a gene is lower (which may be because of a low magnitude of counts, a high dispersion or few degrees of freedom.). A more detailed description is provided in the DESeq2 paper by Love et al. (2014).

Perform shrinkage:
```{r, message = FALSE}
DE_results_DESeq2_shrink_Ensembl <- lfcShrink(dds_Ensembl,
                                               coef = "condition_treated_vs_untreated",
                                               type="apeglm")
```

Function arguments:

- **type**: method to perform shrinkage. We opt for the default 'type = apeglm', however, there are two alternative options the user can choose from. 

- **coef**: indicate the coefficients to be shrunk. We can obtain the right argument from the following function call:

```{r}
resultsNames(dds_Ensembl)
```

This command shows us that we can either shrink the intercept or the "condition_treated_vs_untreated". Since we do not want to shrink the intercept but the estimated log fold changes, we opt for the second option "condition_treated_vs_untreated".

Before generating the ranking, we want to inspect the results table generated with DESeq2 (incl. shrinkage) so that we can identify the relevant columns

```{r}
head(DE_results_DESeq2_shrink_Ensembl, n = 10)
```

Now we proceed to generating the ranking. 

**Step 3: Generate the gene ranking based on the results of differential expression analysis**


```{r}
# 1. Subset the gene expression data set to those genes that have a p-value (i.e.
# which have been NOT been excluded from differential expression analysis)

# indicate those genes WITH a p-value
ind_nonNA_pvalue_Ensembl <- !is.na(DE_results_DESeq2_shrink_Ensembl$pvalue)

# subset gene expression data set to those genes with a p-value
DE_results_noNA_Ensembl <- DE_results_DESeq2_shrink_Ensembl[ind_nonNA_pvalue_Ensembl, ]

# 2. create vector that contains the value of the gene-level ranking metric for each gene
rankvec_DESeq2_Ensembl <- sign(DE_results_noNA_Ensembl$log2FoldChange)*(-1)*log10(DE_results_noNA_Ensembl$pvalue)

# 3. assign respective gene ID to each value in the vector
names(rankvec_DESeq2_Ensembl) <- rownames(DE_results_noNA_Ensembl)

# 4. sort the vector in descending order
rankvec_DESeq2_Ensembl <- sort(rankvec_DESeq2_Ensembl, decreasing=TRUE)
```

Inspect the first entries of the ranking vector
```{r}
head(rankvec_DESeq2_Ensembl , n = 10)
```


#### Option 3: Generate required input using edgeR

Before generating the ranking, we want inspect the results table of differential expression analysis generated with edgeR to identify the relevant columns.

```{r}
head(DE_results_edgeR_Ensembl, n = 10)
```

Now we proceed to generating the ranking.
```{r}
# 1. Subset the gene expression data set to those genes that have a p-value (i.e.
# which have been NOT been excluded from differential expression analysis)

# indicate those genes WITH a p-value
ind_nonNA_pvalue_Ensembl <- !is.na(DE_results_edgeR_Ensembl$PValue)

# subset gene expression data set to those genes with a p-value
DE_results_noNA_Ensembl <- DE_results_edgeR_Ensembl[ind_nonNA_pvalue_Ensembl, ]

# 2. create vector that contains the value of the gene-level ranking metric for each gene
rankvec_edgeR_Ensembl <- sign(DE_results_noNA_Ensembl$logFC)*(-1)*log10(DE_results_noNA_Ensembl$p_adj)

# 3. assign respective gene ID to each value in the vector
names(rankvec_edgeR_Ensembl) <- rownames(DE_results_noNA_Ensembl)

# 4. sort the vector in descending order
rankvec_edgeR_Ensembl <- sort(rankvec_edgeR_Ensembl, decreasing=TRUE)

# inspect ranking vector: 
head(rankvec_edgeR_Ensembl, n = 10)

# 5. special problem here: gene ENSG00000129824 has a ranking value Inf since its adjusted
# p-value in the results table of differential expression analysis amounts to 0

# here, we deal with this issue by resetting this ranking value to the highest ranking value
# that occurs among the remaining genes
# -> note that there is NO common way of dealing with this issue
rankvec_edgeR_Ensembl[rankvec_edgeR_Ensembl == Inf] <- max(rankvec_edgeR_Ensembl[rankvec_edgeR_Ensembl != Inf])
```

Inspect the first entries of the ranking vector: 
```{r}
head(rankvec_edgeR_Ensembl, n = 10)
```


### Step 2: Run GSEA with gene set database GO

For the purpose of simplicity, we work with the ranking generated using limma. However, you can switch around at your discretion.

```{r}
rankvec_Ensembl <- rankvec_limma_Ensembl

# alternatively: 
#rankvec_Ensembl <- rankvec_DESeq2_Ensembl
#rankvec_Ensembl <- rankvec_edgeR_Ensembl
```

Run GSEA with gene set database GO. Note that it is important to set a seed for reproducibility.

```{r, message = FALSE}

# important: set seed for reprodubibility
set.seed(1)

GSEA_GO <- gseGO(geneList = rankvec_Ensembl,
                   ont = "BP",
                   OrgDb = org.Hs.eg.db,
                   keyType = "ENSEMBL",
                   seed = TRUE) # set seed for reproducibility
```

Arguments of function *gseGO()*:

- **geneList**: vector of gene ranking (generated in manner of step 1)

- **ont**: subontology of GO (must be one of "BP", "CC", "MF")

- **OrgDb**: indicates organism from which the gene expression measurements are taken

- **keyTypes**: for our purposes here, argument keyType indicates the gene ID format of the gene names in vector rankvec
  - available keytypes can be found in the following: 

```{r}
keytypes(org.Hs.eg.db)
```

- **seed = TRUE**: set the seed you have indicated above (here: seed = 1 )


- additional arguments (these do not appear in the above code lines since we keep the default options)

- **exponent = 1**: in the calculation of the enrichment score, each gene is weighted by its absolute value of the ranking metric

- **eps = 1e-10**: each p-value that is smaller than 1e-10 is indicated as 1e-10

- **pvalueCutoff = 0.05**: only those gene sets with a p-value < 0.05 are indicated in the results. Note that if you want to inspect ALL gene sets in the results, set pvalueCutoff = 1

- **pAdjustMethod = "BH"**: adjustment for multiple testing using the method by Benjamini and Hochberg

### Step 3: Interpretation of the results 

**Inspect the results table**
```{r}
head(GSEA_GO, n = 10 )
```


Columns in the results table:

- **ID**: ID of gene set

- **Description**: Description of Gene Set

- **setSize**: Size of the gene set

- **enrichmentScore**: Enrichment Score
  - note: this enrichment score has not been normalized for gene set size
  - means: larger gene sets automatically have a bigger (absolute) enrichment score, independent of actual differential enrichment
  - the raw enrichment score is therefore not comparable between different gene sets

- **NES**: Normalized version of column enrichmentScore
  - NES can be compared between different gene sets

- **pvalue**: p-value of enrichment of given gene set
  - note: this raw p-value has not been adjusted for multiple testing
  - therefore: CANNOT be used to assess differential enrichment of a given gene set

- **p.adjust**: **adjusted** p-value of a given gene set. 
  - Note that this p-value now has been adjusted for multiple testing and can therefore be used to assess differential enrichment. For instance, detect all genes with p.adjust < 0.05 as differentially enriched.

- **qvalue**: **adjusted** p-value of a given gene set. Note that a qvalue is the analog to p.adjust, but adjusted for multiple testing using a different method.

Note that you can either use the column p.adjust or qvalue to assess whether a gene set is differentially enriched or not.

- **rank**: position in the ranked list of the genes in which the maximum difference between the two sums occurs, i.e. the rank at which the enrichment score is extracted.

- **leading_edge**:
  + **tags**: The percentage of genes  in the ranked list that are members of gene set before (for positive enrichment score) or after (for negative enrichment score) the position from which the enrichment score is extracted.
  + **list**:  The percentage of genes in the ranked gene list before (for positive enrichment
           score) or after (for negative enrichment score) the position from which the enrichment score is extracted.
  + **signal**: enrichment signal strength
           combines the statistics "tags" and "list"

  + **core_enrichment**: genes that contribute most to the enrichment results

Note that a more detailed information on leading_edge and core_enrichment can be found in the user manual provided for GSEA:
 (https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html): 
 
- search for **Detailed Enrichment Results** (section "Leading Edge")
 
- search for **CORE ENRICHMENT**




## Researchers' degrees of freedom 

In this part, we will go through all parameters that can be adapted in the GOSeq workflow. It is important to note that the intention behind going through the researchers' degrees of freedom is to give you an understanding of what you can do to adapt the given (parameter) setting to the research question. It is even more important to keep in mind that the intention behind going through these flexible parameters is NOT to change them in order to help you obtain the most preferable results by systematically changing these parameters since such behaviour would correspond to "cherry-picking". Any changes in the parameter choice must be documented transparently. 


### Change 1: change exponent 

Note that, in contrast to the web-based tool GSEA, clusterProfiler does not make any indication as to which alternative exponent values are allowed or sensible. One can stick with the web-based tool GSEA which suggests, in addition to the default exponent value of 1, the alternative values 0, 1.5, and 2.

Here, we change the exponent value to 1.5.

```{r, message = FALSE}
# important: set seed for reproducibility
set.seed(1)

GSEA_GO_exponent <- gseGO(geneList = rankvec_Ensembl,
                            ont = "BP",
                            OrgDb = org.Hs.eg.db,
                            keyType = "ENSEMBL",
                            exp = 1.5,
                            seed = TRUE)
```

**Inspect the results**
```{r}
head(GSEA_GO_exponent, n = 10) 
```

When switching the exponent to 2, the number of differentially enriched gene sets has decreased to two. 



