---
title: "Instructions GOSeq"
author: "Milena WÃ¼nsch"
date: "2023-07-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Content of this script 

In this script, we will do the following two things: 
- (1.) Based on the results of differential expression analysis, generated with voom/limma, DESeq2 and edgeR, we will go through all steps required to run GOSeq 
- (2.) We will go through all (meaningful) researchers' degrees of freedom 


### Libraries 

All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.

```{r, eval = FALSE}
install.packages("BiocManager")
BiocManager::install("tweeDEseqCountData")
BiocManager::install("goseq")

```

```{r, message = FALSE}
library(tweeDEseqCountData) # for pickrell data set 
library(goseq)
```

Description of the libraries: 

- tweeDESeqCountData: In addition to the results of differential expression analysis, we also need the raw count data for the case that we want to adjust the bias to account for (see step 4 of 'Researcher's degrees of freedom' at the end of this script).

- goseq: Offers the implementation of the method GOSeq.

#### Load data 

The starting point of GOSeq are results of differential expression analysis. Since GOSeq works with Ensembl gene IDs, we load the corresponding results of differential expression analysis with Ensembl gene IDs. 

```{r}
load("./data/Results_Differential_Expression_Analysis/DE_results_limma_Ensembl.Rdata")
load("./data/Results_Differential_Expression_Analysis/DE_results_DESeq2_Ensembl.Rdata")
load("./data/Results_Differential_Expression_Analysis/DE_results_edgeR_Ensembl.Rdata")
```

Since the relevant columns from the results table are unified for all three methods of differential expression analysis, the subsequent steps are identical for each results table. Therefore, we arbitrarily proceed with only one results table of differential expression analysis, namely DESeq2.

For a simplified readability, we store the corresponding results in an object with a more neutral name:
```{r}
DE_results <- DE_results_DESeq2_Ensembl

# note that you could easily change to one of the results tables of voom/limma or edgeR: 

# DE_results <- DE_results_limma_Ensembl
# or 
# DE_results <- DE_results_DESeq2_Ensembl

```


As mentioned above, we need the expression measurements of the pickrell data set for one of the researchers' degrees of freedom.

```{r}
# load pickrell data set
data(pickrell)

# set gene expression data set 
expression_data <- Biobase::exprs(pickrell.eset)
```


## (1.) Differential expression analysis 


### step 1: Preparation of required input object

GOSeq's required input object differs from the list of differentially expressed genes which is typically required by ORA. Instead, GOSeq requires as input a "named binary vector" with ...
 - (i) values 0 (for not differentially expressed genes) and 1 (for differentially expressed genes).
- (ii) names that correspond to the associated gene IDs.


Note that some methods for differential expression analysis, such as DESeq2, set the adjusted p-values of some genes to NA. This means that these genes are actually not tested for differential expression. Consequently, these genes cannot be assigned neither 1 nor 0 in the input object. We therefore remove genes with an adjusted p-value set to NA from the creation of the named binary vector.

```{r}
# 1. for each gene from the experiment, indicate those that have an EXISTING (i.e.
# NON-MISSING) p-value
indicator_nonNA_p <- !is.na(DE_results$p_adj)

# 2. using the indicator from 1., subset the results of differential expression
# analysis to only those genes with an EXISTING (i.e. NON-MISSING) p-value
DE_results_noNA <- DE_results[indicator_nonNA_p,]

# 3. create vector with binary value (0 or 1) for each gene in DE_results_noNA
# for this, we focus on the column p_adj which contains the adjusted p-values
bin_vec_DEG <- ifelse(DE_results_noNA$p_adj < 0.05 , 1,0)

# note: function ifelse() goes through each adjusted p-value in the results
# of differential expression analysis and in a new vector, assigns value 1
# if the adjusted p-value is < 0.05 and else value 0

# 4. assign each value of the vector the corresponding gene ID
names(bin_vec_DEG) <- rownames(DE_results_noNA)

# 5. results: named binary vector bin_vec_DEG
# note that we here inspect only the first few entries 
head(bin_vec_DEG, n = 20)
```
Note that "bin_vec_DEG" stands for BINary VECtor of Differentially Expressed Genes


For illustration purposes: compare the number of differentially expressed genes from the results table of differential expression analysis to the number of differentially expressed genes from the named binary vector

```{r}
# (i) initial results table of differential expression analysis
table(DE_results$p_adj < 0.05)

# (ii) results table of differential expression analysis subsetted to those genes
# with a non-missing adjusted p-value
table(DE_results_noNA$p_adj < 0.05)

# (iii) named binary vector  
table(bin_vec_DEG)
```


### step 2: Calculate value of probability weighting function PWF

GOSeq works with the function nullp() which provides one row of data for each gene with the following information:

 - DEgenes: indicates status of differential expression (0/1). This information can be directly extracted from the input object
 
 - bias.data: numeric value of bias for each gene concerning the detection of differential expression. This usually corresponds to the length of the gene
 
 - pwf: value of the probability weighting function of the gene

Note that the values of bias.data depend on the function argument bias.data in function nullp().  (further information below)

```{r}
ProbabilityWeighting<- nullp(DEgenes = bin_vec_DEG,
             genome= "hg19",
             id= "ensGene")

```

function arguments:

 - DEgenes: named binary vector generated in step 1
 - genome: indicates organism from which gene expression is measured
 
 - -> specify organism = "hg19" for human
 - -> for other organisms, the argument must be adapted accordingly (such as "mm9" for mouse)

 - id: indicates gene identifier
 
 -> specify id = "knownGene" for Entrez gene ID
 -> specify id = "ensGene" for Ensembl gene ID (which is the case here )
 -> specify id = "geneSymbol" for HGNC gene symbols

 For an overview of available gene identifiers and supported organisms see function supportedGeneIDs() (or for an even better overview: View(supportedGeneIDs())
 

Note 1: argument bias.data specifies data on which the detection of differential expression
 might be dependent on. While the default value bias.data = NULL specifies that the length of each gene is retrieved from the UCSC
 genome browser, the user can alternatively provide a vector which contains an entry for each gene contained by bin_vec_DEG. An alternative suggested by the authors of GOSeq would be to account for the read count bias of each gene, i.e. for the total number of read counts across all samples that are assigned to each individual gene


Note 2: by default, the probability weight function is plotted when nullp() is run. This can be deactivated by adding the function argument plot.fit = FALSE. 


### step 3: Test for differential enrichment

```{r}
GOSeq_results <- goseq(pwf = ProbabilityWeighting,
                       genome = "hg19",
                       id = "ensGene",
                       test.cats = "GO:BP")

```

 function arguments:
 
 - pwf: typically the output of the nullp() function (step 2), i.e. probability weighting function
 genome: indicates organism from which gene expression is measured (see also step 2)
 - -> specify organism = "hg19" for human
 - -> for other organisms, the argument must be adapted accordingly (such as "mm9" for mouse)
 
 - id: indicates gene identifier (see also step 2)
 - -> specify id = "knownGene" for Entrez gene ID
 - -> specify id = "ensGene" for Ensembl gene ID
 - -> specify id = "geneSymbol" for HGNC gene symbols


 - test.cats: gene set database
 - -> "GO:BP", "GO:MF", "GO:CC" : geneset database GO with respective subontology
 - -> "KEGG": geneset database KEGG
 
 
### step 4: Multiple Test Adjustment
 
Note that GOSeq does not perform adjustment for multiple testing internally so that is MUST be performed by the user manually. We therefore add a new column that contains the corresponding adjusted p-values. Relevant for over-representation test are the (unadjusted) p-values from column "over_represented_pvalue" which we adjust for multiple testing in the following: 

```{r}
GOSeq_results$p_adj_overrepresented <- p.adjust(GOSeq_results$over_represented_pvalue)
```

Note that by default, p.adjust() performs multiple test adjustment based on Benjamini and Hochberg.

### step 5: Interpretation of results table 

Inspect final results table: 
```{r}
head(GOSeq_results, n = 10)
```

Columns in results table: 
- Category: provides the ID of the gene set (based on the choice of the gene set
 database in step 4)
- over_represented_pvalue: p-value of over-representation
 IMPORTANT NOTE: DO NOT use these p-values to assess differential enrichment
 of a given gene set, since these p-value have not been adjusted for multiple
 testing
 
- under_represented_p-value: p-value of under-representation
 -> not relevant if you want to test for OVER-REPRESENTATION
 
 IMPORTANT NOTE: if you are interested in testing for UNDER-REPRESENTATION,
 you must adjust the p-values for multiple testing:
 
```{r}
#GOSeq_results$p_adj_underrepresented <- p.adjust(GOSeq_results$under_represented_pvalue)
```
 
 
- numDEInCat: gives the number of differentially expressed genes from the input that are members of the given gene set

- numInCat: number of genes that are members of the given gene set

- term: description of the gene set
 -> NOTE: this column is only provided for the geneset database GO

- ontology: subontology of geneset database GO (based on choice in step 3)
 -> this column is only provided for the geneset database GO



 - p_adj_overrepresented: p-value of over-representation that has been ADJUSTED
 for multiple testing
 -> based on these adjusted p-values, differential enrichment (i.e. significant
 over-representation) can be adjusted
 - typically: detect those gene sets as differentially enriched with a value of
 p_adj_overrepresented < 0.05



## (2.) Researchers' degrees of freedom 

In this part, we will go through all parameters that can be adapted in the GOSeq workflow. It is important to note that the intention behind going through the researchers' degrees of freedom is to give you an understanding of what you can do to adapt the given (parameter) setting to the research question. It is even more important to keep in mind that the intention behind going through these flexible parameters is NOT to change them in order to help you obtain the most preferable results by systematically changing these parameters since such behaviour would correspond to "cherry-picking". Any changes in the parameter choice must be documented transparently. 


### change 1: Change gene set database 

The gene set database can be adjusted in the function goseq() in argument "test.cats", while the arguments and therefore also the results of function pwf() remain untouched. 

Here, we want to change the gene set database to KEGG 

```{r}
GOSeq_results_database <- goseq(pwf = ProbabilityWeighting,
                                genome = "hg19",
                                id = "ensGene",
                                test.cats = "KEGG")


# adjust the p-values of over-representation for multiple testing 
GOSeq_results_database$p_adj_overrepresented <- p.adjust(GOSeq_results_database$over_represented_pvalue)


# inspect the results: 
head(GOSeq_results_database, n = 10)
```

Note that when the gene set database KEGG is specified, then no column "term" is provided in the results table to give a description of the corresponding gene set. 

Other gene set databases can be specified by setting the argument to 


- test.cats = "GO:CC" -> GO with subontology Cellular Component
- test.cats = "GO:MF" -> GO with subontology Molecular Function


### change 2: Include genes that are no member of any gene set in calculation of the p-value 

By default, only those gene sets are included in the calculation of the p-value that are a member to 
at least one gene set. However, the user has the choice to include all genes from the input object, 
thus particularly those that are no member of any gene set. 

The additional inclusion of those genes that are no members of any gene set can by specified in argument "use_genes_without_cat"
in function goseq()

```{r}
GOSeq_results_allgenes <- goseq(pwf = ProbabilityWeighting,
                                    genome = "hg19",
                                    id = "ensGene",
                                    test.cats = "GO:BP",
                                    use_genes_without_cat = TRUE)

# adjust the p-values of over-representation for multiple testing 
GOSeq_results_allgenes$p_adj_overrepresented <- p.adjust(GOSeq_results_allgenes$over_represented_pvalue)

# inspect the results: 
head(GOSeq_results_allgenes, n = 10)
```

Note that the argument "use_genes_without_cat = TRUE" used to be the default in earlier versions of the library goseq. 



### change 3: Change the method for the calculation of the p-value 

The method for the calculation of the p-value can be adapted in function goseq() in the argument "method". 

Here, we want to change the method for the calculation of the p-value from Wallenius approximation to the computationally
expensive resampling. 

```{r, warnings = FALSE, message = FALSE}
GOSeq_results_method <- goseq(pwf = ProbabilityWeighting,
                              genome = "hg19",
                              id = "ensGene",
                              test.cats = "GO:BP",
                              method = "Sampling")


# adjust the p-values of over-representation for multiple testing 
GOSeq_results_method$p_adj_overrepresented <- p.adjust(GOSeq_results_method$over_represented_pvalue)


# inspect the results: 
head(GOSeq_results_method, n = 10 )
```

Note that by default, 2000 samples are calculated in the course of resampling (see argument "repcnt"). This argument should NEVER be played around with to generate preferable results. Instead, it is advisable to keep this parameter in its default configuration.


Also note that, theoretically, GOSeq additionally offers the use of the standard hypergeometric distribution as the method to calculate a p-value of over-representation. However, users are explicitly advised against using this option since the standard hypergeometric distribution does not adjust for ANY biases that might be present in the data. 



### change 4: Change the bias to account for in the analysis 

As an example taken from the user manual of GOSeq, we now want to adjust for the total number of counts of each gene. The idea behind this is that in the context of RNA-Seq data, the magnitude of counts of a given gene sets reflects its overall expression level. In RNA-Seq data, which takes the form of count data, a higher magnitude of counts leads to an increased statistical power (to detect a gene as differentially expressed). It is important to note that accounting for the total number of counts might also remove the bias that results from actual differential expression between both conditions. 


##### step 1: create a vector that contains the total number of read counts for each gene

```{r}
# 1. Compute the sum of read counts of each gene across all samples from the gene expression data set
countbias <- rowSums(expression_data)

# 2. Subset the vector countbias to all genes that are part of the input object bin_vec_DEG

# (i) indicate which genes are a part of the input object:
ind_input <- names(countbias) %in% names(bin_vec_DEG)

# (ii) subset vector countbias
countbias <- countbias[ind_input]
```

##### step 2: Provide countbias to function nullp() 

The custom bias can be provided to function nullp() in argument "bias.data":
```{r}
ProbabilityWeighting_countbias <- nullp(DEgenes = bin_vec_DEG,
                                        genome= "hg19",
                                        id= "ensGene",
                                        bias.data = countbias) # countbias
```

##### step 3: Run function goseq() with adjusted probability weightings 

```{r}
GOSeq_results_countbias <- goseq(pwf = ProbabilityWeighting_countbias, # adjusted probability weightings
                                 genome = "hg19",
                                 id = "ensGene",
                                 test.cats = "GO:BP")
```

##### step 4: Perform adjustment for multiple testing and inspect final results 


```{r}
GOSeq_results_countbias$p_adj_overrepresented <- p.adjust(GOSeq_results_countbias$over_represented_pvalue)
```





