---
title: "PADOG"
author: "Pat Callahan"
date: "2023-06-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

# Content of this script 

In this script, we will go through the process of running PADOG. Note that PADOG requires the transformed RNA-Seq data and the genes in the Entrez ID format. 


## Libraries 

All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine. 

```{r, eval = FALSE}
install.packages("BiocManager")
BiocManager::install("PADOG")
BiocManager::install("tweeDEseqCountData")
BiocManager::install("KEGGREST")

```
Load libraries:

```{r, warning = FALSE, message = FALSE}
BiocManager::install("KEGGREST")
library(PADOG)
library(tweeDEseqCountData)
library(KEGGREST)
```

Provide a brief note on what these libraries are for?

- PADOG: Provides the implementation of the method PADOG 

- tweeDEseqCountData: In addition to the preprocessed Pickrell data set, which we load in the following step, PADOG requires the conditions of the samples 

- KEGGREST: To obtain the character value for a variety of organisms the user is required to provide in the function padog(). While the human organism is indicated by default, it might become necessary to consult the KEGGREST package if working with a different organism. 

# Load Data

This section will change substantially when I reconfigure the project as an R package/book.

For now, place any data you need into the `./data` directory.

```{r}
# we load the voom-transformed Pickrell data set 
load("data/expression_data_voomtransformed_Entrez.Rdata")

# alternatively: load the gene expression measurements that have been transformed using 
# load("data/expression_data_vsttransformed_Entrez.Rdata")

# additionally, we load the pickrell data set so that we can access the sample conditions
data(pickrell)

```

The sample conditions (i.e. phenotype labels) of the pickrell data set can be accessed using

```{r}
pickrell.eset$gender 
```

We proceed with the voom-transformed pickrell data set and the corresponding phenotype labels

```{r}
# gene expression measurements (transformed)
# note: you can also proceed with the vst-transformed gene expression measurements 
expression_data_transformed <- expression_data_voomtransformed_Entrez

# sample conditions
sample_conditions <- pickrell.eset$gender

```

# Step 1: Prepare Sample Conditions

First, we inspect the form of the initial (raw) sample conditions

```{r}
# look at the class: 
class(sample_conditions)
# -> the sample labels are already coded as factor

# the current levels are:
levels(sample_conditions)
```

PADOG requires character vector with class labels of the samples. It can only contain "c" for control samples or "d" for disease samples

```{r}

# prepare sample conditions
# we want to convert 
# (i) "female" to "c"
# (ii) "male" to "d"
sample_conditions_prep <- factor(sample_conditions, 
                                levels=c("female","male"), 
                                labels=c("c","d"))

```

# Step 2: Run PADOG

*It is recommended to set a seed to ensure exact reproducibility of the results if the code is run at multiple time points*

You can specify any integer number as the seed. It is VERY IMPORTANT to choose the seed arbitrarily and WITHOUT INSPECTING the results the seed should NEVER be specified based on which value yields the most preferable results.

```{r, message = FALSE, results = "hide"}
# run PADOG: 
   PADOG_results <- padog(esetm = as.matrix(expression_data_transformed), 
                          group = sample_conditions_prep, 
                          dseed = 1)
```

arguments:

-   esetm: matrix that contains the expression measurements
    -   note: since the expression data is initially stored in a data frame, we transform it to a matrix when running PADOG
-   group: sample conditions (has values "c" and "d")
-   dseed:seed for random number generation (used in the process of phenotype permutation)

additional arguments:

-   paired: indicates whether the samples in both groups are paired
-   block: if the samples are paired (i.e. argument paired = TRUE), then the paired samples must have the same block value
-   gslist: gives instructions on how to cluster the genes into gene sets
    -   gslist = "KEGGRESTpathway": gene sets correspond to KEGG pathways
    -   alternative: provide a user-defined list of gene sets
-   organism: organism from which the gene expression measurements are taken
    -   for human, set organism = "hsa"
    -   the required character value for other organisms can be extracted from the KEGGREST package:
- obtain required organisms from column organism 
- annotation: required if gslist is set to "KEGGRESTpathway" and the rownames of esetm are probe IDs 
- can be set to NULL of gslist is set to "KEGGRESTpathway" and the rownames of esetm are in the Entrez gene ID format 
- if rownames are other gene IDs, then sett annotation = NULL and make sure that the rownames are elements of gslist (and unique!)
- gs.names: contains names of gene sets -> character vector 
   - must have the same length as gslist 
- NI: number of phenotype permutations employed in the assessment of the significance of a given gene set 


We want to take a first look at the results table. Note, however, that that we still have to add the adjusted p-values (see next step), based on which differential enrichment is eventually detected.

```{r}
head(PADOG_results , n = 10)
```




# Step 3: Adjust for Multiple Testing


PADOG does not perform multiple test adjustment internally so that must be performed by the user. We here work with a function from the R package base which uses the method by Benjamini and Hochberg by default. 


```{r}
# add adjusted p-value in column Ppadog_adjusted
PADOG_results$Ppadog_adjusted <- p.adjust(PADOG_results$Ppadog)
```


# Step 4: Interpretation of Results

We want to inspect (a part of) the results and interpret the columns provided in the results table. 

```{r}
head(PADOG_results, n = 10)
```

Differential enrichment of a given gene set can now be assessed based on the adjusted p-value in column Ppadog_adjusted. For instance: detect all gene sets with Ppadog_adjusted < 0.05 as differentially enriched.

Additional columns: 

- Name: Name of gene set 

- ID: Identifier of gene set 

- Size: number of genes in gene set 

- meanAbsT0: Mean of absolute (moderated) t-statistic of all genes that are a member of the given gene set 

- padog0: Mean of abolute weighted moderate t-statistic of all genes that are a member of the given gene set 

- PmeanAbsT: significance of of meanAbsT0

- Ppadog: significance padog0
