<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 8 Cluster Profiler: GSEA GO | RNA Sequencing Guide</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will do the following two things: - 1. Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 8 Cluster Profiler: GSEA GO | RNA Sequencing Guide">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will do the following two things: - 1. Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 8 Cluster Profiler: GSEA GO | RNA Sequencing Guide">
<meta name="twitter:description" content="In this script, we will do the following two things: - 1. Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">RNA Sequencing Guide</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="prefiltering.html"><span class="header-section-number">2</span> Prefiltering</a></li>
<li><a class="" href="gsea-preranking.html"><span class="header-section-number">3</span> GSEA Preranking</a></li>
<li><a class="" href="gene-id-duplicate-removals.html"><span class="header-section-number">4</span> Gene ID Duplicate Removals</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">5</span> Differential Expression Analysis</a></li>
<li><a class="" href="rna-seq-transformation.html"><span class="header-section-number">6</span> RNA-Seq Transformation</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="go-seq.html"><span class="header-section-number">7</span> Go SEQ</a></li>
<li><a class="active" href="cluster-profiler-gsea-go.html"><span class="header-section-number">8</span> Cluster Profiler: GSEA GO</a></li>
<li><a class="" href="david.html"><span class="header-section-number">9</span> DAVID</a></li>
<li><a class="" href="gsea-web.html"><span class="header-section-number">10</span> GSEA WEB</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="cluster-profiler-gsea-kegg.html"><span class="header-section-number">11</span> Cluster Profiler: GSEA KEGG</a></li>
<li><a class="" href="cluster-profiler-ora.html"><span class="header-section-number">12</span> Cluster Profiler: ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">13</span> PADOG</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="cluster-profiler-gsea-go" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> Cluster Profiler: GSEA GO<a class="anchor" aria-label="anchor" href="#cluster-profiler-gsea-go"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will do the following two things:
- 1. Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through
all steps required to run clusterProfiler’s GSEA with gene set database GO.
- 2. We will go through all (meaningful) researchers’ degrees of freedom.</p>
<p>Note that we provide a separate file for the gene set databases GO and KEGG since clusterProfiler differs in the accepted gene ID formats for them. While for GO, a variety of formats, including Ensembl ID, is accepted, the gene IDs must be converted to NCBI (Entrez) ID when choosing KEGG.</p>
<div id="libraries-6" class="section level3" number="8.0.1">
<h3>
<span class="header-section-number">8.0.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-6"><i class="fas fa-link"></i></a>
</h3>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span></code></pre></div>
<div id="load-libraries-3" class="section level4" number="8.0.1.1">
<h4>
<span class="header-section-number">8.0.1.1</span> Load Libraries<a class="anchor" aria-label="anchor" href="#load-libraries-3"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span></code></pre></div>
<p>Description of the packages</p>
<ul>
<li><p>clusterProfiler: We here use clusterProfiler’s implementation of GSEA.</p></li>
<li><p>org.Hs.eg.db: Provides genome-wide annotation for humans. When working with a different organism, the user has to provide a different package (‘Overview_Packages’). Note that this library is required when running clusterProfiler’s GSEA with gene set database GO.</p></li>
<li><p>DESeq2: In contrast detecting the differential expressed genes, which is done solely based on adjusted <span class="math inline">\(p\)</span>-vaules, we additionally use the estimated log fold changes when generating the gene ranking. In DESeq2, the estimated log fold change values are shrunken using an additional function, hence loading the library.</p></li>
</ul>
</div>
</div>
<div id="load-the-data" class="section level3" number="8.0.2">
<h3>
<span class="header-section-number">8.0.2</span> Load the data<a class="anchor" aria-label="anchor" href="#load-the-data"><i class="fas fa-link"></i></a>
</h3>
<p>Note that, as described above, clusterProfiler accepts Ensembl IDs when working with gene set database GO. When generating the ranking of genes with DESeq2, a shrinkage of the log fold change estimates is performed which requires the output of function DESeq(), here named “dds” (see “Instructions_Differential_Expression_Analysis”).</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_limma_Ensembl.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_DESeq2_Ensembl.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_edgeR_Ensembl.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load dds object </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/dds_Ensembl.Rdata"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="run-gsea" class="section level2" number="8.1">
<h2>
<span class="header-section-number">8.1</span> Run GSEA<a class="anchor" aria-label="anchor" href="#run-gsea"><i class="fas fa-link"></i></a>
</h2>
<div id="step-1-generation-of-required-input-object" class="section level3" number="8.1.1">
<h3>
<span class="header-section-number">8.1.1</span> step 1: Generation of Required Input Object<a class="anchor" aria-label="anchor" href="#step-1-generation-of-required-input-object"><i class="fas fa-link"></i></a>
</h3>
<p>In the following, we illustrate the creation of the required input object using the three different methods for differential expression analysis, voom/limma, DESeq2, and edgeR, which shall serve as three options. The required input object for clusterProfiler is an “order ranked geneList”, i.e. a vector with a ranking value for each gene, named with the respective gene IDs and ordered in a descending manner.</p>
<div id="option-1-generate-required-input-using-voomlimma" class="section level4" number="8.1.1.1">
<h4>
<span class="header-section-number">8.1.1.1</span> 1.1 Option 1: Generate required input using voom/limma<a class="anchor" aria-label="anchor" href="#option-1-generate-required-input-using-voomlimma"><i class="fas fa-link"></i></a>
</h4>
<p>The formula for generating the gene ranking from the results table of differential expression analysis is <span class="math inline">\((-1) * log_{10}(p\text{-value}) * \text{sign}(\text{log fold change})\)</span>. Note that “<span class="math inline">\(p\)</span>-value” denotes the non-adjusted p-value.</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_limma_Ensembl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_limma_Ensembl</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA_Ensembl</span> <span class="op">&lt;-</span> <span class="va">DE_results_limma_Ensembl</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_limma_Ensembl</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_limma_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_limma_Ensembl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_limma_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_limma_Ensembl</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect the first entries of the ranking vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_limma_Ensembl</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000154620 </span></span>
<span><span class="co">#&gt;       54.973990       50.395302       38.292623 </span></span>
<span><span class="co">#&gt; ENSG00000120868 ENSG00000110031 ENSG00000151445 </span></span>
<span><span class="co">#&gt;        3.569036        3.262283        3.222725 </span></span>
<span><span class="co">#&gt; ENSG00000065308 ENSG00000095015 ENSG00000198689 </span></span>
<span><span class="co">#&gt;        3.198940        3.033258        2.949591 </span></span>
<span><span class="co">#&gt; ENSG00000102893 </span></span>
<span><span class="co">#&gt;        2.915174</span></span></code></pre></div>
</div>
<div id="option-2-generate-required-input-using-deseq2" class="section level4" number="8.1.1.2">
<h4>
<span class="header-section-number">8.1.1.2</span> 1.2 Option 2: Generate required input using DESeq2<a class="anchor" aria-label="anchor" href="#option-2-generate-required-input-using-deseq2"><i class="fas fa-link"></i></a>
</h4>
<div id="i-perform-differential-expression-analysis" class="section level5" number="8.1.1.2.1">
<h5>
<span class="header-section-number">8.1.1.2.1</span> (i) Perform Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#i-perform-differential-expression-analysis"><i class="fas fa-link"></i></a>
</h5>
<p>This step is performed in the script Instructions_Differential_Expression_Analysis.R. Note that we must exit the DESeq2 workflow from the script a few steps early since there is an additional step required in for the creation of the required input object. We proceed with the object “dds_Ensembl” we’ve obtained in step 3 of DESeq2’s workflow for differential expression analysis.</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds_Ensembl</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 10151 69 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(6): counts mu ... replaceCounts replaceCooks</span></span>
<span><span class="co">#&gt; rownames(10151): ENSG00000000419 ENSG00000000457 ...</span></span>
<span><span class="co">#&gt;   ENSG00000254245 ENSG00000254290</span></span>
<span><span class="co">#&gt; rowData names(23): baseMean baseVar ... maxCooks</span></span>
<span><span class="co">#&gt;   replace</span></span>
<span><span class="co">#&gt; colnames(69): NA18486 NA18498 ... NA19239 NA19257</span></span>
<span><span class="co">#&gt; colData names(3): condition sizeFactor replaceable</span></span></code></pre></div>
</div>
<div id="ii-shrink-estimated-log-fold-change-values" class="section level5" number="8.1.1.2.2">
<h5>
<span class="header-section-number">8.1.1.2.2</span> (ii) Shrink estimated log fold change values<a class="anchor" aria-label="anchor" href="#ii-shrink-estimated-log-fold-change-values"><i class="fas fa-link"></i></a>
</h5>
<p>The intuition behind shrinkage of log fold change values is that, as mentioned in the paper, RNA-Seq data consists (in its raw form) of count data in which is inherently heteroscedastic, i.e. the variance of the count data depends on the mean count of the count data. It is observable that ratios between counts are considerably noisier in low magnitudes of counts compared to higher magnitudes, i.e. the log fold changes between both conditions are higher if the overall magnitude of counts is lower.</p>
<p>DESeq2 addresses this issue by shrinking the estimated log fold changes in the direction of 0 the magnitude of shrinkage is higher if the available information for a gene is lower (which may be because of a low magnitude of counts, a high dispersion
or few degrees of freedom.). A more detailed description is provided in the DESeq2 paper by Love et al. (2014).</p>
<p>Perform shrinkage:</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">DE_results_DESeq2_shrink_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span><span class="op">(</span><span class="va">dds_Ensembl</span>,</span>
<span>                                               coef <span class="op">=</span> <span class="st">"condition_treated_vs_untreated"</span>,</span>
<span>                                               type<span class="op">=</span><span class="st">"apeglm"</span><span class="op">)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li><p>type: method to perform shrinkage. We opt for the default “apeglm” but you can choose from two alternative options
as well</p></li>
<li><p>coef: indicate the coefficients to be shrunk. We can obtain the right argument from the following function call:</p></li>
</ul>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">resultsNames</a></span><span class="op">(</span><span class="va">dds_Ensembl</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Intercept"                     </span></span>
<span><span class="co">#&gt; [2] "condition_treated_vs_untreated"</span></span></code></pre></div>
<p>This command shows us that we can either shrink the intercept or the “condition_treated_vs_untreated”. Since we do not want to shrink the intercept but the estimated log fold changes, we opt for the second option “condition_treated_vs_untreated”.</p>
</div>
<div id="iii-generate-the-ranked-list-of-the-genes-based-on-the-results-of-differential-expression-analysis" class="section level5" number="8.1.1.2.3">
<h5>
<span class="header-section-number">8.1.1.2.3</span> (iii) Generate the ranked list of the genes based on the results of differential expression analysis<a class="anchor" aria-label="anchor" href="#iii-generate-the-ranked-list-of-the-genes-based-on-the-results-of-differential-expression-analysis"><i class="fas fa-link"></i></a>
</h5>
<p>The formula for generating the gene ranking from the results table of differential expression analysis is $ (-1) * log_{10}(p) * ()<span class="math inline">\(. Note that "\)</span>p$-value” denotes the non-adjusted p-value.</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_Ensembl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_shrink_Ensembl</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA_Ensembl</span> <span class="op">&lt;-</span> <span class="va">DE_results_DESeq2_shrink_Ensembl</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_Ensembl</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_DESeq2_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Ensembl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_DESeq2_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Ensembl</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># inspect the first entries of the ranking vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Ensembl</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000198692 </span></span>
<span><span class="co">#&gt;      204.454993      105.507250       75.109639 </span></span>
<span><span class="co">#&gt; ENSG00000154620 ENSG00000157828 ENSG00000183878 </span></span>
<span><span class="co">#&gt;       64.872131       39.560339        7.260319 </span></span>
<span><span class="co">#&gt; ENSG00000143632 ENSG00000172543 ENSG00000082397 </span></span>
<span><span class="co">#&gt;        4.867166        4.469387        3.999177 </span></span>
<span><span class="co">#&gt; ENSG00000107731 </span></span>
<span><span class="co">#&gt;        3.973429</span></span></code></pre></div>
</div>
</div>
</div>
<div id="option-3-generate-required-input-using-edger" class="section level3" number="8.1.2">
<h3>
<span class="header-section-number">8.1.2</span> 1.3 Option 3: Generate required input using edgeR<a class="anchor" aria-label="anchor" href="#option-3-generate-required-input-using-edger"><i class="fas fa-link"></i></a>
</h3>
<p>The formula for generating the gene ranking from the results table of differential expression analysis is <span class="math inline">\((-1) * log_{10}(p\text{-value}) * \text{sign}(\text{log fold change})\)</span>. Note that “<span class="math inline">\(p\)</span>-value” denotes the non-adjusted p-value.</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_Ensembl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Ensembl</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA_Ensembl</span> <span class="op">&lt;-</span> <span class="va">DE_results_edgeR_Ensembl</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_Ensembl</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">p_adj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect ranking vector: </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000154620 </span></span>
<span><span class="co">#&gt;             Inf     192.9498029     118.5930433 </span></span>
<span><span class="co">#&gt; ENSG00000172543 ENSG00000120868 ENSG00000110031 </span></span>
<span><span class="co">#&gt;       1.7058094       1.0563194       1.0370201 </span></span>
<span><span class="co">#&gt; ENSG00000198689 ENSG00000065308 ENSG00000095015 </span></span>
<span><span class="co">#&gt;       0.9865108       0.8790009       0.8790009 </span></span>
<span><span class="co">#&gt; ENSG00000228253 </span></span>
<span><span class="co">#&gt;       0.8654283</span></span>
<span></span>
<span><span class="co"># 5. special problem here: gene ENSG00000129824 has a ranking value Inf since its adjusted</span></span>
<span><span class="co"># p-value in the results table of differential expression analysis amounts to 0</span></span>
<span></span>
<span><span class="co"># here, we deal with this issue by resetting this ranking value to the highest ranking value</span></span>
<span><span class="co"># that occurs among the remaining genes</span></span>
<span><span class="co"># -&gt; note that there is NO common way of dealing with this issue</span></span>
<span><span class="va">rankvec_edgeR_Ensembl</span><span class="op">[</span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">==</span> <span class="cn">Inf</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span><span class="op">[</span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">!=</span> <span class="cn">Inf</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect ranking vector once more: </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000154620 </span></span>
<span><span class="co">#&gt;     192.9498029     192.9498029     118.5930433 </span></span>
<span><span class="co">#&gt; ENSG00000172543 ENSG00000120868 ENSG00000110031 </span></span>
<span><span class="co">#&gt;       1.7058094       1.0563194       1.0370201 </span></span>
<span><span class="co">#&gt; ENSG00000198689 ENSG00000065308 ENSG00000095015 </span></span>
<span><span class="co">#&gt;       0.9865108       0.8790009       0.8790009 </span></span>
<span><span class="co">#&gt; ENSG00000228253 </span></span>
<span><span class="co">#&gt;       0.8654283</span></span></code></pre></div>
</div>
<div id="step-2-run-gsea-with-gene-set-database-go" class="section level3" number="8.1.3">
<h3>
<span class="header-section-number">8.1.3</span> step 2: Run GSEA with gene set database GO<a class="anchor" aria-label="anchor" href="#step-2-run-gsea-with-gene-set-database-go"><i class="fas fa-link"></i></a>
</h3>
<p>For the purpose of simplicity, we work with the ranking generated using limma. However, you can switch around at your discretion.</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rankvec_Ensembl</span> <span class="op">&lt;-</span> <span class="va">rankvec_limma_Ensembl</span></span>
<span></span>
<span><span class="co"># alternatively: </span></span>
<span><span class="co">#rankvec_Ensembl &lt;- rankvec_DESeq2_Ensembl</span></span>
<span><span class="co">#rankvec_Ensembl &lt;- rankvec_edgeR_Ensembl</span></span></code></pre></div>
<p>Run GSEA with gene set database GO. Note that it is important to set a seed for reproducibility.</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># important: set seed for reprodubibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSEA_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html">gseGO</a></span><span class="op">(</span>geneList <span class="op">=</span> <span class="va">rankvec_Ensembl</span>,</span>
<span>                   ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                   OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                   keyType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>                   seed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># set seed for reproducibility</span></span>
<span></span>
<span></span>
<span><span class="co"># inspect the corresponding results </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">GSEA_GO</span>, n <span class="op">=</span> <span class="fl">10</span> <span class="op">)</span></span>
<span><span class="co">#&gt;                    ID</span></span>
<span><span class="co">#&gt; GO:0002694 GO:0002694</span></span>
<span><span class="co">#&gt; GO:0050865 GO:0050865</span></span>
<span><span class="co">#&gt; GO:0045321 GO:0045321</span></span>
<span><span class="co">#&gt; GO:0050863 GO:0050863</span></span>
<span><span class="co">#&gt; GO:0001775 GO:0001775</span></span>
<span><span class="co">#&gt; GO:0051249 GO:0051249</span></span>
<span><span class="co">#&gt; GO:0050867 GO:0050867</span></span>
<span><span class="co">#&gt; GO:0051251 GO:0051251</span></span>
<span><span class="co">#&gt; GO:0002696 GO:0002696</span></span>
<span><span class="co">#&gt; GO:0098609 GO:0098609</span></span>
<span><span class="co">#&gt;                                             Description</span></span>
<span><span class="co">#&gt; GO:0002694           regulation of leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0050865                regulation of cell activation</span></span>
<span><span class="co">#&gt; GO:0045321                         leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0050863              regulation of T cell activation</span></span>
<span><span class="co">#&gt; GO:0001775                              cell activation</span></span>
<span><span class="co">#&gt; GO:0051249          regulation of lymphocyte activation</span></span>
<span><span class="co">#&gt; GO:0050867       positive regulation of cell activation</span></span>
<span><span class="co">#&gt; GO:0051251 positive regulation of lymphocyte activation</span></span>
<span><span class="co">#&gt; GO:0002696  positive regulation of leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0098609                           cell-cell adhesion</span></span>
<span><span class="co">#&gt;            setSize enrichmentScore       NES       pvalue</span></span>
<span><span class="co">#&gt; GO:0002694     219      -0.4538535 -1.871372 2.478751e-07</span></span>
<span><span class="co">#&gt; GO:0050865     232      -0.4464877 -1.843713 1.990043e-07</span></span>
<span><span class="co">#&gt; GO:0045321     334      -0.4142950 -1.780532 9.811474e-08</span></span>
<span><span class="co">#&gt; GO:0050863     146      -0.5022111 -1.977090 4.610938e-07</span></span>
<span><span class="co">#&gt; GO:0001775     377      -0.3951803 -1.717809 4.099930e-07</span></span>
<span><span class="co">#&gt; GO:0051249     185      -0.4665593 -1.899849 5.661556e-07</span></span>
<span><span class="co">#&gt; GO:0050867     149      -0.4915790 -1.941660 7.600078e-07</span></span>
<span><span class="co">#&gt; GO:0051251     130      -0.5062219 -1.963565 8.874654e-07</span></span>
<span><span class="co">#&gt; GO:0002696     146      -0.4938384 -1.944128 1.180204e-06</span></span>
<span><span class="co">#&gt; GO:0098609     268      -0.4192255 -1.756435 1.120187e-06</span></span>
<span><span class="co">#&gt;                p.adjust       qvalue rank</span></span>
<span><span class="co">#&gt; GO:0002694 0.0003077782 0.0002683139  841</span></span>
<span><span class="co">#&gt; GO:0050865 0.0003077782 0.0002683139  778</span></span>
<span><span class="co">#&gt; GO:0045321 0.0003077782 0.0002683139  841</span></span>
<span><span class="co">#&gt; GO:0050863 0.0003435149 0.0002994683  841</span></span>
<span><span class="co">#&gt; GO:0001775 0.0003435149 0.0002994683  782</span></span>
<span><span class="co">#&gt; GO:0051249 0.0003514882 0.0003064193  910</span></span>
<span><span class="co">#&gt; GO:0050867 0.0004044327 0.0003525751 1417</span></span>
<span><span class="co">#&gt; GO:0051251 0.0004132261 0.0003602409 1375</span></span>
<span><span class="co">#&gt; GO:0002696 0.0004396259 0.0003832557 1417</span></span>
<span><span class="co">#&gt; GO:0098609 0.0004396259 0.0003832557  778</span></span>
<span><span class="co">#&gt;                              leading_edge</span></span>
<span><span class="co">#&gt; GO:0002694 tags=26%, list=13%, signal=23%</span></span>
<span><span class="co">#&gt; GO:0050865 tags=24%, list=12%, signal=22%</span></span>
<span><span class="co">#&gt; GO:0045321 tags=23%, list=13%, signal=21%</span></span>
<span><span class="co">#&gt; GO:0050863 tags=29%, list=13%, signal=25%</span></span>
<span><span class="co">#&gt; GO:0001775 tags=22%, list=13%, signal=20%</span></span>
<span><span class="co">#&gt; GO:0051249 tags=28%, list=15%, signal=24%</span></span>
<span><span class="co">#&gt; GO:0050867 tags=40%, list=23%, signal=32%</span></span>
<span><span class="co">#&gt; GO:0051251 tags=42%, list=22%, signal=33%</span></span>
<span><span class="co">#&gt; GO:0002696 tags=41%, list=23%, signal=33%</span></span>
<span><span class="co">#&gt; GO:0098609 tags=24%, list=12%, signal=22%</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            core_enrichment</span></span>
<span><span class="co">#&gt; GO:0002694                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0050865                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000154127/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000010278/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0045321                                                                                 ENSG00000121594/ENSG00000125735/ENSG00000177697/ENSG00000120738/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000197405/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000215021/ENSG00000179388/ENSG00000232629/ENSG00000155760/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000168264/ENSG00000012061/ENSG00000205220/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000153879/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000105438/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000168447/ENSG00000196839/ENSG00000071564/ENSG00000184371/ENSG00000159958/ENSG00000131188/ENSG00000163874/ENSG00000198435/ENSG00000105374/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000103522/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000078902/ENSG00000140968/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000160683/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008/ENSG00000177606</span></span>
<span><span class="co">#&gt; GO:0050863                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000112149/ENSG00000104951/ENSG00000131981/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000160223/ENSG00000113302/ENSG00000167604/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0001775 ENSG00000121594/ENSG00000125735/ENSG00000177697/ENSG00000120738/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000154127/ENSG00000204287/ENSG00000197405/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000215021/ENSG00000179388/ENSG00000232629/ENSG00000155760/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000135124/ENSG00000104951/ENSG00000183624/ENSG00000168264/ENSG00000012061/ENSG00000205220/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000153879/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000105438/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000010278/ENSG00000168447/ENSG00000196839/ENSG00000071564/ENSG00000184371/ENSG00000159958/ENSG00000131188/ENSG00000163874/ENSG00000198435/ENSG00000105374/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000103522/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000078902/ENSG00000140968/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000186222/ENSG00000160683/ENSG00000106211/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008/ENSG00000177606</span></span>
<span><span class="co">#&gt; GO:0051249                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000196126/ENSG00000189114/ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000131981/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0050867                                                                                                                                                                                                                                                                                                                                                                 ENSG00000113811/ENSG00000174130/ENSG00000186827/ENSG00000130475/ENSG00000103653/ENSG00000185950/ENSG00000019582/ENSG00000154096/ENSG00000107968/ENSG00000149273/ENSG00000135426/ENSG00000033327/ENSG00000198502/ENSG00000105639/ENSG00000135334/ENSG00000168040/ENSG00000196126/ENSG00000189114/ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000196839/ENSG00000159958/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000125657/ENSG00000185338/ENSG00000130522/ENSG00000105974/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0051251                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000186827/ENSG00000130475/ENSG00000103653/ENSG00000185950/ENSG00000019582/ENSG00000154096/ENSG00000107968/ENSG00000149273/ENSG00000135426/ENSG00000198502/ENSG00000105639/ENSG00000135334/ENSG00000168040/ENSG00000196126/ENSG00000189114/ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000090776/ENSG00000125726/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000196839/ENSG00000159958/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000125657/ENSG00000185338/ENSG00000105974/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0002696                                                                                                                                                                                                                                                                                                                                                                 ENSG00000113811/ENSG00000174130/ENSG00000186827/ENSG00000130475/ENSG00000103653/ENSG00000185950/ENSG00000019582/ENSG00000154096/ENSG00000107968/ENSG00000149273/ENSG00000135426/ENSG00000033327/ENSG00000198502/ENSG00000105639/ENSG00000135334/ENSG00000168040/ENSG00000196126/ENSG00000189114/ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000196839/ENSG00000159958/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000125657/ENSG00000185338/ENSG00000130522/ENSG00000105974/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0098609                                                                                                                                                                                                                                                                                                 ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000148948/ENSG00000183655/ENSG00000154127/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000140678/ENSG00000172977/ENSG00000172270/ENSG00000171236/ENSG00000091972/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000154134/ENSG00000108622/ENSG00000112149/ENSG00000104951/ENSG00000129474/ENSG00000105376/ENSG00000161638/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000108861/ENSG00000164442/ENSG00000117090/ENSG00000197329/ENSG00000130669/ENSG00000204592/ENSG00000179820/ENSG00000107338/ENSG00000010278/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000253958/ENSG00000160223/ENSG00000113302/ENSG00000167642/ENSG00000232810/ENSG00000167604/ENSG00000169992/ENSG00000076826/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000125398/ENSG00000143507/ENSG00000125657/ENSG00000186222/ENSG00000171840/ENSG00000106211/ENSG00000185338/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span></code></pre></div>
<p>Arguments of function gseGO():</p>
<ul>
<li><p>geneList: vector of gene ranking (generated in manner of step 1)</p></li>
<li><p>ont: subontology of GO (must be one of “BP”, “CC”, “MF”)</p></li>
<li><p>OrgDb: indicates organism from which the gene expression measurements are taken</p></li>
<li><p>keyTypes: for our purposes here, argument keyType indicates the gene ID format of the gene names in vector rankvec</p></li>
<li><p>available keytypes can be found in the following:</p></li>
</ul>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">keytypes</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"     </span></span>
<span><span class="co">#&gt;  [4] "ENSEMBLPROT"  "ENSEMBLTRANS" "ENTREZID"    </span></span>
<span><span class="co">#&gt;  [7] "ENZYME"       "EVIDENCE"     "EVIDENCEALL" </span></span>
<span><span class="co">#&gt; [10] "GENENAME"     "GENETYPE"     "GO"          </span></span>
<span><span class="co">#&gt; [13] "GOALL"        "IPI"          "MAP"         </span></span>
<span><span class="co">#&gt; [16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL" </span></span>
<span><span class="co">#&gt; [19] "PATH"         "PFAM"         "PMID"        </span></span>
<span><span class="co">#&gt; [22] "PROSITE"      "REFSEQ"       "SYMBOL"      </span></span>
<span><span class="co">#&gt; [25] "UCSCKG"       "UNIPROT"</span></span></code></pre></div>
<ul>
<li><p>seed = TRUE: set the seed you have indicated above (here: seed = 1 )</p></li>
<li><p>additional arguments (these do not appear in the above code lines since we keep the default options)</p></li>
<li><p>exponent 1: in the calculation of the enrichment score, each gene is weighted by its absolute value of the ranking metric</p></li>
<li><p>eps = 1e-10: each p-value that is smaller than 1e-10 is indicated as 1e-10</p></li>
<li><p>pvalueCutoff = 0.05: only those gene sets with a p-value &lt; 0.05 are indicated in the results. Note that if you want to inspect ALL gene sets in the results, set pvalueCutoff = 1</p></li>
<li><p>pAdjustMethod = “BH”: adjustment for multiple testing using the method by Benjamini and Hochberg</p></li>
</ul>
</div>
<div id="step-3-interpretation-of-the-results" class="section level3" number="8.1.4">
<h3>
<span class="header-section-number">8.1.4</span> step 3: Interpretation of the results<a class="anchor" aria-label="anchor" href="#step-3-interpretation-of-the-results"><i class="fas fa-link"></i></a>
</h3>
<p>Columns in the results table:</p>
<ul>
<li><p>ID: ID of gene set</p></li>
<li><p>Description: Description of Gene Set</p></li>
<li><p>setSize: Size of the gene set</p></li>
<li><p>enrichmentScore: Enrichment Score
–&gt; note: this enrichment score has not been normalized for gene set size
–&gt; means: larger gene sets automatically have a bigger (absolute) enrichment score,
–&gt; independent of actual differential enrichment
–&gt; the raw enrichment score is therefore not comparable between different gene sets</p></li>
<li><p>NES: Normalized version of column enrichmentScore
-&gt; NES can be compared between different gene sets</p></li>
<li><p>pvalue: p-value of enrichment of given gene set
-&gt; note: this raw p-value has not been adjusted for multiple testing
-&gt; therefore: CANNOT be used to assess differential enrichment of a given gene set</p></li>
<li><p>p.adjust: ADJUSTED p-value of a given gene set. Note that this p-value now has been adjusted for multiple testing and can therefore be used to assess differential enrichment. For instance, detect all genes with p.adjust &lt; 0.05 as differentially enriched.</p></li>
<li><p>qvalue: ADJUSTED p-value of a given gene set. Note that a qvalue is the analog to p.adjust, but adjusted for multiple testing
using a different method.</p></li>
</ul>
<p>-&gt;&gt; you can either use the column p.adjust or qvalue to assess whether a gene set is differentially enriched or not.</p>
<ul>
<li><p>rank: position in the ranked list of the genes in which the maximum difference between the two sums occurs, i.e. the rank at which the enrichment score is extracted.</p></li>
<li>
<p>leading_edge:</p>
<ul>
<li><p>tags: The percentage of genes in the ranked list that are members of gene set before (for positive enrichment score)
or after (for negative enrichment score) the position from which the enrichment score is extracted.</p></li>
<li><p>list: The percentage of genes in the ranked gene list before (for positive enrichment
score) or after (for negative enrichment score) the position from which the enrichment score is extracted.</p></li>
<li><p>signal: enrichment signal strength
combines the statistics “tags” and “list”</p></li>
<li><p>core_enrichment: genes that contribute most to the enrichment results</p></li>
</ul>
</li>
</ul>
<p>Note that a more detailed information on leading_edge and core_enrichment can be found in the user manual provided for GSEA:
<a href="https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html" class="uri">https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html</a> :</p>
<ul>
<li><p>search for Detailed Enrichment Results (section “Leading Edge”)</p></li>
<li><p>search for CORE ENRICHMENT</p></li>
</ul>
</div>
</div>
<div id="researchers-degrees-of-freedom-1" class="section level2" number="8.2">
<h2>
<span class="header-section-number">8.2</span> Researchers’ degrees of freedom<a class="anchor" aria-label="anchor" href="#researchers-degrees-of-freedom-1"><i class="fas fa-link"></i></a>
</h2>
<p>In this part, we will go through all parameters that can be adapted in the GOSeq workflow. It is important to note that the intention behind going through the researchers’ degrees of freedom is to give you an understanding of what you can do to adapt the given (parameter) setting to the research question. It is even more important to keep in mind that the intention behind going through these flexible parameters is NOT to change them in order to help you obtain the most preferable results by systematically changing these parameters since such behaviour would correspond to “cherry-picking”. Any changes in the parameter choice must be documented transparently.</p>
<div id="change-1-change-exponent" class="section level3" number="8.2.1">
<h3>
<span class="header-section-number">8.2.1</span> change 1: change exponent<a class="anchor" aria-label="anchor" href="#change-1-change-exponent"><i class="fas fa-link"></i></a>
</h3>
<p>Note that, in contrast to the web-based tool GSEA, clusterProfiler does not make any indication as to which alternative exponent values are allowed or sensible. oOne could stick with the web-based tool GSEA which suggests, in addition to the default exponent value of 1, the alternative values 0, 1.5, and 2.</p>
<p>Here, we change the exponent value to 2</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># important: set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSEA_GO_exponent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html">gseGO</a></span><span class="op">(</span>geneList <span class="op">=</span> <span class="va">rankvec_Ensembl</span>,</span>
<span>                            ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                            OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                            keyType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>                            exp <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                            seed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect the results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">GSEA_GO_exponent</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] ID              Description     setSize        </span></span>
<span><span class="co">#&gt; [4] enrichmentScore NES             pvalue         </span></span>
<span><span class="co">#&gt; [7] p.adjust        qvalue         </span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre></div>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="go-seq.html"><span class="header-section-number">7</span> Go SEQ</a></div>
<div class="next"><a href="david.html"><span class="header-section-number">9</span> DAVID</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li>
<a class="nav-link" href="#cluster-profiler-gsea-go"><span class="header-section-number">8</span> Cluster Profiler: GSEA GO</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#libraries-6"><span class="header-section-number">8.0.1</span> Libraries</a></li>
<li><a class="nav-link" href="#load-the-data"><span class="header-section-number">8.0.2</span> Load the data</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#run-gsea"><span class="header-section-number">8.1</span> Run GSEA</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#step-1-generation-of-required-input-object"><span class="header-section-number">8.1.1</span> step 1: Generation of Required Input Object</a></li>
<li><a class="nav-link" href="#option-3-generate-required-input-using-edger"><span class="header-section-number">8.1.2</span> 1.3 Option 3: Generate required input using edgeR</a></li>
<li><a class="nav-link" href="#step-2-run-gsea-with-gene-set-database-go"><span class="header-section-number">8.1.3</span> step 2: Run GSEA with gene set database GO</a></li>
<li><a class="nav-link" href="#step-3-interpretation-of-the-results"><span class="header-section-number">8.1.4</span> step 3: Interpretation of the results</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#researchers-degrees-of-freedom-1"><span class="header-section-number">8.2</span> Researchers’ degrees of freedom</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#change-1-change-exponent"><span class="header-section-number">8.2.1</span> change 1: change exponent</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>RNA Sequencing Guide</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-19.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
