<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 13 PADOG | RNA Sequencing Guide</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="13.1 Libraries  13.1.1 Install Libraries All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine. The code below will...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 13 PADOG | RNA Sequencing Guide">
<meta property="og:type" content="book">
<meta property="og:description" content="13.1 Libraries  13.1.1 Install Libraries All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine. The code below will...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 13 PADOG | RNA Sequencing Guide">
<meta name="twitter:description" content="13.1 Libraries  13.1.1 Install Libraries All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine. The code below will...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">RNA Sequencing Guide</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="prefiltering.html"><span class="header-section-number">2</span> Prefiltering</a></li>
<li><a class="" href="gsea-preranking.html"><span class="header-section-number">3</span> GSEA Preranking</a></li>
<li><a class="" href="gene-id-duplicate-removals.html"><span class="header-section-number">4</span> Gene ID Duplicate Removals</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">5</span> Differential Expression Analysis</a></li>
<li><a class="" href="rna-seq-transformation.html"><span class="header-section-number">6</span> RNA-Seq Transformation</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="go-seq.html"><span class="header-section-number">7</span> Go SEQ</a></li>
<li><a class="" href="cluster-profiler-gsea-go.html"><span class="header-section-number">8</span> Cluster Profiler: GSEA GO</a></li>
<li><a class="" href="david.html"><span class="header-section-number">9</span> DAVID</a></li>
<li><a class="" href="gsea-web.html"><span class="header-section-number">10</span> GSEA WEB</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="cluster-profiler-gsea-kegg.html"><span class="header-section-number">11</span> Cluster Profiler: GSEA KEGG</a></li>
<li><a class="" href="cluster-profiler-ora.html"><span class="header-section-number">12</span> Cluster Profiler: ORA</a></li>
<li><a class="active" href="padog.html"><span class="header-section-number">13</span> PADOG</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="padog" class="section level1" number="13">
<h1>
<span class="header-section-number">13</span> PADOG<a class="anchor" aria-label="anchor" href="#padog"><i class="fas fa-link"></i></a>
</h1>
<div id="libraries-10" class="section level2" number="13.1">
<h2>
<span class="header-section-number">13.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-10"><i class="fas fa-link"></i></a>
</h2>
<div id="install-libraries-1" class="section level3" number="13.1.1">
<h3>
<span class="header-section-number">13.1.1</span> Install Libraries<a class="anchor" aria-label="anchor" href="#install-libraries-1"><i class="fas fa-link"></i></a>
</h3>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine. The code below will install {BiocManager} from CRAN, and you can then use this package to install PADOG, tweeDEseqCountDatam, and KEGGREST to your system.</p>
<div class="sourceCode" id="cb178"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"PADOG"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"tweeDEseqCountData"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"KEGGREST"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="load-libraries-7" class="section level3" number="13.1.2">
<h3>
<span class="header-section-number">13.1.2</span> Load Libraries<a class="anchor" aria-label="anchor" href="#load-libraries-7"><i class="fas fa-link"></i></a>
</h3>
<p>Note that loading these libraries will mask many functions from base R packages. If you run into unexpected errors on functions you’re using, it is recommended to use namespacing to explicitly clarify the package from which you need a given function. (I have suppressed the library() loading messages from this document, however.)</p>
<div class="sourceCode" id="cb179"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">PADOG</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.creal.cat/jrgonzalez/software.htm">tweeDEseqCountData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">KEGGREST</span><span class="op">)</span></span></code></pre></div>
<p>Provide a brief note on what these libraries are for?</p>
</div>
<div id="padog-1" class="section level3" number="13.1.3">
<h3>
<span class="header-section-number">13.1.3</span> PADOG<a class="anchor" aria-label="anchor" href="#padog-1"><i class="fas fa-link"></i></a>
</h3>
<p>The PADOG library does XYZ. You can find more information about it online at LINK.</p>
</div>
<div id="tweedeseqcountdata-1" class="section level3" number="13.1.4">
<h3>
<span class="header-section-number">13.1.4</span> tweeDEseqCountData<a class="anchor" aria-label="anchor" href="#tweedeseqcountdata-1"><i class="fas fa-link"></i></a>
</h3>
</div>
<div id="keggrest" class="section level3" number="13.1.5">
<h3>
<span class="header-section-number">13.1.5</span> KEGGREST<a class="anchor" aria-label="anchor" href="#keggrest"><i class="fas fa-link"></i></a>
</h3>
</div>
</div>
<div id="load-data-6" class="section level2" number="13.2">
<h2>
<span class="header-section-number">13.2</span> Load Data<a class="anchor" aria-label="anchor" href="#load-data-6"><i class="fas fa-link"></i></a>
</h2>
<p>This section will change substantially when I reconfigure the project as an R package/book.</p>
<p>For now, place any data you need into the <code>./data</code> directory.</p>
<div class="sourceCode" id="cb180"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we load the voom-transformed Pickrell data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/expression_data_voomtransformed_Entrez.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># alternatively: load the gene expression measurements that have been transformed using </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/expression_data_vsttransformed_Entrez.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># additionally, we load the pickrell data set so that we can access the sample conditions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pickrell</span><span class="op">)</span></span></code></pre></div>
<p>The sample conditions (i.e. phenotype labels) of the pickrell data set can be accessed using</p>
<div class="sourceCode" id="cb181"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span> </span>
<span><span class="co">#&gt;  [1] male   male   female male   female male   female male  </span></span>
<span><span class="co">#&gt;  [9] female male   female male   female male   female male  </span></span>
<span><span class="co">#&gt; [17] female female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [25] female male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [33] female male   female female female female male   female</span></span>
<span><span class="co">#&gt; [41] male   male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [49] female female male   female male   male   female female</span></span>
<span><span class="co">#&gt; [57] male   female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [65] female female female male   female</span></span>
<span><span class="co">#&gt; Levels: female male</span></span></code></pre></div>
<p>We proceed with the voom-transformed pickrell data set and the corresponding phenotype labels</p>
<div class="sourceCode" id="cb182"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># gene expression measurements (transformed)</span></span>
<span><span class="co"># note: you can also proceed with the vst-transformed gene expression measurements </span></span>
<span><span class="va">expression_data_transformed</span> <span class="op">&lt;-</span> <span class="va">expression_data_voomtransformed_Entrez</span></span>
<span><span class="co"># sample conditions</span></span>
<span><span class="va">sample_conditions</span> <span class="op">&lt;-</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span></span></code></pre></div>
</div>
<div id="prepare-sample-conditions" class="section level2" number="13.3">
<h2>
<span class="header-section-number">13.3</span> Prepare Sample Conditions<a class="anchor" aria-label="anchor" href="#prepare-sample-conditions"><i class="fas fa-link"></i></a>
</h2>
<p>First, we inspect the form of the initial (raw) sample conditions</p>
<div class="sourceCode" id="cb183"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## look at the class: </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">sample_conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "factor"</span></span>
<span><span class="co"># -&gt; the sample labels are already coded as factor</span></span>
<span></span>
<span><span class="co"># the current levels are:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">levels</a></span><span class="op">(</span><span class="va">sample_conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "female" "male"</span></span></code></pre></div>
<p>PADOG requires character vector with class labels of the samples. It can only contain “c” for control samples or “d” for disease samples</p>
<div class="sourceCode" id="cb184"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># prepare sample conditions</span></span>
<span><span class="co"># we want to convert </span></span>
<span><span class="co"># (i) "female" to "c"</span></span>
<span><span class="co"># (ii) "male" to "d"</span></span>
<span><span class="va">sample_conditions_prep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">sample_conditions</span>, </span>
<span>                                levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"female"</span>,<span class="st">"male"</span><span class="op">)</span>, </span>
<span>                                labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"c"</span>,<span class="st">"d"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="run-padog" class="section level2" number="13.4">
<h2>
<span class="header-section-number">13.4</span> Run PADOG<a class="anchor" aria-label="anchor" href="#run-padog"><i class="fas fa-link"></i></a>
</h2>
<p><em>It is recommended to set a seed to ensure exact reproducibility of the results if the code is run at multiple time points</em></p>
<p>you can specify any integer number as the seed. It is VERY IMPORTANT to choose the seed arbitrarily and WITHOUT INSPECTING the results the seed should NEVER be specified based on which value yields the most preferable results.</p>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run PADOG: </span></span>
<span> <span class="va">PADOG_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PADOG/man/padog.html">padog</a></span><span class="op">(</span>esetm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">expression_data_transformed</span><span class="op">)</span>, </span>
<span>                        group <span class="op">=</span> <span class="va">sample_conditions_prep</span>, </span>
<span>                        dseed <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>arguments:</p>
<ul>
<li>
<code>esetm</code>: matrix that contains the expression measurements
<ul>
<li>note: since the expression data is initially stored in a data frame, we transform it to a matrix when running PADOG</li>
</ul>
</li>
<li>
<code>group</code>: sample conditions (has values “c” and “d”)</li>
<li>
<code>dseed</code>: seed for random number generation (used in the process of phenotype permutation)</li>
</ul>
<p>additional arguments:</p>
<ul>
<li>
<code>paired</code>: indicates whether the samples in both groups are paired</li>
<li>
<code>block</code>: if the samples are paired (i.e. argument paired = TRUE), then the paired samples must have the same block value</li>
<li>
<code>gslist</code>: gives instructions on how to cluster the genes into gene sets
<ul>
<li>gslist = “KEGGRESTpathway”: gene sets correspond to KEGG pathways</li>
<li>alternative: provide a user-defined list of gene sets</li>
</ul>
</li>
<li>
<code>organism</code>: organism from which the gene expression measurements are taken
<ul>
<li>for human, set organism = “hsa”</li>
<li>the required character value for other organisms can be extracted from the KEGGREST package:</li>
</ul>
</li>
<li>
<code>obtain</code> required organisms from column organism</li>
<li>
<code>annotation</code>: required if gslist is set to “KEGGRESTpathway” and the rownames of esetm are probe IDs</li>
<li>can be set to NULL of gslist is set to “KEGGRESTpathway” and the rownames of esetm are in the Entrez gene ID format</li>
<li>if rownames are other gene IDs, then sett annotation = NULL and make sure that the rownames are elements of gslist (and unique!)</li>
<li>
<code>gs.names</code>: contains names of gene sets -&gt; character vector
<ul>
<li>must have the same length as gslist</li>
</ul>
</li>
<li>
<code>NI</code>: number of phenotype permutations employed in the assessment of the significance of a given gene set</li>
</ul>
</div>
<div id="adjust-for-multiple-testing" class="section level2" number="13.5">
<h2>
<span class="header-section-number">13.5</span> Adjust for Multiple Testing<a class="anchor" aria-label="anchor" href="#adjust-for-multiple-testing"><i class="fas fa-link"></i></a>
</h2>
<!-- I haven't included the content from these parts yet, but I hope this example is clear enough! -->
</div>
<div id="interpretation-of-results" class="section level2" number="13.6">
<h2>
<span class="header-section-number">13.6</span> Interpretation of Results<a class="anchor" aria-label="anchor" href="#interpretation-of-results"><i class="fas fa-link"></i></a>
</h2>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="cluster-profiler-ora.html"><span class="header-section-number">12</span> Cluster Profiler: ORA</a></div>
<div class="empty"></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#padog"><span class="header-section-number">13</span> PADOG</a></li>
<li>
<a class="nav-link" href="#libraries-10"><span class="header-section-number">13.1</span> Libraries</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#install-libraries-1"><span class="header-section-number">13.1.1</span> Install Libraries</a></li>
<li><a class="nav-link" href="#load-libraries-7"><span class="header-section-number">13.1.2</span> Load Libraries</a></li>
<li><a class="nav-link" href="#padog-1"><span class="header-section-number">13.1.3</span> PADOG</a></li>
<li><a class="nav-link" href="#tweedeseqcountdata-1"><span class="header-section-number">13.1.4</span> tweeDEseqCountData</a></li>
<li><a class="nav-link" href="#keggrest"><span class="header-section-number">13.1.5</span> KEGGREST</a></li>
</ul>
</li>
<li><a class="nav-link" href="#load-data-6"><span class="header-section-number">13.2</span> Load Data</a></li>
<li><a class="nav-link" href="#prepare-sample-conditions"><span class="header-section-number">13.3</span> Prepare Sample Conditions</a></li>
<li><a class="nav-link" href="#run-padog"><span class="header-section-number">13.4</span> Run PADOG</a></li>
<li><a class="nav-link" href="#adjust-for-multiple-testing"><span class="header-section-number">13.5</span> Adjust for Multiple Testing</a></li>
<li><a class="nav-link" href="#interpretation-of-results"><span class="header-section-number">13.6</span> Interpretation of Results</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>RNA Sequencing Guide</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-19.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
