<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 2 Prefiltering | RNA Sequencing Guide</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across all samples. We distinguish between these two approaches since the...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 2 Prefiltering | RNA Sequencing Guide">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across all samples. We distinguish between these two approaches since the...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 2 Prefiltering | RNA Sequencing Guide">
<meta name="twitter:description" content="In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across all samples. We distinguish between these two approaches since the...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">RNA Sequencing Guide</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="active" href="prefiltering.html"><span class="header-section-number">2</span> Prefiltering</a></li>
<li><a class="" href="gsea-preranking.html"><span class="header-section-number">3</span> GSEA Preranking</a></li>
<li><a class="" href="gene-id-duplicate-removals.html"><span class="header-section-number">4</span> Gene ID Duplicate Removals</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">5</span> Differential Expression Analysis</a></li>
<li><a class="" href="rna-seq-transformation.html"><span class="header-section-number">6</span> RNA-Seq Transformation</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="go-seq.html"><span class="header-section-number">7</span> Go SEQ</a></li>
<li><a class="" href="cluster-profiler-gsea-go.html"><span class="header-section-number">8</span> Cluster Profiler: GSEA GO</a></li>
<li><a class="" href="david.html"><span class="header-section-number">9</span> DAVID</a></li>
<li><a class="" href="gsea-web.html"><span class="header-section-number">10</span> GSEA WEB</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="cluster-profiler-gsea-kegg.html"><span class="header-section-number">11</span> Cluster Profiler: GSEA KEGG</a></li>
<li><a class="" href="cluster-profiler-ora.html"><span class="header-section-number">12</span> Cluster Profiler: ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">13</span> PADOG</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="prefiltering" class="section level1" number="2">
<h1>
<span class="header-section-number">2</span> Prefiltering<a class="anchor" aria-label="anchor" href="#prefiltering"><i class="fas fa-link"></i></a>
</h1>
<ul>
<li><p>In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across
all samples. We distinguish between these two approaches since the different methods for differential expression analysis (see file “Instructions_Differential_Expression_Analysis”) propose different methods for pre-filtering.</p></li>
<li><p>option 1: pre-filtering using a function provided by edgeR</p></li>
<li><p>option 2: pre-filtering proposed by DESeq2</p></li>
</ul>
<div id="libraries" class="section level2" number="2.1">
<h2>
<span class="header-section-number">2.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"tweeDESeqCountData"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"edgeR"</span><span class="op">)</span></span></code></pre></div>
<div id="load-libraries" class="section level3" number="2.1.1">
<h3>
<span class="header-section-number">2.1.1</span> Load Libraries<a class="anchor" aria-label="anchor" href="#load-libraries"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.creal.cat/jrgonzalez/software.htm">tweeDEseqCountData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span></span></code></pre></div>
<p>Description of the packages:</p>
<ul>
<li><p>tweeDESeqCountData: From this library we obtain the gene expression data set we will use for our illustrations.</p></li>
<li><p>edgeR: edgeR offers a function for pre-filtering we will use below.</p></li>
</ul>
</div>
</div>
<div id="preparation-of-rna-seq-data-set-used-for-illustration" class="section level2" number="2.2">
<h2>
<span class="header-section-number">2.2</span> Preparation of RNA-Seq data set used for illustration<a class="anchor" aria-label="anchor" href="#preparation-of-rna-seq-data-set-used-for-illustration"><i class="fas fa-link"></i></a>
</h2>
<p>The RNA-Seq data set we will use for this illustration is provided by Pickrell et al. (2010) and is part of the library tweeDEseqCountData. In this data set, the sample conditions (i.e. phenotypes) of the respective samples correspond to the genders. For the purpose of simplicity and readability, we store the gene expression measurements and sample conditions in objects with neutral names.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load pickrell data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pickrell</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># access and store gene expression measurements</span></span>
<span><span class="va">expression_data</span> <span class="op">&lt;-</span> <span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">pickrell.eset</span><span class="op">)</span></span>
<span><span class="co"># access and store sample conditions </span></span>
<span><span class="va">sample_conditions</span> <span class="op">&lt;-</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span> </span>
<span></span>
<span></span>
<span><span class="co"># take a look at the gene expression measurements: </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expression_data</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 NA18486 NA18498 NA18499 NA18501 NA18502</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      22     105      40      55      67</span></span>
<span><span class="co">#&gt; ENSG00000000457      22     100     107      53      72</span></span>
<span><span class="co">#&gt; ENSG00000000460       5      23      10      18      15</span></span>
<span><span class="co">#&gt;                 NA18504 NA18505 NA18507 NA18508 NA18510</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       5       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      37      88     127      70      43</span></span>
<span><span class="co">#&gt; ENSG00000000457      38      98      69      66      43</span></span>
<span><span class="co">#&gt; ENSG00000000460       8      11      16      18       7</span></span>
<span><span class="co">#&gt;                 NA18511 NA18516 NA18517 NA18519 NA18520</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419     104      42      66      58      82</span></span>
<span><span class="co">#&gt; ENSG00000000457     103      51      91      87      77</span></span>
<span><span class="co">#&gt; ENSG00000000460       9       7      19      15      22</span></span>
<span><span class="co">#&gt;                 NA18522 NA18523 NA18852 NA18853 NA18855</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      95      40      32      30      34</span></span>
<span><span class="co">#&gt; ENSG00000000457      87      39      65      50      43</span></span>
<span><span class="co">#&gt; ENSG00000000460      17       7      14       8       5</span></span>
<span><span class="co">#&gt;                 NA18856 NA18858 NA18861 NA18862 NA18870</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       1       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      44      47     113      79      38</span></span>
<span><span class="co">#&gt; ENSG00000000457      48      52     117      73      50</span></span>
<span><span class="co">#&gt; ENSG00000000460      13      15      23      15       8</span></span>
<span><span class="co">#&gt;                 NA18871 NA18909 NA18912 NA18913 NA18916</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      41      95      45      54      49</span></span>
<span><span class="co">#&gt; ENSG00000000457      72      93      44      73      81</span></span>
<span><span class="co">#&gt; ENSG00000000460       8      17       2       7       8</span></span>
<span><span class="co">#&gt;                 NA19093 NA19098 NA19099 NA19101 NA19102</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      80      86      73      59      36</span></span>
<span><span class="co">#&gt; ENSG00000000457     103     158     118      89      15</span></span>
<span><span class="co">#&gt; ENSG00000000460      15      15      25      13       2</span></span>
<span><span class="co">#&gt;                 NA19108 NA19114 NA19116 NA19119 NA19127</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419     124      56      64      25     104</span></span>
<span><span class="co">#&gt; ENSG00000000457     108      19     150      16      38</span></span>
<span><span class="co">#&gt; ENSG00000000460      16       6      28       2      18</span></span>
<span><span class="co">#&gt;                 NA19128 NA19130 NA19131 NA19137 NA19138</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       1</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      94     117      62      43      46</span></span>
<span><span class="co">#&gt; ENSG00000000457      50     104     113      71     120</span></span>
<span><span class="co">#&gt; ENSG00000000460      22      25      19      14      10</span></span>
<span><span class="co">#&gt;                 NA19140 NA19143 NA19144 NA19147 NA19152</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      94      82      56      96      66</span></span>
<span><span class="co">#&gt; ENSG00000000457      46     106      74     116     102</span></span>
<span><span class="co">#&gt; ENSG00000000460      21       6      16      13      24</span></span>
<span><span class="co">#&gt;                 NA19153 NA19159 NA19160 NA19171 NA19172</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      78      24      63      34      66</span></span>
<span><span class="co">#&gt; ENSG00000000457     119      29      64      55      70</span></span>
<span><span class="co">#&gt; ENSG00000000460      21       7       9      15      17</span></span>
<span><span class="co">#&gt;                 NA19190 NA19192 NA19193 NA19200 NA19201</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       2</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       1       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      70      85      43      27      77</span></span>
<span><span class="co">#&gt; ENSG00000000457     102      89      31      68      96</span></span>
<span><span class="co">#&gt; ENSG00000000460       8      28      10       5      16</span></span>
<span><span class="co">#&gt;                 NA19203 NA19204 NA19209 NA19210 NA19222</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      82      57      63      89      60</span></span>
<span><span class="co">#&gt; ENSG00000000457      58      86     113      48      71</span></span>
<span><span class="co">#&gt; ENSG00000000460      10      19      17      12      12</span></span>
<span><span class="co">#&gt;                 NA19225 NA19238 NA19239 NA19257</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      76      69      84      76</span></span>
<span><span class="co">#&gt; ENSG00000000457      81      73      87      81</span></span>
<span><span class="co">#&gt; ENSG00000000460       7      21      35      11</span></span>
<span><span class="co"># take a look at the sample conditions:</span></span>
<span><span class="va">sample_conditions</span></span>
<span><span class="co">#&gt;  [1] male   male   female male   female male   female male  </span></span>
<span><span class="co">#&gt;  [9] female male   female male   female male   female male  </span></span>
<span><span class="co">#&gt; [17] female female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [25] female male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [33] female male   female female female female male   female</span></span>
<span><span class="co">#&gt; [41] male   male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [49] female female male   female male   male   female female</span></span>
<span><span class="co">#&gt; [57] male   female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [65] female female female male   female</span></span>
<span><span class="co">#&gt; Levels: female male</span></span>
<span></span>
<span><span class="co"># inspect the number of genes (rows) and the number of samples (columns) in the gene expression data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 52580    69</span></span></code></pre></div>
<div id="option-1-pre-filtering-using-edgers-builting-function" class="section level3" number="2.2.1">
<h3>
<span class="header-section-number">2.2.1</span> Option 1: Pre-Filtering using edgeR’s builting function<a class="anchor" aria-label="anchor" href="#option-1-pre-filtering-using-edgers-builting-function"><i class="fas fa-link"></i></a>
</h3>
<p>This approach works with the function filterByExpr() from the package edgeR and is the proposed method for pre-filtering for the methods for differential expression analysis edgeR and voom/limma. This approach operates on the cpm-transformed count data (cpm: counts-per-million), which excludes all genes that do NOT have a certain number of counts-per-million in a certain number of samples. Note that this approach accounts for differences in the library size between the different samples.</p>
<div id="step-1-generate-input-object-required-by-filterbyexpr" class="section level4" number="2.2.1.1">
<h4>
<span class="header-section-number">2.2.1.1</span> step 1: Generate input object required by filterByExpr()<a class="anchor" aria-label="anchor" href="#step-1-generate-input-object-required-by-filterbyexpr"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_filterByExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">expression_data</span>, </span>
<span>                                        group <span class="op">=</span> <span class="va">sample_conditions</span><span class="op">)</span></span></code></pre></div>
<p>Function description:</p>
<ul>
<li>DGEList(): object to contain RNA sequencing measurements and additional information</li>
</ul>
<p>Argument description:</p>
<ul>
<li>counts: matrix of RNA-Seq data</li>
<li>group: vector that contains the condition of each sample</li>
</ul>
</div>
<div id="step-2-pre-filtering" class="section level4" number="2.2.1.2">
<h4>
<span class="header-section-number">2.2.1.2</span> step 2: Pre-filtering<a class="anchor" aria-label="anchor" href="#step-2-pre-filtering"><i class="fas fa-link"></i></a>
</h4>
<p>The function filterByExpr() creates an indicator which on which genes do and which do not have a sufficient amount of read counts across all samples. Based on this indicator, the gene expression data set is then filtered.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (i) for each gene, indicate if it fulfils the requirements to be kept for the subsequent analysis </span></span>
<span><span class="va">indicator_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">expression_data_filterByExpr</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (ii) filter the gene expression data set such that only those genes are kept which fulfill the requirements</span></span>
<span><span class="va">expression_data_filterByExpr</span> <span class="op">&lt;-</span> <span class="va">expression_data_filterByExpr</span><span class="op">[</span><span class="va">indicator_keep</span>,, keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span></code></pre></div>
<p>Note that the index “keep.lib.sizes = FALSE” ensures that the library size of each sample is recalculated after pre-filtering.</p>
</div>
<div id="step-3-obtain-final-pre-filtered-gene-expression-data-set" class="section level4" number="2.2.1.3">
<h4>
<span class="header-section-number">2.2.1.3</span> step 3: Obtain final pre-filtered gene expression data set<a class="anchor" aria-label="anchor" href="#step-3-obtain-final-pre-filtered-gene-expression-data-set"><i class="fas fa-link"></i></a>
</h4>
<p>At this point, we transform the gene expression measurements back to a data frame. The reason for this is that some subsequent steps (such as the conversion of gene IDs do not work with the DGEList format).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_filterByExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">expression_data_filterByExpr</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect the number of genes and the number of samples in the final pre-filtered gene expression data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression_data_filterByExpr</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6246</span></span></code></pre></div>
<p>Observe that the number of genes has been reduced compared to the original (unfiltered) gene expression data set.</p>
</div>
</div>
<div id="option-2-simpler-pre-filtering-approach-proposed-by-deseq2" class="section level3" number="2.2.2">
<h3>
<span class="header-section-number">2.2.2</span> Option 2: Simpler Pre-filtering approach (proposed by DESeq2)<a class="anchor" aria-label="anchor" href="#option-2-simpler-pre-filtering-approach-proposed-by-deseq2"><i class="fas fa-link"></i></a>
</h3>
<p>A simpler approach for pre-filtering has been proposed by the method for differential expression analysis DESeq2 (see file “Instructions_Differential_Expression_Analysis”. In this approach, only those genes are kept for further analysis that have a pre-specified number of counts X (such as 10) across all samples. A higher value of X thereby leads to more genes being removed.<br>
Note that DESeq2 proposes a stricter version of pre-filtering in which those genes are kept which have X number of counts in at least Y samples.<br>
Note that none of the these two “simpler” approaches to pre-filtering take differences in library size into account.</p>
<div id="step-1-pre-filtering" class="section level4" number="2.2.2.1">
<h4>
<span class="header-section-number">2.2.2.1</span> step 1: Pre-Filtering<a class="anchor" aria-label="anchor" href="#step-1-pre-filtering"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># indicate which genes have at least 10 read counts across all samples:</span></span>
<span><span class="va">indicator_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span> <span class="va">expression_data</span> <span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span> </span>
<span></span>
<span><span class="co"># alternative (and more strict) indicator:</span></span>
<span><span class="co"># indicator_keep &lt;- rowSums( expression_data &gt;=10) &gt;= 10</span></span>
<span></span>
<span><span class="co"># subset gene expression data set accordingly </span></span>
<span><span class="va">expression_data_filterDESEq2</span> <span class="op">&lt;-</span> <span class="va">expression_data</span><span class="op">[</span><span class="va">indicator_keep</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div id="step-2-inspect-final-pre-filtered-gene-expression-data-set" class="section level4" number="2.2.2.2">
<h4>
<span class="header-section-number">2.2.2.2</span> step 2: Inspect final pre-filtered gene expression data set<a class="anchor" aria-label="anchor" href="#step-2-inspect-final-pre-filtered-gene-expression-data-set"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">expression_data_filterDESEq2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 10151    69</span></span></code></pre></div>
<p>Note that the number of genes in the gene expression data set has decreased compared to the initial (unfiltered) gene expression data set.</p>

</div>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="index.html"><span class="header-section-number">1</span> About</a></div>
<div class="next"><a href="gsea-preranking.html"><span class="header-section-number">3</span> GSEA Preranking</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#prefiltering"><span class="header-section-number">2</span> Prefiltering</a></li>
<li>
<a class="nav-link" href="#libraries"><span class="header-section-number">2.1</span> Libraries</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#load-libraries"><span class="header-section-number">2.1.1</span> Load Libraries</a></li></ul>
</li>
<li>
<a class="nav-link" href="#preparation-of-rna-seq-data-set-used-for-illustration"><span class="header-section-number">2.2</span> Preparation of RNA-Seq data set used for illustration</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-pre-filtering-using-edgers-builting-function"><span class="header-section-number">2.2.1</span> Option 1: Pre-Filtering using edgeR’s builting function</a></li>
<li><a class="nav-link" href="#option-2-simpler-pre-filtering-approach-proposed-by-deseq2"><span class="header-section-number">2.2.2</span> Option 2: Simpler Pre-filtering approach (proposed by DESeq2)</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>RNA Sequencing Guide</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-19.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
