% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{book}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{booktabs}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage[]{natbib}
\bibliographystyle{apalike}
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={RNA Sequencing Guide},
  pdfauthor={Milena Wünsch and Pat Callahan},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{RNA Sequencing Guide}
\author{Milena Wünsch and Pat Callahan}
\date{2023-07-02}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\hypertarget{about}{%
\chapter{About}\label{about}}

This is a \emph{sample} book written in \textbf{Markdown}. You can use anything that Pandoc's Markdown supports; for example, a math equation \(a^2 + b^2 = c^2\).

\hypertarget{usage}{%
\section{Usage}\label{usage}}

Each \textbf{bookdown} chapter is an .Rmd file, and each .Rmd file can contain one (and only one) chapter. A chapter \emph{must} start with a first-level heading: \texttt{\#\ A\ good\ chapter}, and can contain one (and only one) first-level heading.

Use second-level and higher headings within chapters like: \texttt{\#\#\ A\ short\ section} or \texttt{\#\#\#\ An\ even\ shorter\ section}.

The \texttt{index.Rmd} file is required, and is also your first book chapter. It will be the homepage when you render the book.

\hypertarget{render-book}{%
\section{Render book}\label{render-book}}

You can render the HTML version of this example book without changing anything:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Find the \textbf{Build} pane in the RStudio IDE, and
\item
  Click on \textbf{Build Book}, then select your output format, or select ``All formats'' if you'd like to use multiple formats from the same book source files.
\end{enumerate}

Or build the book from the R console:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{bookdown}\SpecialCharTok{::}\FunctionTok{render\_book}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

To render this example to PDF as a \texttt{bookdown::pdf\_book}, you'll need to install XeLaTeX. You are recommended to install TinyTeX (which includes XeLaTeX): \url{https://yihui.org/tinytex/}.

\hypertarget{preview-book}{%
\section{Preview book}\label{preview-book}}

As you work, you may start a local server to live preview this HTML book. This preview will update as you edit the book when you save individual .Rmd files. You can start the server in a work session by using the RStudio add-in ``Preview book'', or from the R console:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{bookdown}\SpecialCharTok{::}\FunctionTok{serve\_book}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\hypertarget{part-common-processing-steps}{%
\part{Common Processing Steps}\label{part-common-processing-steps}}

\hypertarget{prefiltering}{%
\chapter{Prefiltering}\label{prefiltering}}

\hypertarget{libraries}{%
\section{Libraries}\label{libraries}}

\hypertarget{install-libraries}{%
\subsection{Install Libraries}\label{install-libraries}}

All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine. The code below will install \{BiocManager\} from CRAN, and you can then use this package to install PADOG, tweeDEseqCountDatam, and KEGGREST to your system.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{install.packages}\NormalTok{(}\StringTok{"BiocManager"}\NormalTok{)}
\NormalTok{BiocManager}\SpecialCharTok{::}\FunctionTok{install}\NormalTok{(}\StringTok{"PADOG"}\NormalTok{)}
\NormalTok{BiocManager}\SpecialCharTok{::}\FunctionTok{install}\NormalTok{(}\StringTok{"tweeDEseqCountData"}\NormalTok{)}
\NormalTok{BiocManager}\SpecialCharTok{::}\FunctionTok{install}\NormalTok{(}\StringTok{"KEGGREST"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{load-libraries}{%
\subsection{Load Libraries}\label{load-libraries}}

Note that loading these libraries will mask many functions from base R packages. If you run into unexpected errors on functions you're using, it is recommended to use namespacing to explicitly clarify the package from which you need a given function. (I have suppressed the library() loading messages from this document, however.)

\begin{Shaded}
\begin{Highlighting}[]

\FunctionTok{library}\NormalTok{(PADOG)}
\FunctionTok{library}\NormalTok{(tweeDEseqCountData)}
\FunctionTok{library}\NormalTok{(KEGGREST)}
\end{Highlighting}
\end{Shaded}

Provide a brief note on what these libraries are for?

\hypertarget{padog}{%
\subsection{PADOG}\label{padog}}

The PADOG library does XYZ. You can find more information about it online at LINK.

\hypertarget{tweedeseqcountdata}{%
\subsection{tweeDEseqCountData}\label{tweedeseqcountdata}}

\hypertarget{keggrest}{%
\subsection{KEGGREST}\label{keggrest}}

\hypertarget{load-data}{%
\section{Load Data}\label{load-data}}

This section will change substantially when I reconfigure the project as an R package/book.

For now, place any data you need into the \texttt{./data} directory.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# we load the voom{-}transformed Pickrell data set }
\FunctionTok{load}\NormalTok{(}\StringTok{"data/expression\_data\_voomtransformed\_Entrez.Rdata"}\NormalTok{)}

\CommentTok{\# alternatively: load the gene expression measurements that have been transformed using }
\FunctionTok{load}\NormalTok{(}\StringTok{"data/expression\_data\_vsttransformed\_Entrez.Rdata"}\NormalTok{)}

\CommentTok{\# additionally, we load the pickrell data set so that we can access the sample conditions}
\FunctionTok{data}\NormalTok{(pickrell)}
\end{Highlighting}
\end{Shaded}

The sample conditions (i.e.~phenotype labels) of the pickrell data set can be accessed using

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pickrell.eset}\SpecialCharTok{$}\NormalTok{gender }
\CommentTok{\#\textgreater{}  [1] male   male   female male   female male   female male  }
\CommentTok{\#\textgreater{}  [9] female male   female male   female male   female male  }
\CommentTok{\#\textgreater{} [17] female female male   female male   female female male  }
\CommentTok{\#\textgreater{} [25] female male   female female male   female female male  }
\CommentTok{\#\textgreater{} [33] female male   female female female female male   female}
\CommentTok{\#\textgreater{} [41] male   male   female female male   female female male  }
\CommentTok{\#\textgreater{} [49] female female male   female male   male   female female}
\CommentTok{\#\textgreater{} [57] male   female male   female male   female female male  }
\CommentTok{\#\textgreater{} [65] female female female male   female}
\CommentTok{\#\textgreater{} Levels: female male}
\end{Highlighting}
\end{Shaded}

We proceed with the voom-transformed pickrell data set and the corresponding phenotype labels

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# gene expression measurements (transformed)}
\CommentTok{\# note: you can also proceed with the vst{-}transformed gene expression measurements }
\NormalTok{expression\_data\_transformed }\OtherTok{\textless{}{-}}\NormalTok{ expression\_data\_voomtransformed\_Entrez}
\CommentTok{\# sample conditions}
\NormalTok{sample\_conditions }\OtherTok{\textless{}{-}}\NormalTok{ pickrell.eset}\SpecialCharTok{$}\NormalTok{gender}
\end{Highlighting}
\end{Shaded}

\hypertarget{prepare-sample-conditions}{%
\section{Prepare Sample Conditions}\label{prepare-sample-conditions}}

First, we inspect the form of the initial (raw) sample conditions

\begin{Shaded}
\begin{Highlighting}[]
\DocumentationTok{\#\# look at the class: }
\FunctionTok{class}\NormalTok{(sample\_conditions)}
\CommentTok{\#\textgreater{} [1] "factor"}
\CommentTok{\# {-}\textgreater{} the sample labels are already coded as factor}

\CommentTok{\# the current levels are:}
\FunctionTok{levels}\NormalTok{(sample\_conditions)}
\CommentTok{\#\textgreater{} [1] "female" "male"}
\end{Highlighting}
\end{Shaded}

PADOG requires character vector with class labels of the samples. It can only contain ``c'' for control samples or ``d'' for disease samples

\begin{Shaded}
\begin{Highlighting}[]

\CommentTok{\# prepare sample conditions}
\CommentTok{\# we want to convert }
\CommentTok{\# (i) "female" to "c"}
\CommentTok{\# (ii) "male" to "d"}
\NormalTok{sample\_conditions\_prep }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(sample\_conditions, }
                                \AttributeTok{levels=}\FunctionTok{c}\NormalTok{(}\StringTok{"female"}\NormalTok{,}\StringTok{"male"}\NormalTok{), }
                                \AttributeTok{labels=}\FunctionTok{c}\NormalTok{(}\StringTok{"c"}\NormalTok{,}\StringTok{"d"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{run-padog}{%
\section{Run PADOG}\label{run-padog}}

\emph{It is recommended to set a seed to ensure exact reproducibility of the results if the code is run at multiple time points}

you can specify any integer number as the seed. It is VERY IMPORTANT to choose the seed arbitrarily and WITHOUT INSPECTING the results the seed should NEVER be specified based on which value yields the most preferable results.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# run PADOG: }
\NormalTok{ PADOG\_results }\OtherTok{\textless{}{-}} \FunctionTok{padog}\NormalTok{(}\AttributeTok{esetm =} \FunctionTok{as.matrix}\NormalTok{(expression\_data\_transformed), }
                        \AttributeTok{group =}\NormalTok{ sample\_conditions\_prep, }
                        \AttributeTok{dseed =} \DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

arguments:

\begin{itemize}
\tightlist
\item
  \texttt{esetm}: matrix that contains the expression measurements

  \begin{itemize}
  \tightlist
  \item
    note: since the expression data is initially stored in a data frame, we transform it to a matrix when running PADOG
  \end{itemize}
\item
  \texttt{group}: sample conditions (has values ``c'' and ``d'')
\item
  \texttt{dseed}: seed for random number generation (used in the process of phenotype permutation)
\end{itemize}

additional arguments:

\begin{itemize}
\tightlist
\item
  \texttt{paired}: indicates whether the samples in both groups are paired
\item
  \texttt{block}: if the samples are paired (i.e.~argument paired = TRUE), then the paired samples must have the same block value
\item
  \texttt{gslist}: gives instructions on how to cluster the genes into gene sets

  \begin{itemize}
  \tightlist
  \item
    gslist = ``KEGGRESTpathway'': gene sets correspond to KEGG pathways
  \item
    alternative: provide a user-defined list of gene sets
  \end{itemize}
\item
  \texttt{organism}: organism from which the gene expression measurements are taken

  \begin{itemize}
  \tightlist
  \item
    for human, set organism = ``hsa''
  \item
    the required character value for other organisms can be extracted from the KEGGREST package:
  \end{itemize}
\item
  \texttt{obtain} required organisms from column organism
\item
  \texttt{annotation}: required if gslist is set to ``KEGGRESTpathway'' and the rownames of esetm are probe IDs
\item
  can be set to NULL of gslist is set to ``KEGGRESTpathway'' and the rownames of esetm are in the Entrez gene ID format
\item
  if rownames are other gene IDs, then sett annotation = NULL and make sure that the rownames are elements of gslist (and unique!)
\item
  \texttt{gs.names}: contains names of gene sets -\textgreater{} character vector

  \begin{itemize}
  \tightlist
  \item
    must have the same length as gslist
  \end{itemize}
\item
  \texttt{NI}: number of phenotype permutations employed in the assessment of the significance of a given gene set
\end{itemize}

\hypertarget{adjust-for-multiple-testing}{%
\section{Adjust for Multiple Testing}\label{adjust-for-multiple-testing}}

\hypertarget{interpretation-of-results}{%
\section{Interpretation of Results}\label{interpretation-of-results}}

\hypertarget{gsea-preranking}{%
\chapter{GSEA Preranking}\label{gsea-preranking}}

Skip to Chapter 4 Differentiaon Expression Analysis if conversion of gene IDs is not needed.

\hypertarget{gene-id-duplicate-removals}{%
\chapter{Gene ID Duplicate Removals}\label{gene-id-duplicate-removals}}

\hypertarget{differential-expression-analysis}{%
\chapter{Differential Expression Analysis}\label{differential-expression-analysis}}

\hypertarget{rna-seq-transformation}{%
\chapter{RNA-Seq Transformation}\label{rna-seq-transformation}}

Branching point: continue to the next chapter if conversion of gene IDs is not necessary. However, if conversion was/is necessary, proceed to Chapter ?? Cluster Profiler.

\hypertarget{part-no-gene-id-conversion}{%
\part{No Gene ID Conversion}\label{part-no-gene-id-conversion}}

\hypertarget{go-seq}{%
\chapter{Go SEQ}\label{go-seq}}

\hypertarget{cluster-profiler-gsea-go}{%
\chapter{Cluster Profiler: GSEA GO}\label{cluster-profiler-gsea-go}}

\hypertarget{david}{%
\chapter{DAVID}\label{david}}

\hypertarget{gsea-web}{%
\chapter{GSEA WEB}\label{gsea-web}}

\hypertarget{part-with-gene-id-conversion}{%
\part{With Gene ID Conversion}\label{part-with-gene-id-conversion}}

\hypertarget{cluster-profiler-gsea-kegg}{%
\chapter{Cluster Profiler: GSEA KEGG}\label{cluster-profiler-gsea-kegg}}

\hypertarget{cluster-profiler-ora}{%
\chapter{Cluster Profiler: ORA}\label{cluster-profiler-ora}}

\hypertarget{padog-1}{%
\chapter{PADOG}\label{padog-1}}

  \bibliography{book.bib,packages.bib}

\end{document}
