<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 9 clusterProfiler’s GSEA (with gene set database KEGG) | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will do the following two things: Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 9 clusterProfiler’s GSEA (with gene set database KEGG) | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will do the following two things: Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 9 clusterProfiler’s GSEA (with gene set database KEGG) | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will do the following two things: Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="active" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clusterprofilers-gsea-with-gene-set-database-kegg" class="section level1" number="9">
<h1>
<span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)<a class="anchor" aria-label="anchor" href="#clusterprofilers-gsea-with-gene-set-database-kegg"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will do the following two things:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run clusterProfiler’s GSEA with gene set database GO.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>We will go through all (meaningful) researchers’ degrees of freedom.</li>
</ol></li>
</ul>
<p>Note that we provide a separate file for the gene set databases GO and KEGG since clusterProfiler differs in the accepted gene ID formats for them. While for GO, a variety of formats, including Entrez ID, is accepted, the gene IDs must be converted to NCBI (Entrez) ID when choosing KEGG.</p>
<div id="libraries-8" class="section level2" number="9.1">
<h2>
<span class="header-section-number">9.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-8"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load Libraries</strong></p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span></code></pre></div>
<p>Description of the packages</p>
<ul>
<li><p><strong>clusterProfiler</strong>: We here use clusterProfiler’s implementation of GSEA.</p></li>
<li><p><strong>DESeq2</strong>: In contrast detecting the differential expressed genes, which is done solely based on adjusted <span class="math inline">\(p\)</span>-vaules, we additionally use the estimated log fold changes when generating the gene ranking. In DESeq2, the estimated log fold change values are shrunken using an additional function, hence loading the library.</p></li>
</ul>
</div>
<div id="load-data-7" class="section level2" number="9.2">
<h2>
<span class="header-section-number">9.2</span> Load data<a class="anchor" aria-label="anchor" href="#load-data-7"><i class="fas fa-link"></i></a>
</h2>
<p>Note that, as described above, clusterProfiler accepts Entrez IDs when working with gene set database GO. When generating the ranking of the genes with DESeq2, a shrinkage of the log fold change estimates is performed which requires the output of function <em>DESeq()</em>, here named <strong>dds</strong> (see Chapter ’Differential Expression Analysis).</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_limma_Entrez.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_DESeq2_Entrez.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_edgeR_Entrez.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load dds object </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/dds_Entrez.Rdata"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="run-gsea-1" class="section level2" number="9.3">
<h2>
<span class="header-section-number">9.3</span> Run GSEA<a class="anchor" aria-label="anchor" href="#run-gsea-1"><i class="fas fa-link"></i></a>
</h2>
<div id="step-1-generation-of-required-input-object-1" class="section level3" number="9.3.1">
<h3>
<span class="header-section-number">9.3.1</span> Step 1: Generation of Required Input Object<a class="anchor" aria-label="anchor" href="#step-1-generation-of-required-input-object-1"><i class="fas fa-link"></i></a>
</h3>
<p>In the following, we illustrate the creation of the required input object based on the results table of the three different methods for differential expression analysis, voom/limma, DESeq2, and edgeR, which shall serve as three options. The required input object for clusterProfiler is an “order ranked geneList”, i.e. a vector with a ranking value for each gene, named with the respective gene IDs and ordered in a descending manner.<br></p>
<p>The formula for generating the gene ranking from the results table of differential expression analysis is
<span class="math display">\[(-1) * \log_{10}(p\text{-value}) * \text{sign}(\text{log fold change})\]</span>.<br>
Note that “<span class="math inline">\(p\)</span>-value” denotes the non-adjusted p-value.</p>
<div id="option-1-generate-required-input-using-voomlimma-1" class="section level4" number="9.3.1.1">
<h4>
<span class="header-section-number">9.3.1.1</span> Option 1: Generate required input using voom/limma<a class="anchor" aria-label="anchor" href="#option-1-generate-required-input-using-voomlimma-1"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_limma_Entrez</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_limma_Entrez</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA_Entrez</span> <span class="op">&lt;-</span> <span class="va">DE_results_limma_Entrez</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_limma_Entrez</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_limma_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_limma_Entrez</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_limma_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_limma_Entrez</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect the first entries of the ranking vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_limma_Entrez</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;      6192      9087       317      9404      9697     63894 </span></span>
<span><span class="co">#&gt; 54.991351 38.217609  3.554179  3.200661  3.164203  3.071138 </span></span>
<span><span class="co">#&gt;      4214     10479      5257     29915 </span></span>
<span><span class="co">#&gt;  3.019720  2.924894  2.875024  2.814519</span></span></code></pre></div>
</div>
<div id="option-2-generate-required-input-using-deseq2-1" class="section level4" number="9.3.1.2">
<h4>
<span class="header-section-number">9.3.1.2</span> Option 2: Generate required input using DESeq2<a class="anchor" aria-label="anchor" href="#option-2-generate-required-input-using-deseq2-1"><i class="fas fa-link"></i></a>
</h4>
<p><strong>(i) Perform Differential Expression Analysis</strong></p>
<p>This step is performed in the script Instructions_Differential_Expression_Analysis.R. Note that we must exit the DESeq2 workflow from the script a few steps early since there is an additional step required in for the creation of the required input object. We proceed with the object “dds_Entrez” we’ve obtained in step 3 of DESeq2’s workflow for differential expression analysis.</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds_Entrez</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 9492 69 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(6): counts mu ... replaceCounts replaceCooks</span></span>
<span><span class="co">#&gt; rownames(9492): 8813 57147 ... 56104 56112</span></span>
<span><span class="co">#&gt; rowData names(23): baseMean baseVar ... maxCooks</span></span>
<span><span class="co">#&gt;   replace</span></span>
<span><span class="co">#&gt; colnames(69): NA18486 NA18498 ... NA19239 NA19257</span></span>
<span><span class="co">#&gt; colData names(3): condition sizeFactor replaceable</span></span></code></pre></div>
<p><strong>(ii) Shrink estimated log fold change values</strong></p>
<p>The intuition behind shrinkage of log fold change values is that, as mentioned in the paper, RNA-Seq data consists (in its raw form) of count data in which is inherently heteroscedastic, i.e. the variance of the count data depends on the mean count of the count data. It is observable that ratios between counts are considerably noisier in low magnitudes of counts compared to higher magnitudes, i.e. the log fold changes between both conditions are higher if the overall magnitude of counts is lower.<br>
DESeq2 addresses this issue by shrinking the estimated log fold changes in the direction of 0 the magnitude of shrinkage is higher if the available information for a gene is lower (which may be because of a low magnitude of counts, a high dispersion or few degrees of freedom.). A more detailed description is provided in the DESeq2 paper by Love et al. (2014).<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;Love, Huber, and Anders, &lt;span&gt;“&lt;span class="nocase"&gt;Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2&lt;/span&gt;.”&lt;/span&gt;&lt;/p&gt;'><sup>19</sup></a></span></p>
<p>Perform shrinkage:</p>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DE_results_DESeq2_shrink_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span><span class="op">(</span><span class="va">dds_Entrez</span>,</span>
<span>                                              coef <span class="op">=</span> <span class="st">"condition_treated_vs_untreated"</span>,</span>
<span>                                              type<span class="op">=</span><span class="st">"apeglm"</span><span class="op">)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li><p><strong>type</strong>: method to perform shrinkage. We opt for the default “apeglm” but you can choose from two alternative options as well</p></li>
<li><p><strong>coef</strong>: indicate the coefficients to be shrunk. We can obtain the right argument from the following function call:</p></li>
</ul>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">resultsNames</a></span><span class="op">(</span><span class="va">dds_Entrez</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Intercept"                     </span></span>
<span><span class="co">#&gt; [2] "condition_treated_vs_untreated"</span></span></code></pre></div>
<p>This command shows us that we can either shrink the intercept or the “condition_treated_vs_untreated”. Since we do not want to shrink the intercept but the estimated log fold changes, we opt for the second option “condition_treated_vs_untreated”.</p>
<p><strong>(iii) Generate the ranked list of the genes</strong></p>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the DESeq2 results to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes with an existing p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_Entrez</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_shrink_Entrez</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset the DESeq2 results to those genes with an existing p-value</span></span>
<span><span class="va">DE_results_noNA_Entrez</span> <span class="op">&lt;-</span> <span class="va">DE_results_DESeq2_shrink_Entrez</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_Entrez</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_DESeq2_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Entrez</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_DESeq2_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Entrez</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># inspect the first entries of the ranking vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Entrez</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;       6192       9086       9087       7404         58 </span></span>
<span><span class="co">#&gt; 204.374390  75.096224  64.815184   7.205688   4.854231 </span></span>
<span><span class="co">#&gt;       1521      23136     219699     126129        317 </span></span>
<span><span class="co">#&gt;   4.467907   3.996186   3.966326   3.918029   3.744345</span></span></code></pre></div>
</div>
<div id="option-3-generate-required-input-using-edger-1" class="section level4" number="9.3.1.3">
<h4>
<span class="header-section-number">9.3.1.3</span> Option 3: Generate required input using edgeR<a class="anchor" aria-label="anchor" href="#option-3-generate-required-input-using-edger-1"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_Entrez</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Entrez</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA_Entrez</span> <span class="op">&lt;-</span> <span class="va">DE_results_edgeR_Entrez</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_Entrez</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_edgeR_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">$</span><span class="va">p_adj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Entrez</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Entrez</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_edgeR_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Entrez</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect ranking vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Entrez</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;        6192        9087        1521         317        9404 </span></span>
<span><span class="co">#&gt;         Inf 116.7303466   1.6605772   1.0317662   0.9984938 </span></span>
<span><span class="co">#&gt;       10479        4214        9697        4509        9645 </span></span>
<span><span class="co">#&gt;   0.9863396   0.8622382   0.8467278   0.8434267   0.8251273</span></span>
<span></span>
<span><span class="co"># 5. special problem here: gene 6192 has a ranking value Inf since its adjusted</span></span>
<span><span class="co"># p-value in the results table of differential expression analysis amounts to 0</span></span>
<span></span>
<span><span class="co"># here, we deal with this issue by resetting this ranking value to the highest ranking value</span></span>
<span><span class="co"># that occurs among the remaining genes</span></span>
<span><span class="co"># -&gt; note that there is NO common way of dealing with this issue</span></span>
<span><span class="va">rankvec_edgeR_Entrez</span><span class="op">[</span><span class="va">rankvec_edgeR_Entrez</span> <span class="op">==</span> <span class="cn">Inf</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Entrez</span><span class="op">[</span><span class="va">rankvec_edgeR_Entrez</span> <span class="op">!=</span> <span class="cn">Inf</span><span class="op">]</span><span class="op">)</span></span>
<span> </span>
<span><span class="co"># inspect ranking vector once more </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Entrez</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;        6192        9087        1521         317        9404 </span></span>
<span><span class="co">#&gt; 116.7303466 116.7303466   1.6605772   1.0317662   0.9984938 </span></span>
<span><span class="co">#&gt;       10479        4214        9697        4509        9645 </span></span>
<span><span class="co">#&gt;   0.9863396   0.8622382   0.8467278   0.8434267   0.8251273</span></span></code></pre></div>
</div>
</div>
<div id="step-2-run-gsea-with-gene-set-database-kegg" class="section level3" number="9.3.2">
<h3>
<span class="header-section-number">9.3.2</span> Step 2: Run GSEA with gene set database KEGG<a class="anchor" aria-label="anchor" href="#step-2-run-gsea-with-gene-set-database-kegg"><i class="fas fa-link"></i></a>
</h3>
<p>For the purpose of simplicity, we work with the ranking generated using limma. However, you can switch around at your discretion.</p>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rankvec_Entrez</span> <span class="op">&lt;-</span> <span class="va">rankvec_limma_Entrez</span></span>
<span></span>
<span><span class="co"># alternatively: </span></span>
<span><span class="co"># rankvec_Entrez &lt;- rankvec_DESeq2_Entrez</span></span>
<span><span class="co"># rankvec_Entrez &lt;- rankvec_edgeR_Entrez</span></span></code></pre></div>
<p>Run clusterProfiler with gene set database KEGG</p>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># important: set seed for reproducibility </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">GSEA_KEGG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseKEGG.html">gseKEGG</a></span><span class="op">(</span>geneList <span class="op">=</span> <span class="va">rankvec_Entrez</span>, </span>
<span>                     seed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  </span></code></pre></div>
<p>additional arguments (these do not appear in the above code lines since we keep the default options)</p>
<ul>
<li><p><strong>‘exponent = 1’</strong>: in the calculation of the enrichment score, each gene is weighted by its absolute value of the ranking metric</p></li>
<li><p><strong>‘organism = “hsa”’</strong>: human organism. Note that the list of accepted organisms is available in the following link:<br><a href="https://www.genome.jp/kegg/catalog/org_list.html" class="uri">https://www.genome.jp/kegg/catalog/org_list.html</a></p></li>
<li><p><strong>‘seed = TRUE’</strong>: set the seed you have indicated above (here: seed = 1 )</p></li>
<li><p><strong>‘eps = 1e-10’</strong>: each (adjusted) p-value that is smaller than 1e-10 is indicated as 1e-10</p></li>
<li><p><strong>‘pvalueCutoff’</strong> = 0.05: only those gene sets with a p-value &lt; 0.05 are indicated in the results</p></li>
<li><p><strong>note</strong>: if you want to inspect ALL gene sets in the results, set pvalueCutoff = 1</p></li>
<li><p><strong>pAdjustMethod</strong> = “BH”: adjustment for multiple testing using Benjamini and Hochberg</p></li>
</ul>
</div>
<div id="step-3-interpretation-of-the-results-1" class="section level3" number="9.3.3">
<h3>
<span class="header-section-number">9.3.3</span> step 3: Interpretation of the results<a class="anchor" aria-label="anchor" href="#step-3-interpretation-of-the-results-1"><i class="fas fa-link"></i></a>
</h3>
<p>We inspect the results for the first 10 gene sets</p>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#head(GSEA_KEGG, n = 10 )</span></span></code></pre></div>
<p>Columns in Result Tables:</p>
<ul>
<li><p><strong>ID</strong>: ID of the gene set</p></li>
<li><p><strong>Description</strong>: Description of Gene Set</p></li>
<li><p><strong>setSize</strong>: Size of the gene set</p></li>
<li>
<p><strong>enrichmentScore</strong>: Enrichment Score</p>
<ul>
<li>note: this enrichment score has not been normalized for gene set size</li>
<li>means: larger gene sets automatically have a bigger (absolute) enrichment score independent of actual differential enrichment</li>
<li>the raw enrichment score is therefore not comparable between different gene sets</li>
</ul>
</li>
<li>
<p><strong>NES</strong>: Normalized version of column enrichmentScore</p>
<ul>
<li>NES can be compared between different gene sets</li>
</ul>
</li>
<li>
<p><strong>pvalue</strong>: p-value of enrichment of given gene set</p>
<ul>
<li>note: this raw p-value has <strong>not</strong> been adjusted for multiple testing</li>
<li>therefore: <strong>cannot</strong> be used to assess differential enrichment of a given gene set</li>
</ul>
</li>
<li>
<p><strong>p.adjust</strong>: <strong>adjusted</strong> p-value of a given gene set</p>
<ul>
<li>this p-value now has been adjusted for multiple testing and can therefore</li>
<li>can therefore be used to assess differential enrichment</li>
<li>example: detect all genes with p.adjust &lt; 0.05 as differentially enriched</li>
</ul>
</li>
<li>
<p><strong>qvalue</strong>: ADJUSTED p-value of a given gene set</p>
<ul>
<li>note: a qvalue is the analog to p.adjust, but adjusted for multiple testing
using a different method</li>
</ul>
</li>
</ul>
<p>You can either use the column p.adjust or qvalue to assess whether a gene set
is differentially enriched or not</p>
<ul>
<li><p><strong>rank</strong>: position in the ranked list of the genes in which the maximum difference between the two sums occurs, i.e. the rank at which the enrichment score is extracted</p></li>
<li>
<p><strong>leading_edge</strong>:</p>
<ul>
<li>
<strong>tags</strong>: The percentage of genes in the ranked list that are members of gene set before (for positive enrichment score) or after (for negative enrichment score) the position from which the enrichment score is extracted.</li>
<li>
<strong>list</strong>: The percentage of genes in the ranked gene list before (for positive enrichment
score) or after (for negative enrichment score) the position from which the enrichment score is extracted.</li>
<li>
<strong>signal</strong>: enrichment signal strength combines the statistics “tags” and “list”</li>
</ul>
</li>
<li><p><strong>core_enrichment</strong>: genes that contribute most to the enrichment results</p></li>
<li>
<p>Note that a more detailed information on leading_edge and core_enrichment can be found in the user
manual provided for GSEA:
<a href="https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html" class="uri">https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html</a></p>
<ul>
<li>search for <strong>Detailed Enrichment Results</strong> (section “Leading Edge”)</li>
<li>search for <strong>CORE ENRICHMENT</strong>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></div>
<div class="next"><a href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clusterprofilers-gsea-with-gene-set-database-kegg"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="nav-link" href="#libraries-8"><span class="header-section-number">9.1</span> Libraries</a></li>
<li><a class="nav-link" href="#load-data-7"><span class="header-section-number">9.2</span> Load data</a></li>
<li>
<a class="nav-link" href="#run-gsea-1"><span class="header-section-number">9.3</span> Run GSEA</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#step-1-generation-of-required-input-object-1"><span class="header-section-number">9.3.1</span> Step 1: Generation of Required Input Object</a></li>
<li><a class="nav-link" href="#step-2-run-gsea-with-gene-set-database-kegg"><span class="header-section-number">9.3.2</span> Step 2: Run GSEA with gene set database KEGG</a></li>
<li><a class="nav-link" href="#step-3-interpretation-of-the-results-1"><span class="header-section-number">9.3.3</span> step 3: Interpretation of the results</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
