<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 6 clusterProfiler’s GSEA (with gene set database GO) | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will do the following two things: Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 6 clusterProfiler’s GSEA (with gene set database GO) | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will do the following two things: Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 6 clusterProfiler’s GSEA (with gene set database GO) | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will do the following two things: Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through all steps required to run...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="active" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clusterprofilers-gsea-with-gene-set-database-go" class="section level1" number="6">
<h1>
<span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)<a class="anchor" aria-label="anchor" href="#clusterprofilers-gsea-with-gene-set-database-go"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will do the following two things:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Based on the results of differential expression analysis from voom/limma, DESeq2, and edgeR, we will go through
all steps required to run clusterProfiler’s GSEA with gene set database GO.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>We will go through all (meaningful) researchers’ degrees of freedom.</li>
</ol></li>
</ul>
<p>Note that we provide a separate file for the gene set databases GO and KEGG since clusterProfiler differs in the accepted gene ID formats for them. While for GO, a variety of formats, including Ensembl ID, is accepted, the gene IDs must be converted to NCBI (Entrez) ID when choosing KEGG.</p>
<div id="libraries-5" class="section level3" number="6.0.1">
<h3>
<span class="header-section-number">6.0.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-5"><i class="fas fa-link"></i></a>
</h3>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"apeglm"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load Libraries </strong></p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span></code></pre></div>
<p>Description of the packages</p>
<ul>
<li><p><strong>clusterProfiler</strong>: We here use clusterProfiler’s implementation of GSEA.</p></li>
<li><p><strong>org.Hs.eg.db</strong>: Provides genome-wide annotation for humans. When working with a different organism, the user has to provide a different package (see chapter ‘About’). Note that this library is required when running clusterProfiler’s GSEA with gene set database GO.</p></li>
<li><p><strong>DESeq2</strong>: In contrast detecting the differential expressed genes, which is done solely based on adjusted <span class="math inline">\(p\)</span>-vaules, we additionally use the estimated log fold changes when generating the gene ranking. In DESeq2, the estimated log fold change values are shrunken using an additional function, hence loading the library.</p></li>
</ul>
</div>
<div id="load-data-4" class="section level2" number="6.1">
<h2>
<span class="header-section-number">6.1</span> Load data<a class="anchor" aria-label="anchor" href="#load-data-4"><i class="fas fa-link"></i></a>
</h2>
<p>Note that, as described above, clusterProfiler accepts Ensembl IDs when working with gene set database GO. When generating the ranking of genes with DESeq2, a shrinkage of the log fold change estimates is performed which requires the output of function <em>DESeq()</em>, here named <strong>‘dds’</strong> (see Chapter ‘Differential expression analysis’).</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_limma_Ensembl.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_DESeq2_Ensembl.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_edgeR_Ensembl.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load dds object </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/dds_Ensembl.Rdata"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="run-gsea" class="section level2" number="6.2">
<h2>
<span class="header-section-number">6.2</span> Run GSEA<a class="anchor" aria-label="anchor" href="#run-gsea"><i class="fas fa-link"></i></a>
</h2>
<div id="step-1-generation-of-required-input-object" class="section level3" number="6.2.1">
<h3>
<span class="header-section-number">6.2.1</span> Step 1: Generation of required input object<a class="anchor" aria-label="anchor" href="#step-1-generation-of-required-input-object"><i class="fas fa-link"></i></a>
</h3>
<p>In the following, we illustrate the creation of the required input object using the three different methods for differential expression analysis, voom/limma, DESeq2, and edgeR, which shall serve as three options. The required input object for clusterProfiler is an “order ranked geneList”, i.e. a vector with a ranking value for each gene, named with the respective gene IDs and ordered in a descending manner.<br>
Here we choose the following gene-level statistic based on which we can transform the results of differential expression analysis to the gene ranking:
<span class="math display">\[(-1) * \log_{10}(p\text{-value}) * \text{sign}(\text{log fold change})\]</span>.<br>
Note that “<span class="math inline">\(p\)</span>-value” denotes the non-adjusted p-value.</p>
<div id="option-1-generate-required-input-using-voomlimma" class="section level4" number="6.2.1.1">
<h4>
<span class="header-section-number">6.2.1.1</span> Option 1: Generate required input using voom/limma<a class="anchor" aria-label="anchor" href="#option-1-generate-required-input-using-voomlimma"><i class="fas fa-link"></i></a>
</h4>
<p>Before starting, we want to have a short look at the results table of differential expression analysis generated with limma so that we can identify the relevant columns.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_limma_Ensembl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;                      logFC   AveExpr         t      P.Value</span></span>
<span><span class="co">#&gt; ENSG00000129824  9.2133770 1.9601987 47.132083 1.061719e-55</span></span>
<span><span class="co">#&gt; ENSG00000099749  6.1709577 0.4153409 40.465602 4.024370e-51</span></span>
<span><span class="co">#&gt; ENSG00000154620  5.0848991 0.4867874 26.729729 5.097737e-39</span></span>
<span><span class="co">#&gt; ENSG00000006757 -0.9214380 5.3073131 -8.805751 4.963544e-13</span></span>
<span><span class="co">#&gt; ENSG00000130021 -0.8516173 2.1365915 -4.388951 3.846411e-05</span></span>
<span><span class="co">#&gt; ENSG00000185753 -0.5479423 3.6902013 -4.217130 7.121081e-05</span></span>
<span><span class="co">#&gt; ENSG00000086712 -0.4017993 5.0899555 -4.023482 1.403927e-04</span></span>
<span><span class="co">#&gt; ENSG00000123689 -1.4679601 5.1661467 -3.850433 2.537689e-04</span></span>
<span><span class="co">#&gt; ENSG00000177606 -0.5844304 8.2469942 -3.910485 2.069770e-04</span></span>
<span><span class="co">#&gt; ENSG00000120868  0.4076940 5.9420527  3.832337 2.697515e-04</span></span>
<span><span class="co">#&gt;                        p_adj           B</span></span>
<span><span class="co">#&gt; ENSG00000129824 6.631497e-52 94.00193786</span></span>
<span><span class="co">#&gt; ENSG00000099749 1.256811e-47 86.16860821</span></span>
<span><span class="co">#&gt; ENSG00000154620 1.061349e-35 68.53453635</span></span>
<span><span class="co">#&gt; ENSG00000006757 7.750575e-10 19.36978341</span></span>
<span><span class="co">#&gt; ENSG00000130021 4.804937e-02  2.04582318</span></span>
<span><span class="co">#&gt; ENSG00000185753 7.413045e-02  1.30583651</span></span>
<span><span class="co">#&gt; ENSG00000086712 1.252704e-01  0.48313770</span></span>
<span><span class="co">#&gt; ENSG00000123689 1.684868e-01 -0.04984613</span></span>
<span><span class="co">#&gt; ENSG00000177606 1.615973e-01 -0.13323415</span></span>
<span><span class="co">#&gt; ENSG00000120868 1.684868e-01 -0.25078873</span></span></code></pre></div>
<p>Generate gene ranking:</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e. which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes with an existing p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_limma_Ensembl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_limma_Ensembl</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with an existing p-value</span></span>
<span><span class="va">DE_results_noNA_Ensembl</span> <span class="op">&lt;-</span> <span class="va">DE_results_limma_Ensembl</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_limma_Ensembl</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_limma_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_limma_Ensembl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_limma_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_limma_Ensembl</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Inspect the first entries of the ranking vector</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_limma_Ensembl</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000154620 </span></span>
<span><span class="co">#&gt;       54.973990       50.395302       38.292623 </span></span>
<span><span class="co">#&gt; ENSG00000120868 ENSG00000110031 ENSG00000151445 </span></span>
<span><span class="co">#&gt;        3.569036        3.262283        3.222725 </span></span>
<span><span class="co">#&gt; ENSG00000065308 ENSG00000095015 ENSG00000198689 </span></span>
<span><span class="co">#&gt;        3.198940        3.033258        2.949591 </span></span>
<span><span class="co">#&gt; ENSG00000102893 </span></span>
<span><span class="co">#&gt;        2.915174</span></span></code></pre></div>
</div>
<div id="option-2-generate-required-input-using-deseq2" class="section level4" number="6.2.1.2">
<h4>
<span class="header-section-number">6.2.1.2</span> Option 2: Generate required input using DESeq2<a class="anchor" aria-label="anchor" href="#option-2-generate-required-input-using-deseq2"><i class="fas fa-link"></i></a>
</h4>
<p><strong>Step 1: Perform differential expression analysis</strong></p>
<p>This step is performed in Chapter ‘Differential expression analysis’. Note that we must exit the DESeq2 workflow from the script a few steps early since there is an additional step required in for the creation of the required input object. We proceed with the object “dds_Ensembl” we’ve obtained in step 2 of DESeq2’s workflow for differential expression analysis.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds_Ensembl</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 10151 69 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(6): counts mu ... replaceCounts replaceCooks</span></span>
<span><span class="co">#&gt; rownames(10151): ENSG00000000419 ENSG00000000457 ...</span></span>
<span><span class="co">#&gt;   ENSG00000254245 ENSG00000254290</span></span>
<span><span class="co">#&gt; rowData names(23): baseMean baseVar ... maxCooks</span></span>
<span><span class="co">#&gt;   replace</span></span>
<span><span class="co">#&gt; colnames(69): NA18486 NA18498 ... NA19239 NA19257</span></span>
<span><span class="co">#&gt; colData names(3): condition sizeFactor replaceable</span></span></code></pre></div>
<p><strong>Step 2: Shrink estimated log fold change values</strong></p>
<p>The intuition behind shrinkage of log fold change values is that, as mentioned in the paper, RNA-Seq data consists (in its raw form) of count data which is inherently heteroscedastic, i.e. the variance of the count data depends on the mean count of the count data. It is observable that ratios between counts are considerably noisier in low magnitudes of counts compared to higher magnitudes, i.e. the log fold changes between both conditions are higher if the overall magnitude of counts is lower.<br>
DESeq2 addresses this issue by shrinking the estimated log fold changes in the direction of 0 the magnitude of shrinkage is higher if the available information for a gene is lower (which may be because of a low magnitude of counts, a high dispersion or few degrees of freedom.). A more detailed description is provided in the DESeq2 paper by Love et al. (2014).<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Love, Huber, and Anders.&lt;/p&gt;"><sup>17</sup></a></span></p>
<p>Perform shrinkage:</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DE_results_DESeq2_shrink_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span><span class="op">(</span><span class="va">dds_Ensembl</span>,</span>
<span>                                               coef <span class="op">=</span> <span class="st">"condition_treated_vs_untreated"</span>,</span>
<span>                                               type<span class="op">=</span><span class="st">"apeglm"</span><span class="op">)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li><p><strong>type</strong>: method to perform shrinkage. We opt for the default ‘type = apeglm’, however, there are two alternative options the user can choose from.</p></li>
<li><p><strong>coef</strong>: indicate the coefficients to be shrunk. We can obtain the right argument from the following function call:</p></li>
</ul>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">resultsNames</a></span><span class="op">(</span><span class="va">dds_Ensembl</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Intercept"                     </span></span>
<span><span class="co">#&gt; [2] "condition_treated_vs_untreated"</span></span></code></pre></div>
<p>This command shows us that we can either shrink the intercept or the “condition_treated_vs_untreated”. Since we do not want to shrink the intercept but the estimated log fold changes, we opt for the second option “condition_treated_vs_untreated”.</p>
<p>Before generating the ranking, we want to inspect the results table generated with DESeq2 (incl. shrinkage) so that we can identify the relevant columns</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_shrink_Ensembl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; log2 fold change (MAP): condition treated vs untreated </span></span>
<span><span class="co">#&gt; Wald test p-value: condition treated vs untreated </span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 5 columns</span></span>
<span><span class="co">#&gt;                  baseMean log2FoldChange      lfcSE</span></span>
<span><span class="co">#&gt;                 &lt;numeric&gt;      &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSG00000000419  62.71639   -1.75964e-05 0.00144263</span></span>
<span><span class="co">#&gt; ENSG00000000457  70.46817    2.47225e-06 0.00144258</span></span>
<span><span class="co">#&gt; ENSG00000000460  13.00098    7.54395e-06 0.00144264</span></span>
<span><span class="co">#&gt; ENSG00000000938  43.56413    7.35510e-06 0.00144267</span></span>
<span><span class="co">#&gt; ENSG00000001036  37.95191    1.68099e-06 0.00144257</span></span>
<span><span class="co">#&gt; ENSG00000001167 228.48025    6.71863e-07 0.00144253</span></span>
<span><span class="co">#&gt; ENSG00000001497  11.52425    7.90345e-07 0.00144265</span></span>
<span><span class="co">#&gt; ENSG00000001561 175.93013    1.18172e-05 0.00144266</span></span>
<span><span class="co">#&gt; ENSG00000002016   4.68668    7.85799e-06 0.00144267</span></span>
<span><span class="co">#&gt; ENSG00000002330  66.34023   -2.62299e-05 0.00144267</span></span>
<span><span class="co">#&gt;                     pvalue      padj</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSG00000000419 0.72425162  0.971027</span></span>
<span><span class="co">#&gt; ENSG00000000457 0.79118089  0.981324</span></span>
<span><span class="co">#&gt; ENSG00000000460 0.27497228  0.820697</span></span>
<span><span class="co">#&gt; ENSG00000000938 0.55928481  0.945493</span></span>
<span><span class="co">#&gt; ENSG00000001036 0.95622491  0.997305</span></span>
<span><span class="co">#&gt; ENSG00000001167 0.59046443  0.953105</span></span>
<span><span class="co">#&gt; ENSG00000001497 0.88864354  0.992301</span></span>
<span><span class="co">#&gt; ENSG00000001561 0.06610429  0.575449</span></span>
<span><span class="co">#&gt; ENSG00000002016 0.14450244  0.703418</span></span>
<span><span class="co">#&gt; ENSG00000002330 0.00949353  0.350398</span></span></code></pre></div>
<p>Now we proceed to generating the ranking.</p>
<p><strong>Step 3: Generate the gene ranking based on the results of differential expression analysis</strong></p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_Ensembl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_shrink_Ensembl</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA_Ensembl</span> <span class="op">&lt;-</span> <span class="va">DE_results_DESeq2_shrink_Ensembl</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_Ensembl</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_DESeq2_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Ensembl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_DESeq2_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Ensembl</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Inspect the first entries of the ranking vector</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_DESeq2_Ensembl</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000198692 </span></span>
<span><span class="co">#&gt;      204.454993      105.507250       75.109639 </span></span>
<span><span class="co">#&gt; ENSG00000154620 ENSG00000157828 ENSG00000183878 </span></span>
<span><span class="co">#&gt;       64.872131       39.560339        7.260319 </span></span>
<span><span class="co">#&gt; ENSG00000143632 ENSG00000172543 ENSG00000082397 </span></span>
<span><span class="co">#&gt;        4.867166        4.469387        3.999177 </span></span>
<span><span class="co">#&gt; ENSG00000107731 </span></span>
<span><span class="co">#&gt;        3.973429</span></span></code></pre></div>
</div>
<div id="option-3-generate-required-input-using-edger" class="section level4" number="6.2.1.3">
<h4>
<span class="header-section-number">6.2.1.3</span> Option 3: Generate required input using edgeR<a class="anchor" aria-label="anchor" href="#option-3-generate-required-input-using-edger"><i class="fas fa-link"></i></a>
</h4>
<p>Before generating the ranking, we want inspect the results table of differential expression analysis generated with edgeR to identify the relevant columns.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Ensembl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;                      logFC   logCPM        PValue</span></span>
<span><span class="co">#&gt; ENSG00000129824 10.1353289 6.130860  0.000000e+00</span></span>
<span><span class="co">#&gt; ENSG00000099749  7.8142855 2.949875 3.594389e-197</span></span>
<span><span class="co">#&gt; ENSG00000154620  5.5800954 2.493788 1.225959e-122</span></span>
<span><span class="co">#&gt; ENSG00000006757 -0.9412925 5.457916  1.516865e-17</span></span>
<span><span class="co">#&gt; ENSG00000130021 -0.7752345 2.545089  3.626637e-06</span></span>
<span><span class="co">#&gt; ENSG00000124216 -1.6423626 3.261210  4.708108e-06</span></span>
<span><span class="co">#&gt; ENSG00000130222 -1.5833419 2.652436  5.638958e-06</span></span>
<span><span class="co">#&gt; ENSG00000196407 -1.1211411 2.276703  1.959274e-05</span></span>
<span><span class="co">#&gt; ENSG00000172543  1.5008703 3.336720  2.836816e-05</span></span>
<span><span class="co">#&gt; ENSG00000086712 -0.4218274 5.181013  3.922485e-05</span></span>
<span><span class="co">#&gt;                         p_adj</span></span>
<span><span class="co">#&gt; ENSG00000129824  0.000000e+00</span></span>
<span><span class="co">#&gt; ENSG00000099749 1.122528e-193</span></span>
<span><span class="co">#&gt; ENSG00000154620 2.552447e-119</span></span>
<span><span class="co">#&gt; ENSG00000006757  2.368584e-14</span></span>
<span><span class="co">#&gt; ENSG00000130021  4.530395e-03</span></span>
<span><span class="co">#&gt; ENSG00000124216  4.901141e-03</span></span>
<span><span class="co">#&gt; ENSG00000130222  5.031562e-03</span></span>
<span><span class="co">#&gt; ENSG00000196407  1.529703e-02</span></span>
<span><span class="co">#&gt; ENSG00000172543  1.968750e-02</span></span>
<span><span class="co">#&gt; ENSG00000086712  2.449984e-02</span></span></code></pre></div>
<p>Now we proceed to generating the ranking.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_Ensembl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Ensembl</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA_Ensembl</span> <span class="op">&lt;-</span> <span class="va">DE_results_edgeR_Ensembl</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_Ensembl</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">$</span><span class="va">p_adj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA_Ensembl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect ranking vector: </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000154620 </span></span>
<span><span class="co">#&gt;             Inf     192.9498029     118.5930433 </span></span>
<span><span class="co">#&gt; ENSG00000172543 ENSG00000120868 ENSG00000110031 </span></span>
<span><span class="co">#&gt;       1.7058094       1.0563194       1.0370201 </span></span>
<span><span class="co">#&gt; ENSG00000198689 ENSG00000065308 ENSG00000095015 </span></span>
<span><span class="co">#&gt;       0.9865108       0.8790009       0.8790009 </span></span>
<span><span class="co">#&gt; ENSG00000228253 </span></span>
<span><span class="co">#&gt;       0.8654283</span></span>
<span></span>
<span><span class="co"># 5. special problem here: gene ENSG00000129824 has a ranking value Inf since its adjusted</span></span>
<span><span class="co"># p-value in the results table of differential expression analysis amounts to 0</span></span>
<span></span>
<span><span class="co"># here, we deal with this issue by resetting this ranking value to the highest ranking value</span></span>
<span><span class="co"># that occurs among the remaining genes</span></span>
<span><span class="co"># -&gt; note that there is NO common way of dealing with this issue</span></span>
<span><span class="va">rankvec_edgeR_Ensembl</span><span class="op">[</span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">==</span> <span class="cn">Inf</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span><span class="op">[</span><span class="va">rankvec_edgeR_Ensembl</span> <span class="op">!=</span> <span class="cn">Inf</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Inspect the first entries of the ranking vector:</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR_Ensembl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ENSG00000129824 ENSG00000099749 ENSG00000154620 </span></span>
<span><span class="co">#&gt;     192.9498029     192.9498029     118.5930433 </span></span>
<span><span class="co">#&gt; ENSG00000172543 ENSG00000120868 ENSG00000110031 </span></span>
<span><span class="co">#&gt;       1.7058094       1.0563194       1.0370201 </span></span>
<span><span class="co">#&gt; ENSG00000198689 ENSG00000065308 ENSG00000095015 </span></span>
<span><span class="co">#&gt;       0.9865108       0.8790009       0.8790009 </span></span>
<span><span class="co">#&gt; ENSG00000228253 </span></span>
<span><span class="co">#&gt;       0.8654283</span></span></code></pre></div>
</div>
</div>
<div id="step-2-run-gsea-with-gene-set-database-go" class="section level3" number="6.2.2">
<h3>
<span class="header-section-number">6.2.2</span> Step 2: Run GSEA with gene set database GO<a class="anchor" aria-label="anchor" href="#step-2-run-gsea-with-gene-set-database-go"><i class="fas fa-link"></i></a>
</h3>
<p>For the purpose of simplicity, we work with the ranking generated using limma. However, you can switch around at your discretion.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rankvec_Ensembl</span> <span class="op">&lt;-</span> <span class="va">rankvec_limma_Ensembl</span></span>
<span></span>
<span><span class="co"># alternatively: </span></span>
<span><span class="co">#rankvec_Ensembl &lt;- rankvec_DESeq2_Ensembl</span></span>
<span><span class="co">#rankvec_Ensembl &lt;- rankvec_edgeR_Ensembl</span></span></code></pre></div>
<p>Run GSEA with gene set database GO. Note that it is important to set a seed for reproducibility.</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># important: set seed for reprodubibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSEA_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html">gseGO</a></span><span class="op">(</span>geneList <span class="op">=</span> <span class="va">rankvec_Ensembl</span>,</span>
<span>                   ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                   OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                   keyType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>                   seed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># set seed for reproducibility</span></span></code></pre></div>
<p>Arguments of function <em>gseGO()</em>:</p>
<ul>
<li><p><strong>geneList</strong>: vector of gene ranking (generated in manner of step 1)</p></li>
<li><p><strong>ont</strong>: subontology of GO (must be one of “BP”, “CC”, “MF”)</p></li>
<li><p><strong>OrgDb</strong>: indicates organism from which the gene expression measurements are taken</p></li>
<li>
<p><strong>keyTypes</strong>: for our purposes here, argument keyType indicates the gene ID format of the gene names in vector rankvec</p>
<ul>
<li>available keytypes can be found in the following:</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">keytypes</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"     </span></span>
<span><span class="co">#&gt;  [4] "ENSEMBLPROT"  "ENSEMBLTRANS" "ENTREZID"    </span></span>
<span><span class="co">#&gt;  [7] "ENZYME"       "EVIDENCE"     "EVIDENCEALL" </span></span>
<span><span class="co">#&gt; [10] "GENENAME"     "GENETYPE"     "GO"          </span></span>
<span><span class="co">#&gt; [13] "GOALL"        "IPI"          "MAP"         </span></span>
<span><span class="co">#&gt; [16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL" </span></span>
<span><span class="co">#&gt; [19] "PATH"         "PFAM"         "PMID"        </span></span>
<span><span class="co">#&gt; [22] "PROSITE"      "REFSEQ"       "SYMBOL"      </span></span>
<span><span class="co">#&gt; [25] "UCSCKG"       "UNIPROT"</span></span></code></pre></div>
<ul>
<li><p><strong>seed = TRUE</strong>: set the seed you have indicated above (here: seed = 1 )</p></li>
<li><p>additional arguments (these do not appear in the above code lines since we keep the default options)</p></li>
<li><p><strong>exponent = 1</strong>: in the calculation of the enrichment score, each gene is weighted by its absolute value of the ranking metric</p></li>
<li><p><strong>eps = 1e-10</strong>: each p-value that is smaller than 1e-10 is indicated as 1e-10</p></li>
<li><p><strong>pvalueCutoff = 0.05</strong>: only those gene sets with a p-value &lt; 0.05 are indicated in the results. Note that if you want to inspect ALL gene sets in the results, set pvalueCutoff = 1</p></li>
<li><p><strong>pAdjustMethod = “BH”</strong>: adjustment for multiple testing using the method by Benjamini and Hochberg</p></li>
</ul>
</div>
<div id="step-3-interpretation-of-the-results" class="section level3" number="6.2.3">
<h3>
<span class="header-section-number">6.2.3</span> Step 3: Interpretation of the results<a class="anchor" aria-label="anchor" href="#step-3-interpretation-of-the-results"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Inspect the results table</strong></p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">GSEA_GO</span>, n <span class="op">=</span> <span class="fl">10</span> <span class="op">)</span></span>
<span><span class="co">#&gt;                    ID</span></span>
<span><span class="co">#&gt; GO:0002694 GO:0002694</span></span>
<span><span class="co">#&gt; GO:0045321 GO:0045321</span></span>
<span><span class="co">#&gt; GO:0001775 GO:0001775</span></span>
<span><span class="co">#&gt; GO:0050865 GO:0050865</span></span>
<span><span class="co">#&gt; GO:0050863 GO:0050863</span></span>
<span><span class="co">#&gt; GO:0051249 GO:0051249</span></span>
<span><span class="co">#&gt; GO:0002696 GO:0002696</span></span>
<span><span class="co">#&gt; GO:0050867 GO:0050867</span></span>
<span><span class="co">#&gt; GO:0046649 GO:0046649</span></span>
<span><span class="co">#&gt; GO:0051240 GO:0051240</span></span>
<span><span class="co">#&gt;                                                        Description</span></span>
<span><span class="co">#&gt; GO:0002694                      regulation of leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0045321                                    leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0001775                                         cell activation</span></span>
<span><span class="co">#&gt; GO:0050865                           regulation of cell activation</span></span>
<span><span class="co">#&gt; GO:0050863                         regulation of T cell activation</span></span>
<span><span class="co">#&gt; GO:0051249                     regulation of lymphocyte activation</span></span>
<span><span class="co">#&gt; GO:0002696             positive regulation of leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0050867                  positive regulation of cell activation</span></span>
<span><span class="co">#&gt; GO:0046649                                   lymphocyte activation</span></span>
<span><span class="co">#&gt; GO:0051240 positive regulation of multicellular organismal process</span></span>
<span><span class="co">#&gt;            setSize enrichmentScore       NES       pvalue</span></span>
<span><span class="co">#&gt; GO:0002694     218      -0.4550306 -1.885682 1.443823e-07</span></span>
<span><span class="co">#&gt; GO:0045321     335      -0.4102866 -1.776603 1.721317e-07</span></span>
<span><span class="co">#&gt; GO:0001775     379      -0.3913563 -1.715310 1.627551e-07</span></span>
<span><span class="co">#&gt; GO:0050865     231      -0.4475860 -1.862707 3.379430e-07</span></span>
<span><span class="co">#&gt; GO:0050863     145      -0.5024266 -1.973006 6.812416e-07</span></span>
<span><span class="co">#&gt; GO:0051249     185      -0.4654907 -1.896787 6.286296e-07</span></span>
<span><span class="co">#&gt; GO:0002696     145      -0.4964230 -1.949430 1.090979e-06</span></span>
<span><span class="co">#&gt; GO:0050867     148      -0.4941648 -1.945621 1.030633e-06</span></span>
<span><span class="co">#&gt; GO:0046649     282      -0.4097538 -1.745495 1.526809e-06</span></span>
<span><span class="co">#&gt; GO:0051240     490      -0.3558447 -1.593477 1.476326e-06</span></span>
<span><span class="co">#&gt;                p.adjust       qvalue rank</span></span>
<span><span class="co">#&gt; GO:0002694 0.0002129843 0.0001866874  841</span></span>
<span><span class="co">#&gt; GO:0045321 0.0002129843 0.0001866874  841</span></span>
<span><span class="co">#&gt; GO:0001775 0.0002129843 0.0001866874  782</span></span>
<span><span class="co">#&gt; GO:0050865 0.0003136111 0.0002748900  778</span></span>
<span><span class="co">#&gt; GO:0050863 0.0004214615 0.0003694242  841</span></span>
<span><span class="co">#&gt; GO:0051249 0.0004214615 0.0003694242  910</span></span>
<span><span class="co">#&gt; GO:0002696 0.0005062142 0.0004437126 1417</span></span>
<span><span class="co">#&gt; GO:0050867 0.0005062142 0.0004437126 1417</span></span>
<span><span class="co">#&gt; GO:0046649 0.0005667516 0.0004967755  910</span></span>
<span><span class="co">#&gt; GO:0051240 0.0005667516 0.0004967755  841</span></span>
<span><span class="co">#&gt;                              leading_edge</span></span>
<span><span class="co">#&gt; GO:0002694 tags=26%, list=13%, signal=23%</span></span>
<span><span class="co">#&gt; GO:0045321 tags=23%, list=13%, signal=21%</span></span>
<span><span class="co">#&gt; GO:0001775 tags=22%, list=13%, signal=20%</span></span>
<span><span class="co">#&gt; GO:0050865 tags=24%, list=12%, signal=22%</span></span>
<span><span class="co">#&gt; GO:0050863 tags=29%, list=13%, signal=26%</span></span>
<span><span class="co">#&gt; GO:0051249 tags=28%, list=15%, signal=24%</span></span>
<span><span class="co">#&gt; GO:0002696 tags=42%, list=23%, signal=33%</span></span>
<span><span class="co">#&gt; GO:0050867 tags=41%, list=23%, signal=33%</span></span>
<span><span class="co">#&gt; GO:0046649 tags=24%, list=15%, signal=22%</span></span>
<span><span class="co">#&gt; GO:0051240 tags=24%, list=13%, signal=23%</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            core_enrichment</span></span>
<span><span class="co">#&gt; GO:0002694                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0045321                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000121594/ENSG00000125735/ENSG00000177697/ENSG00000120738/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000197405/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000215021/ENSG00000179388/ENSG00000232629/ENSG00000155760/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000168264/ENSG00000012061/ENSG00000205220/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000153879/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000105438/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000168447/ENSG00000196839/ENSG00000071564/ENSG00000184371/ENSG00000159958/ENSG00000131188/ENSG00000163874/ENSG00000198435/ENSG00000105374/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000103522/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000078902/ENSG00000140968/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000160683/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008/ENSG00000177606</span></span>
<span><span class="co">#&gt; GO:0001775                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000121594/ENSG00000125735/ENSG00000177697/ENSG00000120738/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000154127/ENSG00000204287/ENSG00000197405/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000215021/ENSG00000179388/ENSG00000232629/ENSG00000155760/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000135124/ENSG00000104951/ENSG00000183624/ENSG00000168264/ENSG00000012061/ENSG00000205220/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000153879/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000105438/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000010278/ENSG00000168447/ENSG00000196839/ENSG00000071564/ENSG00000184371/ENSG00000159958/ENSG00000131188/ENSG00000163874/ENSG00000198435/ENSG00000105374/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000103522/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000078902/ENSG00000140968/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000186222/ENSG00000160683/ENSG00000106211/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008/ENSG00000177606</span></span>
<span><span class="co">#&gt; GO:0050865                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000154127/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000010278/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0050863                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000112149/ENSG00000104951/ENSG00000131981/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000160223/ENSG00000113302/ENSG00000167604/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0051249                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000196126/ENSG00000189114/ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000131981/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0002696                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000113811/ENSG00000174130/ENSG00000186827/ENSG00000130475/ENSG00000103653/ENSG00000185950/ENSG00000019582/ENSG00000154096/ENSG00000066336/ENSG00000107968/ENSG00000149273/ENSG00000135426/ENSG00000033327/ENSG00000198502/ENSG00000105639/ENSG00000135334/ENSG00000168040/ENSG00000196126/ENSG00000189114/ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000196839/ENSG00000159958/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000125657/ENSG00000185338/ENSG00000130522/ENSG00000105974/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0050867                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000113811/ENSG00000174130/ENSG00000186827/ENSG00000130475/ENSG00000103653/ENSG00000185950/ENSG00000019582/ENSG00000154096/ENSG00000066336/ENSG00000107968/ENSG00000149273/ENSG00000135426/ENSG00000033327/ENSG00000198502/ENSG00000105639/ENSG00000135334/ENSG00000168040/ENSG00000196126/ENSG00000189114/ENSG00000121594/ENSG00000125735/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000196839/ENSG00000159958/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000125657/ENSG00000185338/ENSG00000130522/ENSG00000105974/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0046649                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000196126/ENSG00000189114/ENSG00000114383/ENSG00000121594/ENSG00000125735/ENSG00000177697/ENSG00000120738/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000215021/ENSG00000179388/ENSG00000232629/ENSG00000155760/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000168264/ENSG00000012061/ENSG00000205220/ENSG00000131981/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000153879/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000105438/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000071564/ENSG00000159958/ENSG00000131188/ENSG00000163874/ENSG00000198435/ENSG00000105374/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000103522/ENSG00000223496/ENSG00000167604/ENSG00000140968/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000160683/ENSG00000185338/ENSG00000171223/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0051240 ENSG00000128604/ENSG00000111012/ENSG00000064393/ENSG00000089094/ENSG00000092820/ENSG00000196126/ENSG00000189114/ENSG00000114383/ENSG00000136997/ENSG00000121594/ENSG00000125735/ENSG00000126262/ENSG00000095739/ENSG00000186350/ENSG00000121966/ENSG00000120738/ENSG00000073150/ENSG00000095059/ENSG00000131979/ENSG00000116670/ENSG00000126353/ENSG00000148834/ENSG00000183655/ENSG00000102103/ENSG00000103811/ENSG00000136717/ENSG00000204287/ENSG00000148926/ENSG00000197405/ENSG00000171791/ENSG00000237541/ENSG00000140678/ENSG00000133321/ENSG00000172977/ENSG00000213015/ENSG00000111725/ENSG00000171236/ENSG00000091972/ENSG00000185507/ENSG00000179388/ENSG00000187109/ENSG00000139146/ENSG00000232629/ENSG00000213626/ENSG00000124920/ENSG00000198053/ENSG00000110092/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000176788/ENSG00000060138/ENSG00000161638/ENSG00000111424/ENSG00000119508/ENSG00000140995/ENSG00000090776/ENSG00000158470/ENSG00000125726/ENSG00000153879/ENSG00000164442/ENSG00000117090/ENSG00000105372/ENSG00000197329/ENSG00000108561/ENSG00000130669/ENSG00000153487/ENSG00000204592/ENSG00000107338/ENSG00000173511/ENSG00000196839/ENSG00000184371/ENSG00000159958/ENSG00000111087/ENSG00000170345/ENSG00000163794/ENSG00000163874/ENSG00000073712/ENSG00000103257/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000168056/ENSG00000176170/ENSG00000232810/ENSG00000168779/ENSG00000186652/ENSG00000223496/ENSG00000101670/ENSG00000215301/ENSG00000131408/ENSG00000167604/ENSG00000124216/ENSG00000169992/ENSG00000140564/ENSG00000140968/ENSG00000105245/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000125398/ENSG00000143507/ENSG00000156127/ENSG00000196843/ENSG00000125657/ENSG00000106211/ENSG00000130222/ENSG00000185338/ENSG00000143878/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008/ENSG00000123689</span></span></code></pre></div>
<p>Columns in the results table:</p>
<ul>
<li><p><strong>ID</strong>: ID of gene set</p></li>
<li><p><strong>Description</strong>: Description of Gene Set</p></li>
<li><p><strong>setSize</strong>: Size of the gene set</p></li>
<li>
<p><strong>enrichmentScore</strong>: Enrichment Score</p>
<ul>
<li>note: this enrichment score has not been normalized for gene set size</li>
<li>means: larger gene sets automatically have a bigger (absolute) enrichment score, independent of actual differential enrichment</li>
<li>the raw enrichment score is therefore not comparable between different gene sets</li>
</ul>
</li>
<li>
<p><strong>NES</strong>: Normalized version of column enrichmentScore</p>
<ul>
<li>NES can be compared between different gene sets</li>
</ul>
</li>
<li>
<p><strong>pvalue</strong>: p-value of enrichment of given gene set</p>
<ul>
<li>note: this raw p-value has not been adjusted for multiple testing</li>
<li>therefore: CANNOT be used to assess differential enrichment of a given gene set</li>
</ul>
</li>
<li>
<p><strong>p.adjust</strong>: <strong>adjusted</strong> p-value of a given gene set.</p>
<ul>
<li>Note that this p-value now has been adjusted for multiple testing and can therefore be used to assess differential enrichment. For instance, detect all genes with p.adjust &lt; 0.05 as differentially enriched.</li>
</ul>
</li>
<li><p><strong>qvalue</strong>: <strong>adjusted</strong> p-value of a given gene set. Note that a qvalue is the analog to p.adjust, but adjusted for multiple testing using a different method.</p></li>
</ul>
<p>Note that you can either use the column p.adjust or qvalue to assess whether a gene set is differentially enriched or not.</p>
<ul>
<li><p><strong>rank</strong>: position in the ranked list of the genes in which the maximum difference between the two sums occurs, i.e. the rank at which the enrichment score is extracted.</p></li>
<li>
<p><strong>leading_edge</strong>:</p>
<ul>
<li><p><strong>tags</strong>: The percentage of genes in the ranked list that are members of gene set before (for positive enrichment score) or after (for negative enrichment score) the position from which the enrichment score is extracted.</p></li>
<li><p><strong>list</strong>: The percentage of genes in the ranked gene list before (for positive enrichment
score) or after (for negative enrichment score) the position from which the enrichment score is extracted.</p></li>
<li><p><strong>signal</strong>: enrichment signal strength
combines the statistics “tags” and “list”</p></li>
<li><p><strong>core_enrichment</strong>: genes that contribute most to the enrichment results</p></li>
</ul>
</li>
</ul>
<p>Note that a more detailed information on leading_edge and core_enrichment can be found in the user manual provided for GSEA:
(<a href="https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html" class="uri">https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html</a>):</p>
<ul>
<li><p>search for <strong>Detailed Enrichment Results</strong> (section “Leading Edge”)</p></li>
<li><p>search for <strong>CORE ENRICHMENT</strong></p></li>
</ul>
</div>
</div>
<div id="researchers-degrees-of-freedom-1" class="section level2" number="6.3">
<h2>
<span class="header-section-number">6.3</span> Researchers’ degrees of freedom<a class="anchor" aria-label="anchor" href="#researchers-degrees-of-freedom-1"><i class="fas fa-link"></i></a>
</h2>
<p>In this part, we will go through all parameters that can be adapted in the GOSeq workflow. It is important to note that the intention behind going through the researchers’ degrees of freedom is to give you an understanding of what you can do to adapt the given (parameter) setting to the research question. It is even more important to keep in mind that the intention behind going through these flexible parameters is NOT to change them in order to help you obtain the most preferable results by systematically changing these parameters since such behaviour would correspond to “cherry-picking”. Any changes in the parameter choice must be documented transparently.</p>
<div id="change-1-change-exponent" class="section level3" number="6.3.1">
<h3>
<span class="header-section-number">6.3.1</span> Change 1: change exponent<a class="anchor" aria-label="anchor" href="#change-1-change-exponent"><i class="fas fa-link"></i></a>
</h3>
<p>Note that, in contrast to the web-based tool GSEA, clusterProfiler does not make any indication as to which alternative exponent values are allowed or sensible. One can stick with the web-based tool GSEA which suggests, in addition to the default exponent value of 1, the alternative values 0, 1.5, and 2.</p>
<p>Here, we change the exponent value to 1.5.</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># important: set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSEA_GO_exponent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html">gseGO</a></span><span class="op">(</span>geneList <span class="op">=</span> <span class="va">rankvec_Ensembl</span>,</span>
<span>                            ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                            OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                            keyType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>                            exp <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>                            seed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><strong>Inspect the results</strong></p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">GSEA_GO_exponent</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> </span>
<span><span class="co">#&gt;                    ID                         Description</span></span>
<span><span class="co">#&gt; GO:0042572 GO:0042572           retinol metabolic process</span></span>
<span><span class="co">#&gt; GO:0045321 GO:0045321                leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0001775 GO:0001775                     cell activation</span></span>
<span><span class="co">#&gt; GO:0006721 GO:0006721         terpenoid metabolic process</span></span>
<span><span class="co">#&gt; GO:0001523 GO:0001523          retinoid metabolic process</span></span>
<span><span class="co">#&gt; GO:0002694 GO:0002694  regulation of leukocyte activation</span></span>
<span><span class="co">#&gt; GO:0050865 GO:0050865       regulation of cell activation</span></span>
<span><span class="co">#&gt; GO:0055088 GO:0055088                   lipid homeostasis</span></span>
<span><span class="co">#&gt; GO:0051249 GO:0051249 regulation of lymphocyte activation</span></span>
<span><span class="co">#&gt; GO:0006629 GO:0006629             lipid metabolic process</span></span>
<span><span class="co">#&gt;            setSize enrichmentScore       NES       pvalue</span></span>
<span><span class="co">#&gt; GO:0042572      16      -0.9375514 -1.999582 1.516074e-06</span></span>
<span><span class="co">#&gt; GO:0045321     335      -0.5323438 -1.717534 8.735337e-06</span></span>
<span><span class="co">#&gt; GO:0001775     379      -0.5155165 -1.681090 8.893550e-06</span></span>
<span><span class="co">#&gt; GO:0006721      24      -0.8807910 -2.037460 1.655703e-05</span></span>
<span><span class="co">#&gt; GO:0001523      20      -0.9263665 -2.057425 3.114786e-05</span></span>
<span><span class="co">#&gt; GO:0002694     218      -0.5788244 -1.797112 2.686939e-05</span></span>
<span><span class="co">#&gt; GO:0050865     231      -0.5719874 -1.781633 3.022575e-05</span></span>
<span><span class="co">#&gt; GO:0055088      52      -0.7733441 -2.026849 5.967564e-05</span></span>
<span><span class="co">#&gt; GO:0051249     185      -0.5863351 -1.798671 7.530031e-05</span></span>
<span><span class="co">#&gt; GO:0006629     441      -0.4774551 -1.573053 6.777393e-05</span></span>
<span><span class="co">#&gt;               p.adjust      qvalue rank</span></span>
<span><span class="co">#&gt; GO:0042572 0.005627665 0.005122733   22</span></span>
<span><span class="co">#&gt; GO:0045321 0.011004286 0.010016946  782</span></span>
<span><span class="co">#&gt; GO:0001775 0.011004286 0.010016946  782</span></span>
<span><span class="co">#&gt; GO:0006721 0.015364923 0.013986333   22</span></span>
<span><span class="co">#&gt; GO:0001523 0.016517266 0.015035284   22</span></span>
<span><span class="co">#&gt; GO:0002694 0.016517266 0.015035284  778</span></span>
<span><span class="co">#&gt; GO:0050865 0.016517266 0.015035284  778</span></span>
<span><span class="co">#&gt; GO:0055088 0.027689495 0.025205104  285</span></span>
<span><span class="co">#&gt; GO:0051249 0.027951474 0.025443577  778</span></span>
<span><span class="co">#&gt; GO:0006629 0.027951474 0.025443577  836</span></span>
<span><span class="co">#&gt;                              leading_edge</span></span>
<span><span class="co">#&gt; GO:0042572  tags=12%, list=0%, signal=12%</span></span>
<span><span class="co">#&gt; GO:0045321 tags=22%, list=13%, signal=20%</span></span>
<span><span class="co">#&gt; GO:0001775 tags=21%, list=13%, signal=19%</span></span>
<span><span class="co">#&gt; GO:0006721  tags=12%, list=0%, signal=13%</span></span>
<span><span class="co">#&gt; GO:0001523  tags=10%, list=0%, signal=10%</span></span>
<span><span class="co">#&gt; GO:0002694 tags=25%, list=12%, signal=22%</span></span>
<span><span class="co">#&gt; GO:0050865 tags=24%, list=12%, signal=22%</span></span>
<span><span class="co">#&gt; GO:0055088  tags=29%, list=5%, signal=28%</span></span>
<span><span class="co">#&gt; GO:0051249 tags=25%, list=12%, signal=23%</span></span>
<span><span class="co">#&gt; GO:0006629 tags=19%, list=13%, signal=18%</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            core_enrichment</span></span>
<span><span class="co">#&gt; GO:0042572                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000062282/ENSG00000006757</span></span>
<span><span class="co">#&gt; GO:0045321                                                                                                                                                 ENSG00000120738/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000197405/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000215021/ENSG00000179388/ENSG00000232629/ENSG00000155760/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000168264/ENSG00000012061/ENSG00000205220/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000153879/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000105438/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000168447/ENSG00000196839/ENSG00000071564/ENSG00000184371/ENSG00000159958/ENSG00000131188/ENSG00000163874/ENSG00000198435/ENSG00000105374/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000103522/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000078902/ENSG00000140968/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000160683/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008/ENSG00000177606</span></span>
<span><span class="co">#&gt; GO:0001775                                                                 ENSG00000120738/ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000154127/ENSG00000204287/ENSG00000197405/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000215021/ENSG00000179388/ENSG00000232629/ENSG00000155760/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000135124/ENSG00000104951/ENSG00000183624/ENSG00000168264/ENSG00000012061/ENSG00000205220/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000090339/ENSG00000125726/ENSG00000153879/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000105438/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000010278/ENSG00000168447/ENSG00000196839/ENSG00000071564/ENSG00000184371/ENSG00000159958/ENSG00000131188/ENSG00000163874/ENSG00000198435/ENSG00000105374/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000103522/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000078902/ENSG00000140968/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000186222/ENSG00000160683/ENSG00000106211/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008/ENSG00000177606</span></span>
<span><span class="co">#&gt; GO:0006721                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000145545/ENSG00000062282/ENSG00000006757</span></span>
<span><span class="co">#&gt; GO:0001523                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000062282/ENSG00000006757</span></span>
<span><span class="co">#&gt; GO:0002694                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0050865                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000154127/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000091972/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000126368/ENSG00000131981/ENSG00000119508/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000010278/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000176170/ENSG00000232810/ENSG00000223496/ENSG00000167604/ENSG00000130164/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000130522/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0055088                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000126368/ENSG00000109084/ENSG00000073060/ENSG00000167695/ENSG00000119655/ENSG00000087237/ENSG00000105698/ENSG00000204444/ENSG00000101670/ENSG00000215301/ENSG00000131408/ENSG00000130164/ENSG00000062282/ENSG00000105974/ENSG00000006757</span></span>
<span><span class="co">#&gt; GO:0051249                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSG00000095059/ENSG00000116670/ENSG00000126353/ENSG00000183655/ENSG00000204287/ENSG00000171791/ENSG00000237541/ENSG00000172977/ENSG00000179388/ENSG00000232629/ENSG00000198053/ENSG00000164430/ENSG00000127666/ENSG00000112149/ENSG00000104951/ENSG00000183624/ENSG00000131981/ENSG00000090776/ENSG00000125726/ENSG00000108861/ENSG00000117090/ENSG00000197329/ENSG00000204592/ENSG00000107338/ENSG00000213689/ENSG00000196839/ENSG00000159958/ENSG00000163874/ENSG00000198435/ENSG00000167775/ENSG00000160223/ENSG00000113302/ENSG00000223496/ENSG00000167604/ENSG00000101017/ENSG00000110944/ENSG00000105246/ENSG00000002330/ENSG00000143507/ENSG00000156127/ENSG00000125657/ENSG00000185338/ENSG00000171223/ENSG00000172216/ENSG00000105974/ENSG00000101665/ENSG00000115008</span></span>
<span><span class="co">#&gt; GO:0006629 ENSG00000101255/ENSG00000100889/ENSG00000141858/ENSG00000184207/ENSG00000099377/ENSG00000221968/ENSG00000079459/ENSG00000166908/ENSG00000120738/ENSG00000074660/ENSG00000126353/ENSG00000183260/ENSG00000183655/ENSG00000149084/ENSG00000130766/ENSG00000148926/ENSG00000133321/ENSG00000172977/ENSG00000163964/ENSG00000242372/ENSG00000110536/ENSG00000166394/ENSG00000111725/ENSG00000165175/ENSG00000174915/ENSG00000215021/ENSG00000103502/ENSG00000181513/ENSG00000169710/ENSG00000112531/ENSG00000121769/ENSG00000129474/ENSG00000100979/ENSG00000204386/ENSG00000133328/ENSG00000103642/ENSG00000126368/ENSG00000119508/ENSG00000125741/ENSG00000158470/ENSG00000145545/ENSG00000123179/ENSG00000213689/ENSG00000168447/ENSG00000073060/ENSG00000158715/ENSG00000119655/ENSG00000167468/ENSG00000165782/ENSG00000087237/ENSG00000083807/ENSG00000100294/ENSG00000176170/ENSG00000232810/ENSG00000196743/ENSG00000179598/ENSG00000167969/ENSG00000101670/ENSG00000167130/ENSG00000131408/ENSG00000173599/ENSG00000124216/ENSG00000120833/ENSG00000130164/ENSG00000086544/ENSG00000167508/ENSG00000089163/ENSG00000125398/ENSG00000087076/ENSG00000111666/ENSG00000110921/ENSG00000174326/ENSG00000100600/ENSG00000147383/ENSG00000185338/ENSG00000125652/ENSG00000062282/ENSG00000105974/ENSG00000172893/ENSG00000101846/ENSG00000196407/ENSG00000115008/ENSG00000006757</span></span></code></pre></div>
<p>When switching the exponent to 2, the number of differentially enriched gene sets has decreased to two.</p>

</div>
</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></div>
<div class="next"><a href="david.html"><span class="header-section-number">7</span> DAVID</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li>
<a class="nav-link" href="#clusterprofilers-gsea-with-gene-set-database-go"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#libraries-5"><span class="header-section-number">6.0.1</span> Libraries</a></li></ul>
</li>
<li><a class="nav-link" href="#load-data-4"><span class="header-section-number">6.1</span> Load data</a></li>
<li>
<a class="nav-link" href="#run-gsea"><span class="header-section-number">6.2</span> Run GSEA</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#step-1-generation-of-required-input-object"><span class="header-section-number">6.2.1</span> Step 1: Generation of required input object</a></li>
<li><a class="nav-link" href="#step-2-run-gsea-with-gene-set-database-go"><span class="header-section-number">6.2.2</span> Step 2: Run GSEA with gene set database GO</a></li>
<li><a class="nav-link" href="#step-3-interpretation-of-the-results"><span class="header-section-number">6.2.3</span> Step 3: Interpretation of the results</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#researchers-degrees-of-freedom-1"><span class="header-section-number">6.3</span> Researchers’ degrees of freedom</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#change-1-change-exponent"><span class="header-section-number">6.3.1</span> Change 1: change exponent</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
