<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 10 clusterProfiler’s ORA | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will run clusterProfiler’s ORA tool based on the results table of differential expression analysis for the gene set databases KEGG GO (with default subontology “MF”) go through...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 10 clusterProfiler’s ORA | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will run clusterProfiler’s ORA tool based on the results table of differential expression analysis for the gene set databases KEGG GO (with default subontology “MF”) go through...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 10 clusterProfiler’s ORA | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will run clusterProfiler’s ORA tool based on the results table of differential expression analysis for the gene set databases KEGG GO (with default subontology “MF”) go through...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="active" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clusterprofilers-ora" class="section level1" number="10">
<h1>
<span class="header-section-number">10</span> clusterProfiler’s ORA<a class="anchor" aria-label="anchor" href="#clusterprofilers-ora"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will</p>
<ol style="list-style-type: decimal">
<li>run clusterProfiler’s ORA tool based on the results table of differential expression analysis for the gene set databases</li>
</ol>
<ul>
<li><p>KEGG</p></li>
<li><p>GO (with default subontology “MF”)</p></li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>go through all meaningful researchers’ degrees of freedom</li>
</ol>
<p>Note that clusterProfiler’s ORA accepts the gene IDs in the NCBI (Entrez) ID format, independent of the chosen gene set database.</p>
<div id="libraries-9" class="section level2" number="10.1">
<h2>
<span class="header-section-number">10.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-9"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load Libraries</strong></p>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre></div>
<p>Description of the packages</p>
<ul>
<li><p><strong>clusterProfiler</strong>: We here use clusterProfiler’s implementation of ORA.</p></li>
<li><p><strong>org.Hs.eg.db</strong>: Provides genome-wide annotation for humans. When working with a different organism, the user has to provide a different package (see Chapter 1 ‘About’). Note that this library is required when running clusterProfiler’s ORA with gene set database GO.</p></li>
</ul>
</div>
<div id="load-data-8" class="section level2" number="10.2">
<h2>
<span class="header-section-number">10.2</span> Load data<a class="anchor" aria-label="anchor" href="#load-data-8"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_limma_Entrez.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_DESeq2_Entrez.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_Differential_Expression_Analysis/DE_results_edgeR_Entrez.Rdata"</span><span class="op">)</span></span></code></pre></div>
<p>We arbitrarily set the object DE_results we will work with resulting from DESeq2. However, you can switch around at your discretion.</p>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DE_results</span> <span class="op">&lt;-</span> <span class="va">DE_results_DESeq2_Entrez</span></span>
<span></span>
<span><span class="co"># alternatively: </span></span>
<span><span class="co">#DE_results &lt;- DE_results_limma_Entrez</span></span>
<span><span class="co">#DE_results &lt;- DE_results_edgeR_Entrez</span></span></code></pre></div>
</div>
<div id="run-clusterprofilers-ora" class="section level2" number="10.3">
<h2>
<span class="header-section-number">10.3</span> Run clusterProfiler’s ORA<a class="anchor" aria-label="anchor" href="#run-clusterprofilers-ora"><i class="fas fa-link"></i></a>
</h2>
<div id="step-1-preparation-of-required-input-object-1" class="section level3" number="10.3.1">
<h3>
<span class="header-section-number">10.3.1</span> step 1: Preparation of required input object<a class="anchor" aria-label="anchor" href="#step-1-preparation-of-required-input-object-1"><i class="fas fa-link"></i></a>
</h3>
<p>clusterProfiler’s ORA requires as input a list of differentially expressed genes. We extract such list from the results table of differential expression analysis.<br>
For each gene from the results of differential expression analysis, we indicate whether it is differentially expressed (TRUE) or not differentially expressed (FALSE) based on the following two criteria:</p>
<ol style="list-style-type: decimal">
<li><p>it was tested for differential expression, i.e. has a non-missing adjusted p-value AND</p></li>
<li><p>it has an adjusted p-value &lt; 0.05</p></li>
</ol>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># indicate which gene fulfills BOTH criteria: </span></span>
<span><span class="va">ind_differentially_expressed</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">$</span><span class="va">p_adj</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">DE_results</span><span class="op">$</span><span class="va">p_adj</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using this indicator, we extract the vector of differentially expressed genes from the results</span></span>
<span><span class="co"># of differential expression analysis </span></span>
<span><span class="va">DEG_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">[</span><span class="va">ind_differentially_expressed</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>inspect the first few genes from the input list of differentially expressed genes</p>
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DEG_vec</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "8228"   "23136"  "55787"  "10732"  "412"    "219699"</span></span>
<span><span class="co">#&gt;  [7] "6615"   "6192"   "8226"   "10912"</span></span></code></pre></div>
</div>
<div id="step-2-run-clusterprofilers-ora" class="section level3" number="10.3.2">
<h3>
<span class="header-section-number">10.3.2</span> step 2: Run clusterProfiler’s ORA<a class="anchor" aria-label="anchor" href="#step-2-run-clusterprofilers-ora"><i class="fas fa-link"></i></a>
</h3>
<p>ORA can be run with the common geneset databases KEGG and GO as well as user-defined gene set databases. Here, we focus on the two common gene set databases GO and KEGG.</p>
<div id="option-1-gene-set-database-go" class="section level4" number="10.3.2.1">
<h4>
<span class="header-section-number">10.3.2.1</span> option 1: gene set database GO<a class="anchor" aria-label="anchor" href="#option-1-gene-set-database-go"><i class="fas fa-link"></i></a>
</h4>
<p>Here, we work the the subontology specified by default, namely Molecular Function (“MF”)</p>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ORA_results_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">DEG_vec</span>, </span>
<span>                           OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, </span>
<span>                           ont <span class="op">=</span> <span class="st">"MF"</span><span class="op">)</span> </span></code></pre></div>
<p>Inspect the results table</p>
<div class="sourceCode" id="cb159"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ORA_results_GO</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] ID          Description GeneRatio   BgRatio    </span></span>
<span><span class="co">#&gt; [5] pvalue      p.adjust    qvalue      geneID     </span></span>
<span><span class="co">#&gt; [9] Count      </span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li><p><strong>gene</strong>: vector of differentially expressed genes</p></li>
<li><p><strong>OrgDb</strong>: annotation package for organism at hand (here: human)</p></li>
<li><p><strong>ont</strong>: subontology (“MF” by default, alternatives: “BP” and “CC”)</p></li>
</ul>
</div>
<div id="option-2-gene-set-database-kegg" class="section level4" number="10.3.2.2">
<h4>
<span class="header-section-number">10.3.2.2</span> option 2: gene set database KEGG<a class="anchor" aria-label="anchor" href="#option-2-gene-set-database-kegg"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb160"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ORA_results_KEGG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichKEGG.html">enrichKEGG</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">DEG_vec</span>, </span>
<span>                               organism <span class="op">=</span> <span class="st">"hsa"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading KEGG annotation online: "https://rest.kegg.jp/link/hsa/pathway"...</span></span>
<span><span class="co">#&gt; Reading KEGG annotation online: "https://rest.kegg.jp/list/pathway/hsa"...</span></span>
<span></span>
<span><span class="co"># inspect results table </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ORA_results_KEGG</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] ID          Description GeneRatio   BgRatio    </span></span>
<span><span class="co">#&gt; [5] pvalue      p.adjust    qvalue      geneID     </span></span>
<span><span class="co">#&gt; [9] Count      </span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li><p><strong>gene</strong>: vector of differentially expressed genes</p></li>
<li>
<p><strong>organism</strong>: organism from which gene expression measurements are obtained</p>
<ul>
<li><p>by default, it is set ‘organism = “hsa”’.</p></li>
<li><p>must be adapted for other organisms (such as organism = “mmu” for mice)</p></li>
</ul>
</li>
</ul>
</div>
</div>
<div id="step-4-interpretation-of-the-results" class="section level3" number="10.3.3">
<h3>
<span class="header-section-number">10.3.3</span> step 4: Interpretation of the results<a class="anchor" aria-label="anchor" href="#step-4-interpretation-of-the-results"><i class="fas fa-link"></i></a>
</h3>
<p>Description of columns in results table:</p>
<ul>
<li><p><strong>GeneRatio</strong>: number of genes from the input list that are members of the given gene set divided by the number of genes from the input list that are NOT members of the given gene set</p></li>
<li><p><strong>BgRatio</strong>: number of genes from the universe that are members of the gene set divided by the total number of genes in the universe</p></li>
<li><p><strong>pvalue</strong>: p-value of enrichment of the given gene set</p></li>
<li><p><strong>p.adjust</strong>: p-value of enrichment ADJUSTED for multiple testing</p></li>
<li>
<p><strong>qvalue</strong>: p-value of enrichment ADJUSTED for multiple testing</p>
<ul>
<li>note: p.adjust and qvalue are adjusted using slightly different approaches</li>
</ul>
</li>
<li><p><strong>geneID</strong>: list of genes from the input list that are members of the given gene set</p></li>
<li><p><strong>count</strong>: number of genes from the input list that are members of the given gene set</p></li>
</ul>
</div>
</div>
<div id="researchers-degrees-of-freedom-3" class="section level2" number="10.4">
<h2>
<span class="header-section-number">10.4</span> Researchers’ degrees of freedom<a class="anchor" aria-label="anchor" href="#researchers-degrees-of-freedom-3"><i class="fas fa-link"></i></a>
</h2>
<p>In this part, we will go through all parameters that can be adapted in the GOSeq workflow. It is important to note that the intention behind going through the researchers’ degrees of freedom is to give you an understanding of what you can do to adapt the given (parameter) setting to the research question. It is even more important to keep in mind that the intention behind going through these flexible parameters is NOT to change them in order to help you obtain the most preferable results by systematically changing these parameters since such behaviour would correspond to “cherry-picking”. Any changes in the parameter choice must be documented transparently.</p>
<div id="change-1-change-universe" class="section level3" number="10.4.1">
<h3>
<span class="header-section-number">10.4.1</span> change 1: Change universe<a class="anchor" aria-label="anchor" href="#change-1-change-universe"><i class="fas fa-link"></i></a>
</h3>
<p>Here, we change the universe to all genes measured in the experiment.</p>
<p>Note that we do not change the universe to the interception between all genes from the experiment and the list of genes annotated to the given gene set database since we found no way to obtain the latter list of genes.<br>
Also note that we want to restrict ourselves to all genes in the experiment that HAVE an adjusted p-value (i.e. whose expression was indeed measured). The intuition is that, e.g. for DESeq2, some genes are filtered out internally and
therefore do not have an adjusted p-value. These genes therefore neither be detected as differentially expressed or not differentially expressed so it would not be feasible to include them in the universe.</p>
<p><strong>Step 1: Set up alternative universe</strong></p>
<div class="sourceCode" id="cb161"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (i) indicate which genes have an adjusted p-value</span></span>
<span><span class="va">ind_adjp</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">$</span><span class="va">p_adj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (ii) filter the genes from the experiment to those genes that do have an adjusted p-value </span></span>
<span><span class="va">alternative_universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">)</span><span class="op">[</span><span class="va">ind_adjp</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># inspect the first few genes in the universe</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">alternative_universe</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "8813"  "57147" "55732" "2268"  "2519"  "4800"  "81887"</span></span>
<span><span class="co">#&gt;  [8] "22875" "5893"  "572"</span></span></code></pre></div>
<p><strong>Step 2: Add alternative universe as a parameter to ORA</strong></p>
<p>In both functions <em>enrichGO()</em> and <em>enrichKEGG()</em>, an alternative universe can be specified in argument <strong>universe</strong>.</p>
<p><strong>(a) gene set database GO: specify parameter universe</strong></p>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ORA_results_GO_universe</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">DEG_vec</span>, </span>
<span>                                     OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, </span>
<span>                                     universe <span class="op">=</span> <span class="va">alternative_universe</span><span class="op">)</span></span></code></pre></div>
<p>Inspect the results table:</p>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ORA_results_GO_universe</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] ID          Description GeneRatio   BgRatio    </span></span>
<span><span class="co">#&gt; [5] pvalue      p.adjust    qvalue      geneID     </span></span>
<span><span class="co">#&gt; [9] Count      </span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre></div>
<p><strong>(b) gene set database KEGG: specify parameter universe</strong></p>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ORA_results_KEGG_universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichKEGG.html">enrichKEGG</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">DEG_vec</span>, </span>
<span>                                        organism <span class="op">=</span> <span class="st">"hsa"</span>,</span>
<span>                                        universe <span class="op">=</span> <span class="va">alternative_universe</span><span class="op">)</span></span></code></pre></div>
<p>Inspect the results table:</p>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ORA_results_KEGG_universe</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] ID          Description GeneRatio   BgRatio    </span></span>
<span><span class="co">#&gt; [5] pvalue      p.adjust    qvalue      geneID     </span></span>
<span><span class="co">#&gt; [9] Count      </span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre></div>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></div>
<div class="next"><a href="padog.html"><span class="header-section-number">11</span> PADOG</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clusterprofilers-ora"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="nav-link" href="#libraries-9"><span class="header-section-number">10.1</span> Libraries</a></li>
<li><a class="nav-link" href="#load-data-8"><span class="header-section-number">10.2</span> Load data</a></li>
<li>
<a class="nav-link" href="#run-clusterprofilers-ora"><span class="header-section-number">10.3</span> Run clusterProfiler’s ORA</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#step-1-preparation-of-required-input-object-1"><span class="header-section-number">10.3.1</span> step 1: Preparation of required input object</a></li>
<li><a class="nav-link" href="#step-2-run-clusterprofilers-ora"><span class="header-section-number">10.3.2</span> step 2: Run clusterProfiler’s ORA</a></li>
<li><a class="nav-link" href="#step-4-interpretation-of-the-results"><span class="header-section-number">10.3.3</span> step 4: Interpretation of the results</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#researchers-degrees-of-freedom-3"><span class="header-section-number">10.4</span> Researchers’ degrees of freedom</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#change-1-change-universe"><span class="header-section-number">10.4.1</span> change 1: Change universe</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
