<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 3 Differential expression analysis | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will perform differential expression analysis for each of the three parametric methods: voom/limma DESeq2 edgeR Here, we illustrate this process for two gene ID format Entrez...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 3 Differential expression analysis | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will perform differential expression analysis for each of the three parametric methods: voom/limma DESeq2 edgeR Here, we illustrate this process for two gene ID format Entrez...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 3 Differential expression analysis | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will perform differential expression analysis for each of the three parametric methods: voom/limma DESeq2 edgeR Here, we illustrate this process for two gene ID format Entrez...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="active" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="differential-expression-analysis" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will perform differential expression analysis for each of the three parametric methods:</p>
<ul>
<li><p>voom/limma</p></li>
<li><p>DESeq2</p></li>
<li><p>edgeR</p></li>
</ul>
<p>Here, we illustrate this process for two gene ID format Entrez IDs, while we additionally present the identical process for the format Ensembl ID in the R script “Instructions_Differential_Expression_Analysis.R”.</p>
<p>Note that at at the end of the illustration for each method, we rename some of the columns in the corresponding results tables to unify them across the different methods. While this step is not required for differential expression analysis or for the subsequent use of gene set analysis IN GENERAL, renaming of some columns (such as adjusted p-value) is necessary in this context so that we can use the same code to illustrate further conduct of GSA in the following chapters, independent of the method for differential expression analysis used.</p>
<div id="libraries-2" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-2"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"tweeDESeqCountData"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"edgeR"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load libraries</strong></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.creal.cat/jrgonzalez/software.htm">tweeDEseqCountData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<p>Description of the libraries:</p>
<ul>
<li><p><strong>tweeDEseqCountData</strong>: In addition the the pre-filtered gene expression data set (with converted gene IDs) that we load in the next step, we need the conditions of the samples which are provided by tweeDEseqCountData.</p></li>
<li><p><strong>limma</strong>: First option of methods for differential expression analysis.</p></li>
<li><p><strong>DESeq2</strong>: Second option of methods for differential expression analysis.</p></li>
<li><p><strong>edgeR</strong>: Third option of methods for differential expression analysis.</p></li>
<li><p><strong>dplyr</strong>: Used to unify some of the columns in the results table of the three methods for differential expression analysis.</p></li>
</ul>
<div id="load-data-1" class="section level3" number="3.1.1">
<h3>
<span class="header-section-number">3.1.1</span> Load data<a class="anchor" aria-label="anchor" href="#load-data-1"><i class="fas fa-link"></i></a>
</h3>
<p>Note that depending on the method for differential expression analysis, different methods for pre-filtering are proposed:</p>
<ul>
<li>
<strong>voom/limma</strong> and <strong>egdeR</strong>: pre-filtering using edgeR’s built-in function filterByExpr()</li>
<li>
<strong>DESeq2</strong>: manual/simple pre-filtering</li>
</ul>
<p>We therefore work with both pre-filtered gene expression data sets generated in Chapter ‘Pre-Filtering’:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_GeneID_Conversion_DuplicatesRemoval/exprdat_filter_conv_DESeq2.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_GeneID_Conversion_DuplicatesRemoval/exprdat_filter_conv_filterByExpr.Rdata"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Obtain the sample conditions</strong>
Obtain the sample conditions (i.e. phenotypes) of pickrell data set and store them in object “sample_conditions”:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load pickrell data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pickrell</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the conditions/phenotypes of the samples can be addressed with the following syntax: </span></span>
<span><span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span></span>
<span><span class="co">#&gt;  [1] male   male   female male   female male   female male  </span></span>
<span><span class="co">#&gt;  [9] female male   female male   female male   female male  </span></span>
<span><span class="co">#&gt; [17] female female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [25] female male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [33] female male   female female female female male   female</span></span>
<span><span class="co">#&gt; [41] male   male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [49] female female male   female male   male   female female</span></span>
<span><span class="co">#&gt; [57] male   female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [65] female female female male   female</span></span>
<span><span class="co">#&gt; Levels: female male</span></span>
<span></span>
<span><span class="co"># store sample conditions </span></span>
<span><span class="va">sample_conditions</span> <span class="op">&lt;-</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span></span></code></pre></div>
</div>
</div>
<div id="differential-expression-analysis-with-entrez-gene-id" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Differential Expression Analysis (with Entrez gene ID)<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-with-entrez-gene-id"><i class="fas fa-link"></i></a>
</h2>
<p>A brief description of the methods for differential expression analysis is provided in the supplement.</p>
<div id="option-1-differential-expression-analysis-using-limma" class="section level3" number="3.2.1">
<h3>
<span class="header-section-number">3.2.1</span> Option 1: Differential expression analysis using limma<a class="anchor" aria-label="anchor" href="#option-1-differential-expression-analysis-using-limma"><i class="fas fa-link"></i></a>
</h3>
<p><strong>1. Generate required input object</strong></p>
<p>At this point, the gene expression data set ‘exprdat_filter_conv_filterByExpr’ is
a data frame and needs to be converted to a DGEList object.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># store expression data with corresponding sample conditions in object of class DGEList</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">exprdat_filter_conv_filterByExpr</span>, </span>
<span>              group <span class="op">=</span> <span class="va">sample_conditions</span><span class="op">)</span></span></code></pre></div>
<p>Required function arguments:</p>
<ul>
<li>
<strong>counts</strong>: matrix that contains RNA-Seq data (i.e. count data)</li>
</ul>
<p>Optional function arguments:</p>
<ul>
<li>
<strong>group</strong>: indicates condition of each sample/library. This argument must be specified sooner or later (such as in subsequent functions) so we just specify it at this point</li>
</ul>
<p>Note: we leave the remaining arguments in their default state since the corresponding info will be added through the following pipeline of functions.</p>
<p><strong>2. Normalization</strong></p>
<p>The following piece of code generates a normalization factor for each sample which accounts for sample-specific effects (such as differences in library sizes and effects of compositionality). If not accounted for, these effects prevent a comparison between the samples.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p>Note that this function does not transform the count data, but rather generates a normalization factor for each sample which is separately incorporated into the subsequent analysis.</p>
<p><strong>3. voom transformation</strong></p>
<p>Perform the voom<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;Charity W Law et al., &lt;span&gt;“Voom: Precision Weights Unlock Linear Model Analysis Tools for &lt;span class="nocase"&gt;RNA-seq&lt;/span&gt; Read Counts,”&lt;/span&gt; &lt;em&gt;Genome Biology&lt;/em&gt; 15, no. 2 (2014): 1–17.&lt;/p&gt;'><sup>14</sup></a></span> transformation to make limma, which was initially developed for microarray measurements, available to RNA-Seq data. voom transforms the RNA-Seq data to log counts-per-million and generates a precision weight for each entry in the gene expression data sets to be incorporated into the limma pipeline.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (i) generate design matrix (rows correspond to samples, columns indicate which coefficients are to be estimated)</span></span>
<span><span class="va">design_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">sample_conditions</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (ii) voom transformation </span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html">voom</a></span><span class="op">(</span><span class="va">y</span>, design <span class="op">=</span> <span class="va">design_matrix</span><span class="op">)</span></span></code></pre></div>
<p><strong>4. Test for differential expression</strong></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (i) fit a linear model for each gene</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (ii) calculate the statistics for the assessment of differential expression </span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (iii) Get the result table for each gene whose differential expression was assessed </span></span>
<span><span class="va">DE_results_limma_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span>fit <span class="op">=</span> <span class="va">y</span>, </span>
<span>                                    number <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># note: 'number = nrow(y)' ensures that all genes are displayed in results</span></span></code></pre></div>
<p><strong>5. Rename columns in results of differential expression analysis</strong></p>
<p>We rename some of the columns in the results table to unify them across the different methods for differential expression analysis. For this, we use the function <em>rename()</em> provided by the library dplyr.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, we transform the results table to a data frame so that we see the results table directly when accessing it through the name "res"</span></span>
<span><span class="va">DE_results_limma_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">DE_results_limma_Entrez</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rename column that contains adjusted p-values: rename padj to p_adj</span></span>
<span><span class="va">DE_results_limma_Entrez</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="va">DE_results_limma_Entrez</span>, </span>
<span>                                         p_adj <span class="op">=</span> <span class="st">"adj.P.Val"</span><span class="op">)</span></span></code></pre></div>
<p><strong>6. Take a look at top 10 genes in the results table generated with limma</strong></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_limma_Entrez</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;             logFC   AveExpr         t      P.Value</span></span>
<span><span class="co">#&gt; 6192    9.2105431 1.9863048 47.186499 1.020114e-55</span></span>
<span><span class="co">#&gt; 9087    5.0806396 0.5128935 26.668835 6.058859e-39</span></span>
<span><span class="co">#&gt; 8228   -0.9240302 5.3334193 -8.779270 5.579187e-13</span></span>
<span><span class="co">#&gt; 8226   -0.8544468 2.1626976 -4.415343 3.497072e-05</span></span>
<span><span class="co">#&gt; 159013 -0.5508457 3.7163074 -4.251355 6.307821e-05</span></span>
<span><span class="co">#&gt; 55787  -0.4042722 5.1160616 -4.045191 1.302619e-04</span></span>
<span><span class="co">#&gt; 50486  -1.4703261 5.1922528 -3.855673 2.493894e-04</span></span>
<span><span class="co">#&gt; 3725   -0.5880074 8.2731003 -3.920361 2.001855e-04</span></span>
<span><span class="co">#&gt; 317     0.4046249 5.9681588  3.822268 2.791390e-04</span></span>
<span><span class="co">#&gt; 3552   -1.4855077 3.5435981 -3.666767 4.682471e-04</span></span>
<span><span class="co">#&gt;               p_adj           B</span></span>
<span><span class="co">#&gt; 6192   6.133944e-52 84.20170055</span></span>
<span><span class="co">#&gt; 9087   1.821596e-35 62.68255252</span></span>
<span><span class="co">#&gt; 8228   1.118255e-09 19.22260877</span></span>
<span><span class="co">#&gt; 8226   5.256973e-02  2.15126317</span></span>
<span><span class="co">#&gt; 159013 7.585786e-02  1.55871370</span></span>
<span><span class="co">#&gt; 55787  1.305441e-01  0.74145045</span></span>
<span><span class="co">#&gt; 50486  1.864959e-01  0.15615850</span></span>
<span><span class="co">#&gt; 3725   1.719594e-01  0.11995311</span></span>
<span><span class="co">#&gt; 317    1.864959e-01 -0.07537785</span></span>
<span><span class="co">#&gt; 3552   2.616375e-01 -0.24387530</span></span></code></pre></div>
</div>
<div id="option-2-differential-expression-analysis-using-deseq2" class="section level3" number="3.2.2">
<h3>
<span class="header-section-number">3.2.2</span> option 2: Differential expression analysis using DESeq2<a class="anchor" aria-label="anchor" href="#option-2-differential-expression-analysis-using-deseq2"><i class="fas fa-link"></i></a>
</h3>
<p><strong>1. Generate required input object</strong></p>
<p>DESeq2 operates on the format DESeqDataSet which contains information on the count data, the conditions of the samples and the design. Note that for gene expression data sets that must be imported to R, additional steps are necessary before the following code can be run.</p>
<p><strong>Generate a data frame which contains the condition of each sample</strong></p>
<p>Here, the information on the sample conditions is the only column in the data frame. However, if further variables (such as batch effects) are to be controlled for, the corresponding variables must additionally be added to coldata.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the names of the samples are stored as the row names of the data frame</span></span>
<span><span class="co"># -&gt; important: make sure that the order of the conditions in sample_conditions corresponds to the order of the samples in expression_data</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">sample_conditions</span>, </span>
<span>                      row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">exprdat_filter_conv_DESeq2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rename the column header to "condition" </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"condition"</span></span>
<span></span>
<span><span class="co"># recode the variable condition as a factor </span></span>
<span><span class="co"># rename the sample conditions (in this case from "female" and "male") to "untreated" and "treated"</span></span>
<span><span class="co"># note: make sure that the control level in variable condition is coded as the first level (i.e. "untreated")</span></span>
<span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span>, </span>
<span>                            labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"untreated"</span>,<span class="st">"treated"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Generate the DESeqDataSet</strong></p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">exprdat_filter_conv_DESeq2</span>, </span>
<span>                                     colData <span class="op">=</span> <span class="va">coldata</span>, </span>
<span>                                     design <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span><span class="op">)</span></span></code></pre></div>
<p>Relevant arguments in function DESeqDataSetFromMatrix:</p>
<ul>
<li><p><strong>countData</strong>: count data from the gene expression data set</p></li>
<li><p><strong>colData</strong>: data frame that contains information on the samples (see above) such as the conditions of the samples (required) and possibly further variables to correct for (such as batch effects)</p></li>
<li>
<p><strong>design</strong>: indicates which variables from colData are used for modelling</p>
<ul>
<li><p>more detailed: the argument design is used to estimate the dispersions and the log2 fold changes of the model</p></li>
<li><p>if more than one variable from colData are used in argument design (e.g. a second variable “batch”), the syntax changes to the following formula: ‘design ~ batch + condition’</p></li>
<li><p>make sure that the variable of interest (here: variable that represents conditions of the samples) is placed at the end of the formula</p></li>
</ul>
</li>
</ul>
<p><strong>2. Test for differential expression</strong></p>
<p>DESeq2 calculates the normalization factor of each sample, estimates the dispersion parameter, fits a negative binomial model. A p-value of differential expression for each gene is then computed using a Wald test.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># perform default differential expression analysis </span></span>
<span><span class="va">dds_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span><span class="op">(</span><span class="va">dds_Entrez</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate results table which provides</span></span>
<span><span class="va">DE_results_DESeq2_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span><span class="op">(</span><span class="va">dds_Entrez</span><span class="op">)</span></span></code></pre></div>
<p>Note: function <em>results()</em> provides the base mean across all samples, the estimated log2 fold changes, standard errors, test statistics, p-values, and adjusted p-values for each gene.</p>
<p><strong>3. Rename columns in results of differential expression analysis</strong></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, we transform to results table to a data frame so that we see the results</span></span>
<span><span class="co"># table directly when accessing it through the name "res"</span></span>
<span><span class="va">DE_results_DESeq2_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_Entrez</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rename column that contains adjusted p-values: rename padj to p_adj</span></span>
<span><span class="va">DE_results_DESeq2_Entrez</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_Entrez</span>, </span>
<span>                                          p_adj <span class="op">=</span> <span class="va">padj</span><span class="op">)</span></span></code></pre></div>
<p><strong>4. Take a look at top 10 genes in the results table generated with DESeq2</strong></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_Entrez</span>, n <span class="op">=</span> <span class="fl">10</span> <span class="op">)</span></span>
<span><span class="co">#&gt;         baseMean log2FoldChange      lfcSE        stat</span></span>
<span><span class="co">#&gt; 8813   62.726871    0.039434784 0.11345444  0.34758255</span></span>
<span><span class="co">#&gt; 57147  70.453343    0.028533065 0.11131002  0.25633868</span></span>
<span><span class="co">#&gt; 55732  13.001060    0.164117365 0.15102073  1.08672080</span></span>
<span><span class="co">#&gt; 2268   43.577756    0.113711805 0.19545410  0.58178266</span></span>
<span><span class="co">#&gt; 2519   37.963998   -0.006527064 0.10803176 -0.06041801</span></span>
<span><span class="co">#&gt; 4800  228.478329    0.051147440 0.09641906  0.53047024</span></span>
<span><span class="co">#&gt; 81887  11.534031    0.025191913 0.18501229  0.13616346</span></span>
<span><span class="co">#&gt; 22875 175.833526    0.299999462 0.16347952  1.83508896</span></span>
<span><span class="co">#&gt; 5893    4.688308    0.280425754 0.19291917  1.45359198</span></span>
<span><span class="co">#&gt; 572    66.369975   -0.264896424 0.10233495 -2.58852362</span></span>
<span><span class="co">#&gt;            pvalue     p_adj</span></span>
<span><span class="co">#&gt; 8813  0.728153713 0.9313348</span></span>
<span><span class="co">#&gt; 57147 0.797689329 0.9506228</span></span>
<span><span class="co">#&gt; 55732 0.277160217 0.7502690</span></span>
<span><span class="co">#&gt; 2268  0.560713083 0.8849136</span></span>
<span><span class="co">#&gt; 2519  0.951822712 0.9884098</span></span>
<span><span class="co">#&gt; 4800  0.595785935 0.8982377</span></span>
<span><span class="co">#&gt; 81887 0.891692065 0.9763145</span></span>
<span><span class="co">#&gt; 22875 0.066492509 0.5209390</span></span>
<span><span class="co">#&gt; 5893  0.146059463 0.6396912</span></span>
<span><span class="co">#&gt; 572   0.009638834 0.3141234</span></span></code></pre></div>
</div>
<div id="option-3-differential-expression-analysis-using-edger" class="section level3" number="3.2.3">
<h3>
<span class="header-section-number">3.2.3</span> option 3: Differential expression analysis using edgeR<a class="anchor" aria-label="anchor" href="#option-3-differential-expression-analysis-using-edger"><i class="fas fa-link"></i></a>
</h3>
<p><strong>1. Generate required input object</strong></p>
<p>Note that edgeR operates on a DGEList object.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">exprdat_filter_conv_filterByExpr</span>, </span>
<span>             group <span class="op">=</span> <span class="va">sample_conditions</span><span class="op">)</span></span></code></pre></div>
<p>Required function arguments:</p>
<ul>
<li>
<strong>counts</strong>: matrix that contains RNA-Seq data (i.e. count data)</li>
</ul>
<p>Optional function arguments:</p>
<ul>
<li>
<strong>group</strong>: indicates condition of each sample/library
<ul>
<li>this argument must be specified sooner or later (such as in subsequent functions) so we just specify it at this point</li>
</ul>
</li>
</ul>
<p>Note that we leave the remaining arguments in their default state since the corresponding info will be added through the following pipeline of functions.</p>
<p><strong>2. Normalization</strong></p>
<p>The following piece of code generates a normalization factor for each sample. This accounts for sample-specific effect (such as differences in library sizes and effects of compositionality). If not accounted for, these effects prevent a comparison between the samples.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p>Note that this function does not transform the count data, but rather generates a normalization factor for each sample which is incorporated into the subsequent analysis separately.</p>
<p><strong>3. Estimation of dispersion</strong></p>
<p>The following code estimates common and tagwise dispersion, namely the variation of the true abundance of a given gene between different samples. This is required to assess differential expression realistically.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p><strong>4. Test for differential expression</strong></p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># test each gene for differential expression: </span></span>
<span><span class="va">DE_results_edgeR_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/exactTest.html">exactTest</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract pre-specified (n) number of genes</span></span>
<span><span class="va">DE_results_edgeR_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Entrez</span>, </span>
<span>                                   n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Entrez</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that argument ‘n’ specifies the number of top differentially expressed genes to be displayed in the results. Setting it to ‘n = nrow(DE_results_Entrez)’ ensures the results of all genes whose differential expression was assessed are displayed.</p>
<p><strong>5. Rename columns in the results of differential expression analysis</strong></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, we transform the results table to a data frame so that we see the results</span></span>
<span><span class="co"># table directly when accessing it through the name "res"</span></span>
<span><span class="va">DE_results_edgeR_Entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Entrez</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rename column that contains adjusted p-values: rename padj to p_adj</span></span>
<span><span class="va">DE_results_edgeR_Entrez</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Entrez</span>, </span>
<span>                                         p_adj <span class="op">=</span> <span class="va">FDR</span><span class="op">)</span></span></code></pre></div>
<p><strong>4. Take a look at top 10 genes in the results table generated with edgeR</strong></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_edgeR_Entrez</span>, n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt;                logFC   logCPM        PValue         p_adj</span></span>
<span><span class="co">#&gt; 6192      10.1330599 6.154999  0.000000e+00  0.000000e+00</span></span>
<span><span class="co">#&gt; 9087       5.5791707 2.519656 6.188599e-121 1.860602e-117</span></span>
<span><span class="co">#&gt; 8228      -0.9448255 5.485319  1.821705e-17  3.651303e-14</span></span>
<span><span class="co">#&gt; 8226      -0.7768228 2.571002  3.067361e-06  4.611011e-03</span></span>
<span><span class="co">#&gt; 6615      -1.6460041 3.288135  4.698096e-06  5.627893e-03</span></span>
<span><span class="co">#&gt; 10912     -1.5870378 2.679647  5.615726e-06  5.627893e-03</span></span>
<span><span class="co">#&gt; 284486    -1.1257052 2.302952  1.949732e-05  1.674820e-02</span></span>
<span><span class="co">#&gt; 1521       1.5019085 3.365855  2.906843e-05  2.184856e-02</span></span>
<span><span class="co">#&gt; 55787     -0.4253221 5.207647  3.522694e-05  2.269295e-02</span></span>
<span><span class="co">#&gt; 10732     -0.6028374 5.756558  3.773981e-05  2.269295e-02</span></span>
<span><span class="co">#&gt; 100191040 -1.1506301 3.121643  4.727503e-05  2.584225e-02</span></span>
<span><span class="co">#&gt; 412       -0.7312091 6.861545  1.064460e-04  5.333830e-02</span></span>
<span><span class="co">#&gt; 159013    -0.5219775 3.911283  1.225919e-04  5.670349e-02</span></span>
<span><span class="co">#&gt; 317        0.4048454 6.062183  2.164067e-04  9.294666e-02</span></span>
<span><span class="co">#&gt; 51655     -1.3164084 3.094734  2.570863e-04  1.003474e-01</span></span>
<span><span class="co">#&gt; 6662      -0.9774711 8.100796  2.804477e-04  1.003474e-01</span></span>
<span><span class="co">#&gt; 9404       0.3361246 9.579044  2.837030e-04  1.003474e-01</span></span>
<span><span class="co">#&gt; 171177    -0.6747490 6.350249  3.161790e-04  1.031954e-01</span></span>
<span><span class="co">#&gt; 10479      0.3635303 6.142157  3.320976e-04  1.031954e-01</span></span>
<span><span class="co">#&gt; 3725      -0.5320863 8.429198  3.593785e-04  1.031954e-01</span></span></code></pre></div>

</div>
</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></div>
<div class="next"><a href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#differential-expression-analysis"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li>
<a class="nav-link" href="#libraries-2"><span class="header-section-number">3.1</span> Libraries</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#load-data-1"><span class="header-section-number">3.1.1</span> Load data</a></li></ul>
</li>
<li>
<a class="nav-link" href="#differential-expression-analysis-with-entrez-gene-id"><span class="header-section-number">3.2</span> Differential Expression Analysis (with Entrez gene ID)</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-differential-expression-analysis-using-limma"><span class="header-section-number">3.2.1</span> Option 1: Differential expression analysis using limma</a></li>
<li><a class="nav-link" href="#option-2-differential-expression-analysis-using-deseq2"><span class="header-section-number">3.2.2</span> option 2: Differential expression analysis using DESeq2</a></li>
<li><a class="nav-link" href="#option-3-differential-expression-analysis-using-edger"><span class="header-section-number">3.2.3</span> option 3: Differential expression analysis using edgeR</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
