<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 12 GSEAPreranked | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will go through the process of preparing the required input for GSEAPreranked20 (which is part of the web-based tool GSEA21 ). Note that in contrast to the remaining tools...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 12 GSEAPreranked | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will go through the process of preparing the required input for GSEAPreranked20 (which is part of the web-based tool GSEA21 ). Note that in contrast to the remaining tools...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 12 GSEAPreranked | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will go through the process of preparing the required input for GSEAPreranked20 (which is part of the web-based tool GSEA21 ). Note that in contrast to the remaining tools...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="active" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="gseapreranked" class="section level1" number="12">
<h1>
<span class="header-section-number">12</span> GSEAPreranked<a class="anchor" aria-label="anchor" href="#gseapreranked"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will go through the process of preparing the required input for GSEAPreranked<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;Aravind Subramanian et al., &lt;span&gt;“Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles,”&lt;/span&gt; &lt;em&gt;Proceedings of the National Academy of Sciences&lt;/em&gt; 102, no. 43 (2005): 15545–50; Vamsi K Mootha et al., &lt;span&gt;“&lt;span&gt;PGC&lt;/span&gt;-1&lt;span class="math inline"&gt;\(\alpha\)&lt;/span&gt;-Responsive Genes Involved in Oxidative Phosphorylation Are Coordinately Downregulated in Human Diabetes,”&lt;/span&gt; &lt;em&gt;Nature Genetics&lt;/em&gt; 34, no. 3 (2003): 267–73.&lt;/p&gt;'><sup>20</sup></a></span> (which is part of the web-based tool GSEA<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;Subramanian et al., &lt;span&gt;“Gene Set Enrichment Analysis”&lt;/span&gt;; Mootha et al., &lt;span&gt;“&lt;span&gt;PGC&lt;/span&gt;-1&lt;span class="math inline"&gt;\(\alpha\)&lt;/span&gt;-Responsive Genes Involved in Oxidative Phosphorylation Are Coordinately Downregulated in Human Diabetes.”&lt;/span&gt;&lt;/p&gt;'><sup>21</sup></a></span> ).<br>
Note that in contrast to the remaining tools considered in this work, it is recom-
mended for GSEAPreranked to provide the input with the genes ID as HGNC (HUGO) gene
symbols.<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Ruth L Seal et al., &lt;span&gt;“Genenames. Org: The HGNC Resources in 2023,”&lt;/span&gt; &lt;em&gt;Nucleic Acids Research&lt;/em&gt; 51, no. D1 (2023): D1003–9.&lt;/p&gt;"><sup>22</sup></a></span> The reason for this is described in the supplement of the paper.</p>
<p>Therefore, our starting point is the pre-filtered gene expression data set with the
gene IDs in the initial (Ensembl) ID format.</p>
<p>We proceed in the following order:</p>
<ol style="list-style-type: decimal">
<li><p>Conversion of the gene IDs to HGNC (HUGO) symbols and removal of resulting duplicated gene IDs</p></li>
<li><p>Differential expression analysis using voom/limma, DESeq2, and edgeR</p></li>
<li><p>Generation of the gene ranking from each results table generated in (ii)</p></li>
<li><p>Export of gene rankings to text file</p></li>
<li><p>Link to information on how to further process the gene rankings in Excel</p></li>
</ol>
<div id="libraries-11" class="section level2" number="12.1">
<h2>
<span class="header-section-number">12.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-11"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb178"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"tweeDEseqCountData"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"edgeR"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load libraries</strong>:</p>
<div class="sourceCode" id="cb179"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.creal.cat/jrgonzalez/software.htm">tweeDEseqCountData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<p>Description of the libraries:</p>
<ul>
<li><p><strong>clusterProfiler</strong>: To convert Ensembl ID format to HGNC (HUGO) gene symbols</p></li>
<li><p><strong>org.Hs.eg.db</strong>: To indicate that we work with human gene expression data (must be adapted when working with a different organism)</p></li>
<li><p><strong>tweeDEseqCountData</strong>: To obtain the conditions of the samples in the gene expression data set</p></li>
<li><p><strong>limma</strong>: For differential expression analysis (based on which the gene ranking is generated)</p></li>
<li><p><strong>edgeR</strong>: For differential expression analysis (based on which the gene ranking is generated)</p></li>
<li><p><strong>DESeq2</strong>: For differential expression analysis (based on which the gene ranking is generated)</p></li>
<li><p><strong>dplyr</strong>: To unify the relevant columns in the results of differential expression analysis across the three methods</p></li>
</ul>
</div>
<div id="load-data-10" class="section level2" number="12.2">
<h2>
<span class="header-section-number">12.2</span> Load data<a class="anchor" aria-label="anchor" href="#load-data-10"><i class="fas fa-link"></i></a>
</h2>
<p>We will proceed with the gene expression data set which was pre-filtered using edgeR’s <em>filterByExpr()</em>.</p>
<div class="sourceCode" id="cb180"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_PreFiltering/expression_data_filterByExpr.Rdata"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can also load the gene expression data set which was manually pre-filtered (as proposed by DESeq2):</p>
<div class="sourceCode" id="cb181"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load("./data/Results_PreFiltering/expression_data_filterDESeq2.Rdata")</span></span></code></pre></div>
<p>For a simplified readability, we will proceed with the pre-filtered gene expression data set under a different name:</p>
<div class="sourceCode" id="cb182"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_filtered</span> <span class="op">&lt;-</span> <span class="va">expression_data_filterByExpr</span></span></code></pre></div>
</div>
<div id="step-1-convert-ensembl-ids-to-hgnc-symbols" class="section level2" number="12.3">
<h2>
<span class="header-section-number">12.3</span> Step 1: Convert Ensembl IDs to HGNC symbols<a class="anchor" aria-label="anchor" href="#step-1-convert-ensembl-ids-to-hgnc-symbols"><i class="fas fa-link"></i></a>
</h2>
<p><strong>(i) Obtain the mapping of initial (Ensembl) gene IDs to required HGNC gene symbols</strong></p>
<div class="sourceCode" id="cb183"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bitr_EnsToSymb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/bitr.html">bitr</a></span><span class="op">(</span>geneID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">expression_data_filtered</span><span class="op">)</span>,</span>
<span>                       fromType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>                       toType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                       OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'select()' returned 1:many mapping between keys and</span></span>
<span><span class="co">#&gt; columns</span></span>
<span><span class="co">#&gt; Warning in bitr(geneID =</span></span>
<span><span class="co">#&gt; rownames(expression_data_filtered), fromType = "ENSEMBL", :</span></span>
<span><span class="co">#&gt; 3.84% of input gene IDs are fail to map...</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li><p><strong>geneID</strong>: gene IDs to be converted (stored in the rownames of the gene expression data set)</p></li>
<li><p><strong>fromType</strong>: initial gene ID format that is to be converted</p></li>
<li><p><strong>toType</strong>: gene ID format to be converted to</p></li>
<li><p><strong>OrgDb</strong>: annotation data base of organisms from which the gene expression measurements originate. The corresponding argument must be loaded as a library (see above)</p></li>
</ul>
<p>Note that the arguments “fromType” and “toType” must be set as one of the following, depending on the given and the required gene ID format:<br></p>
<p>ACCNUM, ALIAS, Ensembl, EnsemblPROT, EnsemblTRANS, ENTREZID, ENZYME, EVIDENCE, EVIDENCEALL, GENENAME, GENETYPE, GO, GOALL, IPI, MAP, OMIM, ONTOLOGY, ONTOLOGYALL, PATH, PFAM, PMID, PROSITE, REFSEQ, SYMBOL, UCSCKG, UNIPROT.</p>
<p><strong>(ii) Concatenate the initial gene expression data set with the mapping from initial (Ensembl) to required HGNC symbols</strong></p>
<p>Merge by row names of expression data set and ENSEMBL ID of gene ID mapping.</p>
<div class="sourceCode" id="cb184"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression_data_filtered</span>,</span>
<span>                                             y <span class="op">=</span> <span class="va">bitr_EnsToSymb</span>,</span>
<span>                                             by.x <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                             by.y <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>                                             sort<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Description of function arguments:</p>
<ul>
<li>
<p><strong>by.x</strong> and <strong>by.y</strong>: specify by which columns in expression_data_filterByExpr and bitr_EnsToEntr, respectively, both data sets are concatenated:</p>
<ul>
<li><p><strong>by.x = 0</strong>: use row names of expression_data_filterByExpr (the row names contain the gene IDs)</p></li>
<li><p><strong>by.y = “ENSEMBL”</strong>: use the column that contains the Ensembl gene IDs</p></li>
</ul>
</li>
</ul>
<p>Analogous Chapter ‘Gene ID conversion and removal of the resulting duplicated gene IDs’, we work with three separate ways to remove the duplicate gene IDs that result fromvgene ID conversion. We will apply all three of these approaches to removal.</p>
</div>
<div id="step-2-take-a-closer-look-at-the-duplicated-gene-ids-1" class="section level2" number="12.4">
<h2>
<span class="header-section-number">12.4</span> Step 2: Take a closer look at the duplicated gene IDs<a class="anchor" aria-label="anchor" href="#step-2-take-a-closer-look-at-the-duplicated-gene-ids-1"><i class="fas fa-link"></i></a>
</h2>
<p><strong>Case 1: single ENSEMBL ID is mapped to multiple HGNC symbols</strong></p>
<p>In this case, we denote the corresponding ENSEMBL IDs as the duplicated IDs</p>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtain number of cases in which an ENSEMBL gene ID was converted to several HGNC symbols, i.e. the number of</span></span>
<span><span class="co"># times an Ensembl ID appears more than once in the mapping</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">ENSEMBL</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 34</span></span>
<span><span class="co"># determine all duplicated ENSEMBL gene IDS (i.e. all ENSEMBL gene IDs that were mapped to multiple distinct HGNC gene symbols):</span></span>
<span><span class="va">dupl_ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">ENSEMBL</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">ENSEMBL</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># number of ENSEMBL IDs that have at least one duplicate</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dupl_ensembl</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 34</span></span></code></pre></div>
<p>In the following, we can inspect the conversion pattern for each Ensembl ID that was mapped to multiple distinct HGNC symbols:</p>
<div class="sourceCode" id="cb186"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get subset conversion pattern of all duplicated Ensembl IDs </span></span>
<span><span class="va">duplicated_conversion_ens</span><span class="op">&lt;-</span><span class="va">bitr_EnsToSymb</span><span class="op">[</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">ENSEMBL</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="va">dupl_ensembl</span>,<span class="op">]</span></span>
<span><span class="co"># take a look at conversion pattern for some of the duplicated Ensembl gene IDs:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">duplicated_conversion_ens</span>, n <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt;             ENSEMBL          SYMBOL</span></span>
<span><span class="co">#&gt; 303 ENSG00000063587          ZNF275</span></span>
<span><span class="co">#&gt; 304 ENSG00000063587    LOC105373378</span></span>
<span><span class="co">#&gt; 615 ENSG00000088298           EDEM2</span></span>
<span><span class="co">#&gt; 616 ENSG00000088298 MMP24-AS1-EDEM2</span></span>
<span><span class="co">#&gt; 683 ENSG00000090857            PDPR</span></span>
<span><span class="co">#&gt; 684 ENSG00000090857    LOC124907803</span></span></code></pre></div>
<p>We can see perfectly that these Ensembl ID are each mapped to multiple individual HGNC symbols.</p>
<p><strong>Case 2: multiple ENSEMBL IDs are mapped to the same HGNC symbol</strong></p>
<p>In this case, we denote the corresponding HGNC symbols as the duplicated IDs.</p>
<p>Obtain number of cases in which multiple distinct Ensembl IDs are converted to the same HGNC gene symbol</p>
<div class="sourceCode" id="cb187"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># in these cases, the corresponding HGNC symbols appear repeatedly in the mapping:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3</span></span>
<span></span>
<span><span class="co"># determine all HGNC gene symbols affected by this duplication</span></span>
<span><span class="va">dupl_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect the first few HGNC symbols affected by a duplication</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">dupl_symbol</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "H4C15" "OPN3"</span></span>
<span></span>
<span><span class="co"># get number of HGNC symbols that have at least one duplicate</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span></code></pre></div>
<p>In the following, we want to take a look at the conversion pattern of some of the duplicated HGNC symbols.</p>
<div class="sourceCode" id="cb188"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset the conversion pattern to that of the duplicated HGNC symbols</span></span>
<span><span class="va">duplicated_conversion</span> <span class="op">&lt;-</span> <span class="va">bitr_EnsToSymb</span><span class="op">[</span><span class="va">bitr_EnsToSymb</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="va">dupl_symbol</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># for visibility: order the conversion pattern by the HGNC symbols</span></span>
<span><span class="va">duplicated_conversion</span> <span class="op">&lt;-</span> <span class="va">duplicated_conversion</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html">order</a></span><span class="op">(</span><span class="va">duplicated_conversion</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># take a look at conversion pattern of the first few duplicated HGNC symbols:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">duplicated_conversion</span>, n <span class="op">=</span> <span class="fl">6</span> <span class="op">)</span></span>
<span><span class="co">#&gt;              ENSEMBL SYMBOL</span></span>
<span><span class="co">#&gt; 3605 ENSG00000158406  H4C15</span></span>
<span><span class="co">#&gt; 5675 ENSG00000197061  H4C15</span></span>
<span><span class="co">#&gt; 5736 ENSG00000197837  H4C15</span></span>
<span><span class="co">#&gt; 250  ENSG00000054277   OPN3</span></span>
<span><span class="co">#&gt; 5875 ENSG00000203668   OPN3</span></span></code></pre></div>
</div>
<div id="step-3-remove-the-duplicated-gene-ids" class="section level2" number="12.5">
<h2>
<span class="header-section-number">12.5</span> Step 3: Remove the duplicated gene IDs<a class="anchor" aria-label="anchor" href="#step-3-remove-the-duplicated-gene-ids"><i class="fas fa-link"></i></a>
</h2>
<p>For the subsequent analysis (whether differential expression analysis of gene set analysis), we need to deal with the duplicated gene IDs and remove them in a suitable way such that only one unique gene expression measurement among the duplicates remains. There is no recommended way to proceed, i.e. no common approach presented in official scientific puplicatios, but instead several approaches suggested by users in corresponding user platforms. We have observed that the manner of duplicate gene ID removal seems to be chosen at the discretion of the user. We therefore present three approaches to duplicate gene ID removal.</p>
<div id="option-1-keep-the-duplicate-with-the-lowest-subscript" class="section level3" number="12.5.1">
<h3>
<span class="header-section-number">12.5.1</span> Option 1: Keep the duplicate with the lowest subscript<a class="anchor" aria-label="anchor" href="#option-1-keep-the-duplicate-with-the-lowest-subscript"><i class="fas fa-link"></i></a>
</h3>
<p>This is the simplest approach among the three.</p>
<p><strong>1. Remove duplicated HGNC gene symbols</strong></p>
<div class="sourceCode" id="cb189"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># indicate which HGNC gene symbols are NO duplicate of any other gene ID</span></span>
<span><span class="va">ind_nodupl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">expression_data_merged</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter the gene expression data set to the genes that are NO duplicate of any other gene ID </span></span>
<span><span class="va">exprdat_symbol_dupl1</span> <span class="op">&lt;-</span> <span class="va">expression_data_merged</span><span class="op">[</span><span class="va">ind_nodupl</span>,<span class="op">]</span></span></code></pre></div>
<p><strong>2. Remove duplicated ENSEMBL gene IDs</strong></p>
<div class="sourceCode" id="cb190"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># indicate which Ensembl IDs are NO duplicate of any other gene ID </span></span>
<span><span class="va">ind_nodupl</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl1</span><span class="op">$</span><span class="va">Row.names</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter the gene expression data set to the genes that are NO duplicate of any other gene ID </span></span>
<span><span class="va">exprdat_symbol_dupl1</span> <span class="op">&lt;-</span> <span class="va">exprdat_symbol_dupl1</span><span class="op">[</span><span class="va">ind_nodupl</span>,<span class="op">]</span></span></code></pre></div>
<p><strong>3. Set HGNC gene symbols as rownames</strong></p>
<div class="sourceCode" id="cb191"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">exprdat_symbol_dupl1</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span></span>
<span><span class="co">#Remove columns containing ENSEMBL and HGNC symbols</span></span>
<span><span class="va">exprdat_symbol_dupl1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl1</span>, select<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Row.names</span>,<span class="va">SYMBOL</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the dimension of the resulting gene expression data set</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6006   69</span></span>
<span><span class="co"># we particularly see that the number of columns is now at its initial number</span></span></code></pre></div>
<p>We now want to inspect a small part of the gene expression data set:</p>
<div class="sourceCode" id="cb192"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exprdat_symbol_dupl1</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;        NA18486 NA18498 NA18499 NA18501 NA18502 NA18504</span></span>
<span><span class="co">#&gt; DPM1        22     105      40      55      67      37</span></span>
<span><span class="co">#&gt; SCYL3       22     100     107      53      72      38</span></span>
<span><span class="co">#&gt; FIRRM        5      23      10      18      15       8</span></span>
<span><span class="co">#&gt; FGR         36      70      41      33      59      29</span></span>
<span><span class="co">#&gt; FUCA2       29      79      33      31      29      21</span></span>
<span><span class="co">#&gt; NFYA       301     351     344     176     340     170</span></span>
<span><span class="co">#&gt; LAS1L       32      22       4       8      15       5</span></span>
<span><span class="co">#&gt; ENPP4       52     419     173     137     127      72</span></span>
<span><span class="co">#&gt; BAD         48      81      65      19      59      57</span></span>
<span><span class="co">#&gt; HS3ST1       8      15       8      18       4      27</span></span>
<span><span class="co">#&gt;        NA18505 NA18507 NA18508 NA18510</span></span>
<span><span class="co">#&gt; DPM1        88     127      70      43</span></span>
<span><span class="co">#&gt; SCYL3       98      69      66      43</span></span>
<span><span class="co">#&gt; FIRRM       11      16      18       7</span></span>
<span><span class="co">#&gt; FGR         22      71      12      43</span></span>
<span><span class="co">#&gt; FUCA2       42      62      41      25</span></span>
<span><span class="co">#&gt; NFYA       238     247     226     227</span></span>
<span><span class="co">#&gt; LAS1L       15      16      19      15</span></span>
<span><span class="co">#&gt; ENPP4      267     247     154      81</span></span>
<span><span class="co">#&gt; BAD         44      79      71      79</span></span>
<span><span class="co">#&gt; HS3ST1       3      11       5       3</span></span></code></pre></div>
</div>
<div id="option-2-keep-the-rounded-mean-expression-value-of-all-duplicated-gene-ids-1" class="section level3" number="12.5.2">
<h3>
<span class="header-section-number">12.5.2</span> Option 2: Keep the (rounded) mean expression value of all duplicated gene IDs<a class="anchor" aria-label="anchor" href="#option-2-keep-the-rounded-mean-expression-value-of-all-duplicated-gene-ids-1"><i class="fas fa-link"></i></a>
</h3>
<p>Here, we remove duplicated HGNC gene symbols (case 2) before removing the duplicated Ensembl gene IDs (case 1). The reason for this is elaborated below.</p>
<p><strong>1. Remove the duplicated HGNC gene symbols</strong></p>
<div class="sourceCode" id="cb193"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#generate matrix to contain (rounded) mean expression values of all rows that have same HGNC gene symbol</span></span>
<span><span class="co">#ncol=ncol(expression_data_filterByExpr)-2 since data set contains 2 columns with IDs at this point</span></span>
<span><span class="va">mean_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span>nrow<span class="op">=</span><span class="fl">0</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">expression_data_merged</span><span class="op">)</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># for each duplicated HGNC gene symbol separately, we gather all rows with the corresponding gene expression data and then extract the (rounded) mean expression value of all rows</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span></span>
<span><span class="co"># go through each HGNC symbols which occurs multiple times</span></span>
<span><span class="co"># determine all rows whose HGNC symbols correspond to currently considered HGNC symbol</span></span>
<span><span class="va">counts_dupl</span> <span class="op">&lt;-</span> <span class="va">expression_data_merged</span><span class="op">[</span><span class="va">expression_data_merged</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co">#compute the mean expression value of all rows that contain to the given HGNC gene symbol</span></span>
<span><span class="va">dupl_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="va">counts_dupl</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">expression_data_merged</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># store rounded mean expression value in matrix</span></span>
<span><span class="co"># this matrix is extended by a single row of gene expression data which corresponds to the (rounded) mean expression data that corresponds to the given HGNC gene symbol</span></span>
<span><span class="va">mean_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">rbind</a></span><span class="op">(</span><span class="va">mean_symbol</span>,<span class="va">dupl_id</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#set corresponding HGNC gene symbols as rownames</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">mean_symbol</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># after completing the for-loop, mean_symbol contains the mean expression measurements of each HGNC gene symbol which contains duplicates resulting from gene ID conversion</span></span>
<span><span class="co"># We want to take a look at a part of the data set </span></span>
<span><span class="va">mean_symbol</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span></span>
<span><span class="co">#&gt;       NA18486 NA18498 NA18499 NA18501 NA18502 NA18504</span></span>
<span><span class="co">#&gt; H4C15      27     109      31      50      57      18</span></span>
<span><span class="co">#&gt; OPN3       50     182      82     118      60      52</span></span>
<span><span class="co">#&gt;       NA18505 NA18507 NA18508</span></span>
<span><span class="co">#&gt; H4C15      56      34      52</span></span>
<span><span class="co">#&gt; OPN3      102      80      78</span></span>
<span></span>
<span></span>
<span><span class="co"># test whether the number of rows in mean_symbol corresponds to the number HGNC symbols</span></span>
<span><span class="co"># that occur more than once</span></span>
<span><span class="co"># result should be TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">mean_symbol</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co"># remove all rows from the expression data whose HGNC symbol has at least one duplicate</span></span>
<span><span class="co"># intuition: we have just dealt with the corresponding rows and do not want them to be considered in the second step (which deals with case 2</span></span>
<span><span class="va">exprdat_symbol_dupl2</span> <span class="op">&lt;-</span> <span class="va">expression_data_merged</span><span class="op">[</span><span class="op">!</span><span class="va">expression_data_merged</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="va">dupl_symbol</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># test whether the number of rows in resulting data set equals nrow of inital data set minus number of genes with at least one duplicate</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl2</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression_data_merged</span><span class="op">)</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">duplicated_conversion</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p><strong>2. Remove the duplicated ENSEMBL IDs</strong></p>
<p>Caution: Single ENSEMBL IDs that are mapped to multiple HGNC symbol naturally generate identical count data for all corresponding HGNC symbols. It is therefore pointless to compute mean expression values. This is verifiable by looking at data set only containing those ENSEMBL IDs that are mapped by multiple HGNC symbols:</p>
<div class="sourceCode" id="cb194"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_dupl_ensembl</span> <span class="op">&lt;-</span> <span class="va">exprdat_symbol_dupl2</span><span class="op">[</span><span class="va">exprdat_symbol_dupl2</span><span class="op">$</span><span class="va">Row.names</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="va">dupl_ensembl</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># take a look at a few entries of the data set:</span></span>
<span><span class="co"># note that we additionally include the last column of the data set which corresponds to the HGNC symbols</span></span>
<span><span class="va">test_dupl_ensembl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">test_dupl_ensembl</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt;           Row.names NA18486 NA18498 NA18499 NA18501</span></span>
<span><span class="co">#&gt; 299 ENSG00000063587     143     116     244      76</span></span>
<span><span class="co">#&gt; 300 ENSG00000063587     143     116     244      76</span></span>
<span><span class="co">#&gt; 610 ENSG00000088298     202     384     368     207</span></span>
<span><span class="co">#&gt; 611 ENSG00000088298     202     384     368     207</span></span>
<span><span class="co">#&gt; 678 ENSG00000090857      59      41     102      96</span></span>
<span><span class="co">#&gt; 679 ENSG00000090857      59      41     102      96</span></span>
<span><span class="co">#&gt;              SYMBOL</span></span>
<span><span class="co">#&gt; 299          ZNF275</span></span>
<span><span class="co">#&gt; 300    LOC105373378</span></span>
<span><span class="co">#&gt; 610           EDEM2</span></span>
<span><span class="co">#&gt; 611 MMP24-AS1-EDEM2</span></span>
<span><span class="co">#&gt; 678            PDPR</span></span>
<span><span class="co">#&gt; 679    LOC124907803</span></span></code></pre></div>
<p>We see that those rows that originate from the same Ensembl ID contain identical count data.<br></p>
<p>We therefore proceed as in option 1 and use the HGNC symbol that occurs first while we remove the rest.</p>
<div class="sourceCode" id="cb195"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Keep HGNC symbol that occurs first</span></span>
<span><span class="va">exprdat_symbol_dupl2</span><span class="op">&lt;-</span><span class="va">exprdat_symbol_dupl2</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl2</span><span class="op">$</span><span class="va">Row.names</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<p><strong>3. Set HGNC symbol as rownames</strong></p>
<div class="sourceCode" id="cb196"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl2</span><span class="op">)</span><span class="op">&lt;-</span><span class="va">exprdat_symbol_dupl2</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span></span>
<span><span class="co"># Remove the columns containing ENSEMBL and HGNC symbols</span></span>
<span><span class="va">exprdat_symbol_dupl2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl2</span>,select<span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Row.names</span>,<span class="va">SYMBOL</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add those rows to the data set that contain mean expression values of duplicate HGNC symbols</span></span>
<span><span class="va">exprdat_symbol_dupl2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">rbind</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl2</span>,<span class="va">mean_symbol</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#Inspect the dimension of the remaining gene expression data set</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6007   69</span></span></code></pre></div>
</div>
<div id="option-3-among-the-duplicates-keep-row-with-highest-overall-expression-values-i.e-highest-counts-across-all-samples" class="section level3" number="12.5.3">
<h3>
<span class="header-section-number">12.5.3</span> Option 3: Among the duplicates, keep row with highest overall expression values (i.e highest counts across all samples)<a class="anchor" aria-label="anchor" href="#option-3-among-the-duplicates-keep-row-with-highest-overall-expression-values-i.e-highest-counts-across-all-samples"><i class="fas fa-link"></i></a>
</h3>
<p>The intuition behind this approach is that the row with highest counts values has highest power of being detected as differentially expressed. As in option 2, this applies only to the duplicates that result from multiple ENSEMBL IDs being mapped to the same HGNC symbol.</p>
<p><strong>1. Remove duplicated HGNC symbols</strong></p>
<div class="sourceCode" id="cb197"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate a matrix to later contain row with highest count values among ID duplicates this data set is to be filled gradually and with each iteration of the follwing for-loop</span></span>
<span><span class="va">highest_count_symbol</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span>, nrow<span class="op">=</span><span class="fl">0</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">expression_data_filterByExpr</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># For each duplicated HGNC gene symbol separately, we gather all rows with the corresponding gene expression data and then extract the row with the highest overall magnitude of counts</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span></span>
<span><span class="co"># Go through each HGNC symbols which occurs multiple times</span></span>
<span><span class="co"># determine all rows whose HGNC symbols correspond to currently considered HGNC symbol</span></span>
<span><span class="va">counts_dupl</span><span class="op">&lt;-</span><span class="va">expression_data_merged</span><span class="op">[</span><span class="va">expression_data_merged</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Order rows in decreasing manner by their number of read counts across all samples</span></span>
<span><span class="va">order_rowsums</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts_dupl</span><span class="op">[</span>,<span class="fl">2</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">counts_dupl</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># Detect row with highest number of read counts across all samples (i.e. row with rank 1)</span></span>
<span><span class="va">dupl_id</span><span class="op">&lt;-</span><span class="va">counts_dupl</span><span class="op">[</span><span class="va">order_rowsums</span><span class="op">==</span><span class="fl">1</span>,<span class="op">]</span></span>
<span>  </span>
<span><span class="co">#store corresponding expression</span></span>
<span><span class="va">highest_count_symbol</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">rbind</a></span><span class="op">(</span><span class="va">highest_count_symbol</span>,<span class="va">dupl_id</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co">#Remove all initial values with HGNC duplicates from the dataset initial gene expression data set</span></span>
<span><span class="va">exprdat_symbol_dupl3</span><span class="op">&lt;-</span><span class="va">expression_data_merged</span><span class="op">[</span><span class="op">!</span> <span class="va">expression_data_merged</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">dupl_symbol</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<p><strong>2. Remove duplicated ENSEMBL IDs</strong></p>
<p>As in option 2, it is pointless to detect the row with highest count values as all rows that correspond to the same ENSEMBL ID as these rows naturally contain identical count data. We therefore remove the duplicate ENSEMBL ID that occurs first (such as in option 1).</p>
<div class="sourceCode" id="cb198"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Keep the corresponding ENSEMBL ID that occurs first</span></span>
<span><span class="va">exprdat_symbol_dupl3</span><span class="op">&lt;-</span><span class="va">exprdat_symbol_dupl3</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl3</span><span class="op">$</span><span class="va">Row.names</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Add the gene expression rows of all HGNC gene symbols that were initially duplicated</span></span>
<span><span class="va">exprdat_symbol_dupl3</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">rbind</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl3</span>,<span class="va">highest_count_symbol</span> <span class="op">)</span></span></code></pre></div>
<p><strong>3. Set HGNC symbols as rownames</strong></p>
<div class="sourceCode" id="cb199"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl3</span><span class="op">)</span><span class="op">&lt;-</span><span class="va">exprdat_symbol_dupl3</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span><span class="co">#Remove any column that contains information on gene IDs</span></span>
<span><span class="va">exprdat_symbol_dupl3</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl3</span>, select<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Row.names</span>,<span class="va">SYMBOL</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the dimension of the resulting gene expression data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">exprdat_symbol_dupl3</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6007   69</span></span>
<span><span class="co"># we see that the sample size (number of columns) is now back at its initial number</span></span></code></pre></div>
</div>
</div>
<div id="intermediate-step-choose-which-converted-gene-expression-data-sets-to-proceed-with" class="section level2" number="12.6">
<h2>
<span class="header-section-number">12.6</span> Intermediate step: Choose which converted gene expression data sets to proceed with<a class="anchor" aria-label="anchor" href="#intermediate-step-choose-which-converted-gene-expression-data-sets-to-proceed-with"><i class="fas fa-link"></i></a>
</h2>
<p>In this illustration, we will proceed with only one of the three gene expression data sets that result from the conversion of the gene IDs and the removal of the duplications. Here, we will proceed with the first approach to
the removal of duplicated gene IDs. However, you can easily switch to another approach at your discretion.</p>
<p>We will proceed with the gene expression data set stored as object “expression_data_filt_symbol”.</p>
<div class="sourceCode" id="cb200"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_filt_symbol</span> <span class="op">&lt;-</span> <span class="va">exprdat_symbol_dupl1</span></span></code></pre></div>
</div>
<div id="step-3-differential-expression-analysis" class="section level2" number="12.7">
<h2>
<span class="header-section-number">12.7</span> Step 3: Differential expression analysis<a class="anchor" aria-label="anchor" href="#step-3-differential-expression-analysis"><i class="fas fa-link"></i></a>
</h2>
<p>From the corresponding results table of differential expression analysis, we will later transform certain metrics (i.e. columns) into a gene-level statistic for each gene based on which the gene ranking is generated.</p>
<p>We will proceed analogously to Chapter ‘Differential Expression Analysis’, in which we perform differential expression analysis using three established methods, namely</p>
<ol style="list-style-type: decimal">
<li>voom/limma</li>
<li>DESeq2</li>
<li>edgeR</li>
</ol>
<p>While we already have the count data at hand that we will be working with, we additionally need the conditions of the samples. Since we work with the Pickrell data set in these illustrations, we have to load the corresponding data.</p>
<div class="sourceCode" id="cb201"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load pickrell data </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pickrell</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Store the sample conditions in the object sample_conditions </span></span>
<span><span class="va">sample_conditions</span> <span class="op">&lt;-</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span></span></code></pre></div>
<div id="option-1-differential-expression-analysis-using-limma-1" class="section level3" number="12.7.1">
<h3>
<span class="header-section-number">12.7.1</span> Option 1: Differential expression analysis using limma<a class="anchor" aria-label="anchor" href="#option-1-differential-expression-analysis-using-limma-1"><i class="fas fa-link"></i></a>
</h3>
<p>Note that the code illustrations are based on the following user manual:  (<a href="https://bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf" class="uri">https://bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf</a>)</p>
<p><strong>1. Generate the required input object for limma</strong></p>
<p>As described in “Instructions_Differential_Expression_Analysis”, the gene expression data set expression_data_filt_symbol is currently a data frame and needs to be converted to a DGEList object.</p>
<div class="sourceCode" id="cb202"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># store expression data with corresponding sample conditions in object of class DGEList</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">expression_data_filt_symbol</span>,</span>
<span>             group <span class="op">=</span> <span class="va">sample_conditions</span><span class="op">)</span></span></code></pre></div>
<p>Required function arguments:</p>
<ul>
<li>
<strong>counts</strong>: matrix that contains RNA-Seq data (i.e. count data)</li>
</ul>
<p>Optional function arguments:</p>
<ul>
<li>
<strong>group</strong>: indicates the condition of each sample/library
<ul>
<li>this argument must be specified sooner or later (such as in subsequent functions) so we just specify it at this point.</li>
</ul>
</li>
</ul>
<p>Note that we leave the remaining arguments in their default state since the corresponding info will be added through the following pipeline of functions.</p>
<p><strong>2. Normalization</strong></p>
<p>The following piece of code generates a normalization factor for each sample. This accounts for sample-specific effects (such as differences in library sizes and effects of compositionality). If not accounted for, these effects prevent a comparison between the samples.</p>
<div class="sourceCode" id="cb203"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p>Note that this function does not transform the count data, but rather generates a normalization factor for each sample which is incorporated into the subsequent analysis separately.</p>
<p><strong>3. voom transformation</strong></p>
<p>Note that the voom transformation facilitates the use of the subsequent functions that were initially developed for microarray data.</p>
<div class="sourceCode" id="cb204"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># design matrix (rows correspond to samples, columns indicate which coefficients are to be estimated)</span></span>
<span><span class="va">design_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample_conditions</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># voom transformation (transformation of count data to log-cpm values, computation of observation weight for each</span></span>
<span><span class="co"># gene based on mean-variance relationship)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html">voom</a></span><span class="op">(</span><span class="va">y</span>, design <span class="op">=</span> <span class="va">design_matrix</span><span class="op">)</span></span></code></pre></div>
<p><strong>4. Differential expression analysis</strong></p>
<div class="sourceCode" id="cb205"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fit a linear model for each gene</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate the statistics for the assessment of differential expression</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the results table for each gene whose differential expression was assessed</span></span>
<span><span class="va">DE_results_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">y</span>, number <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Removing intercept from test coefficients</span></span>
<span><span class="co"># Note that number = nrow(y) ensures that all genes are displayed in results, not just a subset</span></span></code></pre></div>
<p><strong>5. Rename columns in results table</strong></p>
<p>This step is not required for differential expression analysis or for the subsequent use of gene set analysis IN GENERAL.</p>
<p>A renaming of some columns (such as adjusted p-value) is necessary in the context of these illustrations since the different methods for differential expression analysis typically differ in the column names of the results tables. A unification of the column names is required so that we can use the same code to illustrate further conduct of GSA in the subsequent code.</p>
<div class="sourceCode" id="cb206"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, we transform to results table to a data frame so that we see the results table directly when accessing it through the name "res"</span></span>
<span><span class="va">DE_results_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">DE_results_limma</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rename the column that contains the adjusted p-values: rename padj to p_adj</span></span>
<span><span class="va">DE_results_limma</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="va">DE_results_limma</span>, p_adj <span class="op">=</span> <span class="va">`adj.P.Val`</span><span class="op">)</span></span></code></pre></div>
<p><strong>6. Inspect the first 10 genes in the results table </strong></p>
<div class="sourceCode" id="cb207"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_limma</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;              logFC  AveExpr         t      P.Value</span></span>
<span><span class="co">#&gt; RPS4Y1   9.2106424 1.986675 47.198797 1.005487e-55</span></span>
<span><span class="co">#&gt; TMSB4Y   5.0807298 0.513264 26.672932 6.011554e-39</span></span>
<span><span class="co">#&gt; PNPLA4  -0.9239961 5.333790 -8.777656 5.619670e-13</span></span>
<span><span class="co">#&gt; PUDP    -0.8543759 2.163068 -4.414608 3.506569e-05</span></span>
<span><span class="co">#&gt; CXorf38 -0.5507332 3.716678 -4.250337 6.330976e-05</span></span>
<span><span class="co">#&gt; TXLNG   -0.4042256 5.116432 -4.043565 1.310034e-04</span></span>
<span><span class="co">#&gt; G0S2    -1.4700846 5.192623 -3.855239 2.497623e-04</span></span>
<span><span class="co">#&gt; JUN     -0.5879335 8.273471 -3.919967 2.004610e-04</span></span>
<span><span class="co">#&gt; APAF1    0.4047167 5.968529  3.823276 2.782012e-04</span></span>
<span><span class="co">#&gt; IL1A    -1.4856099 3.543969 -3.666732 4.683127e-04</span></span>
<span><span class="co">#&gt;                p_adj           B</span></span>
<span><span class="co">#&gt; RPS4Y1  6.038958e-52 84.22462726</span></span>
<span><span class="co">#&gt; TMSB4Y  1.805270e-35 62.69727996</span></span>
<span><span class="co">#&gt; PNPLA4  1.125058e-09 19.21578210</span></span>
<span><span class="co">#&gt; PUDP    5.265113e-02  2.14888617</span></span>
<span><span class="co">#&gt; CXorf38 7.604769e-02  1.55522518</span></span>
<span><span class="co">#&gt; TXLNG   1.311344e-01  0.73590696</span></span>
<span><span class="co">#&gt; G0S2    1.856530e-01  0.15458026</span></span>
<span><span class="co">#&gt; JUN     1.719955e-01  0.11850631</span></span>
<span><span class="co">#&gt; APAF1   1.856530e-01 -0.07239398</span></span>
<span><span class="co">#&gt; IL1A    2.623670e-01 -0.24410772</span></span></code></pre></div>
</div>
<div id="option-2-differential-expression-analysis-using-deseq2-1" class="section level3" number="12.7.2">
<h3>
<span class="header-section-number">12.7.2</span> Option 2: Differential expression analysis using DESeq2<a class="anchor" aria-label="anchor" href="#option-2-differential-expression-analysis-using-deseq2-1"><i class="fas fa-link"></i></a>
</h3>
<p>The official DESeq2 vignette can be found through the following link:<br><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></p>
<p><strong>1. Generate the required input object for DESeq2</strong></p>
<p>DESeq2 operates on the format DESeqDataSet which contains information on the count data, the conditions of the samples and the design (for further information see below). Note that for gene expression data sets that must be imported to R, additional steps are necessary before the following code can be run.</p>
<p><strong>Generate a data frame which contains the condition of each sample</strong></p>
<p>Here, the information on the sample conditions is the only column in the data frame. However, if further variables (such as batch effects) are to be controlled for, the corresponding variables must additionally be added to coldata.</p>
<div class="sourceCode" id="cb208"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here, the information on the sample conditions is the only column in the data frame</span></span>
<span><span class="co"># However, if further variables (such as batch effects) are to be controlled for, the corresponding variables must additionally be added to coldata</span></span>
<span></span>
<span><span class="co"># The names of the samples are stored as the row names of the data frame</span></span>
<span><span class="co"># Important: make sure that the order of the conditions in sample_conditions corresponds to the order of the</span></span>
<span><span class="co"># samples in expression_data</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">sample_conditions</span>,</span>
<span>                      row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">expression_data_filt_symbol</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rename the column header to "condition"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"condition"</span></span>
<span></span>
<span><span class="co"># Recode the variable "condition" as a factor</span></span>
<span><span class="co"># Rename the sample conditions (in this case from "female" and "male") to "untreated" and "treated"</span></span>
<span><span class="co"># Note: make sure that the control level in variable condition is coded as the first level (i.e. "untreated")</span></span>
<span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span>,</span>
<span>                            labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"untreated"</span>,<span class="st">"treated"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Generate the DESeqDataSet</strong></p>
<div class="sourceCode" id="cb209"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">expression_data_filt_symbol</span>,</span>
<span>                            colData <span class="op">=</span> <span class="va">coldata</span>,</span>
<span>                            design <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span><span class="op">)</span></span></code></pre></div>
<p>Relevant function arguments:</p>
<ul>
<li><p><strong>countData</strong>: count data from gene expression data set</p></li>
<li>
<p><strong>colData</strong>: data frame that contains information on the samples (see above)</p>
<ul>
<li>conditions of the samples (required) and possibly further variables to correct for (such as batch effects)</li>
</ul>
</li>
<li>
<p><strong>design</strong>: indicates which variables from colData are used for modelling</p>
<ul>
<li><p><strong>more detailed</strong>: the argument design is used to estimate the dispersions and the log2 fold changes of the model</p></li>
<li><p>if more than one variable from colData are used in argument design (e.g. a second variable “batch”), the syntax changes to the following formula: design ~ batch + condition</p></li>
<li><p>make sure that the variable of interest (here: variable that represents conditions of the samples) is placed at the end of the formula</p></li>
</ul>
</li>
</ul>
<p><strong>2. Differential expression analysis</strong></p>
<div class="sourceCode" id="cb210"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Perform default differential expression analysis</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co"># 2. Generate results table which provides</span></span>
<span><span class="va">DE_results_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre></div>
<p>Description of the functions:</p>
<ul>
<li>
<strong>DESeq()</strong>: Estimation normalization factors, estimation of dispersions, fitting of generalized linear model, Wald statistics</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>results()</strong>: Provides base mean across all samples, estimated log2 fold changes, standard errors, test statistics, p-values, adjusted p-values</li>
</ol>
<p><strong>3. Rename columns in results table</strong></p>
<p>This step is not required for differential expression analysis or for the subsequent use of gene set analysis IN GENERAL.</p>
<p>A renaming of some columns (such as adjusted p-value) is necessary in the context of these illustrations since the different methods for differential expression analysis typically differ in the column names of the results tables. A unification of the column names is required so that we can use the same code to illustrate further conduct of GSA in the subsequent code.</p>
<div class="sourceCode" id="cb211"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, we transform to results table to a data frame so that we see the results table directly when accessing it through the name "res"</span></span>
<span><span class="va">DE_results_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">DE_results_DESeq2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rename the column that contains the adjusted p-values: rename padj to p_adj</span></span>
<span><span class="va">DE_results_DESeq2</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="va">DE_results_DESeq2</span>, p_adj <span class="op">=</span> <span class="st">"padj"</span><span class="op">)</span></span></code></pre></div>
<p><strong>4. Inspect the first 10 genes in the results table</strong></p>
<div class="sourceCode" id="cb212"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_DESeq2</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;         baseMean log2FoldChange      lfcSE        stat</span></span>
<span><span class="co">#&gt; DPM1    62.75664    0.039694123 0.11391433  0.34845593</span></span>
<span><span class="co">#&gt; SCYL3   70.48320    0.028962410 0.11169085  0.25930870</span></span>
<span><span class="co">#&gt; FIRRM   13.00620    0.165415797 0.15394501  1.07451223</span></span>
<span><span class="co">#&gt; FGR     43.59415    0.114144611 0.19558025  0.58362034</span></span>
<span><span class="co">#&gt; FUCA2   37.98066   -0.006462503 0.10902544 -0.05927518</span></span>
<span><span class="co">#&gt; NFYA   228.56697    0.051476343 0.09634771  0.53427676</span></span>
<span><span class="co">#&gt; LAS1L   11.53865    0.025240854 0.18770526  0.13447068</span></span>
<span><span class="co">#&gt; ENPP4  175.91248    0.300292682 0.16292585  1.84312480</span></span>
<span><span class="co">#&gt; BAD     66.39379   -0.264391459 0.10275324 -2.57307183</span></span>
<span><span class="co">#&gt; HS3ST1  10.12757   -0.199051398 0.31889483 -0.62419137</span></span>
<span><span class="co">#&gt;            pvalue     p_adj</span></span>
<span><span class="co">#&gt; DPM1   0.72749780 0.9221933</span></span>
<span><span class="co">#&gt; SCYL3  0.79539707 0.9423095</span></span>
<span><span class="co">#&gt; FIRRM  0.28259317 0.7199487</span></span>
<span><span class="co">#&gt; FGR    0.55947577 0.8668104</span></span>
<span><span class="co">#&gt; FUCA2  0.95273293 0.9859714</span></span>
<span><span class="co">#&gt; NFYA   0.59315007 0.8787295</span></span>
<span><span class="co">#&gt; LAS1L  0.89303039 0.9706589</span></span>
<span><span class="co">#&gt; ENPP4  0.06531079 0.4826632</span></span>
<span><span class="co">#&gt; BAD    0.01008003 0.2998836</span></span>
<span><span class="co">#&gt; HS3ST1 0.53250191 0.8522833</span></span></code></pre></div>
</div>
<div id="option-3-differential-expression-analysis-using-edger-1" class="section level3" number="12.7.3">
<h3>
<span class="header-section-number">12.7.3</span> Option 3: Differential expression analysis using edgeR<a class="anchor" aria-label="anchor" href="#option-3-differential-expression-analysis-using-edger-1"><i class="fas fa-link"></i></a>
</h3>
<p>The instructions for edgeR are based on the following vignette:  (<a href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf" class="uri">https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf</a>)</p>
<p><strong>1. Generate the required input object for edgeR</strong></p>
<p>Store the expression data with corresponding sample conditions in object of class DGEList</p>
<div class="sourceCode" id="cb213"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">expression_data_filt_symbol</span>,</span>
<span>             group <span class="op">=</span> <span class="va">sample_conditions</span><span class="op">)</span></span></code></pre></div>
<p>Required function arguments:</p>
<ul>
<li>
<strong>counts</strong>: matrix that contains RNA-Seq data (i.e. count data)</li>
</ul>
<p>Optional function arguments:</p>
<ul>
<li>
<strong>group</strong>: indicates condition of each sample/library
<ul>
<li>This argument must be specified sooner or later (such as in subsequent functions) so we just specify it at this point</li>
</ul>
</li>
</ul>
<p>Note that we leave the remaining arguments in their default state since the corresponding info will be added through the subsequent pipeline of functions.</p>
<p><strong>2. Normalization</strong></p>
<p>The following piece of code generates a normalization factor for each sample. This accounts for sample-specific effects (such as differences in library sizes and effects of compositionality). If not accounted for, these effects prevent a comparison between the samples</p>
<div class="sourceCode" id="cb214"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p>Note: this function does not transform the count data, but rather generates a normalization factor for each sample which is incorporated into the subsequent analysis separately.</p>
<p><strong>3. Estimation of the dispersion</strong></p>
<p>Estimate the common and tagwise dispersion which quantify the variation of the true abundance of a given gene between different samples and is required to assess differential expression realistically.</p>
<div class="sourceCode" id="cb215"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using classic mode.</span></span></code></pre></div>
<p><strong>4. Differential expression analysis</strong></p>
<div class="sourceCode" id="cb216"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test each gene for differential expression:</span></span>
<span><span class="va">DE_results_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/exactTest.html">exactTest</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract pre-specified number (n) of genes</span></span>
<span><span class="va">DE_results_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span><span class="op">(</span><span class="va">DE_results_edgeR</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">DE_results_edgeR</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Note: argument n specifies the number of top differentially expressed genes to be displayed in the results</span></span>
<span><span class="co"># n = nrow(DE_results) ensures the results of all genes whose differential expression was assessed are displayed</span></span></code></pre></div>
<p><strong>5.Rename columns in results table</strong></p>
<p>This step is not required for differential expression analysis or for the subsequent use of gene set analysis IN GENERAL.</p>
<p>A renaming of some columns (such as adjusted p-value) is necessary in the context of these illustrations since the different methods for differential expression analysis typically differ in the column names of the results tables. A unification of the column names is required so that we can use the same code to illustrate further conduct of GSA in the subsequent code.</p>
<div class="sourceCode" id="cb217"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, we transform to results table to a data frame so that we see the results</span></span>
<span><span class="co"># table directly when accessing it through the name "res"</span></span>
<span><span class="va">DE_results_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">DE_results_edgeR</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rename column that contains adjusted p-values: rename padj to p_adj</span></span>
<span><span class="va">DE_results_edgeR</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="va">DE_results_edgeR</span>, p_adj <span class="op">=</span> <span class="st">"FDR"</span><span class="op">)</span></span></code></pre></div>
<p><strong>6. Inspect the first 10 genes in the results table</strong></p>
<div class="sourceCode" id="cb218"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_edgeR</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;              logFC   logCPM        PValue         p_adj</span></span>
<span><span class="co">#&gt; RPS4Y1  10.1331218 6.155421  0.000000e+00  0.000000e+00</span></span>
<span><span class="co">#&gt; TMSB4Y   5.5792120 2.520038 6.284371e-121 1.887197e-117</span></span>
<span><span class="co">#&gt; PNPLA4  -0.9446936 5.485702  1.873415e-17  3.750577e-14</span></span>
<span><span class="co">#&gt; PUDP    -0.7768043 2.571376  3.072736e-06  4.613713e-03</span></span>
<span><span class="co">#&gt; SNAI1   -1.6460183 3.288600  4.708446e-06  5.469265e-03</span></span>
<span><span class="co">#&gt; GADD45G -1.5871184 2.680046  5.463801e-06  5.469265e-03</span></span>
<span><span class="co">#&gt; THEM5   -1.1256488 2.303307  1.949582e-05  1.672742e-02</span></span>
<span><span class="co">#&gt; CTSW     1.5020170 3.366261  2.908599e-05  2.183631e-02</span></span>
<span><span class="co">#&gt; TXLNG   -0.4252126 5.208043  3.558631e-05  2.271802e-02</span></span>
<span><span class="co">#&gt; TCFL5   -0.6027710 5.756923  3.782554e-05  2.271802e-02</span></span></code></pre></div>
<p>Note that one gene has an adjusted p-value of 0. We will deal with this issue when generating the gene ranking from edgeR.</p>
</div>
</div>
<div id="step-4-generate-gene-ranking" class="section level2" number="12.8">
<h2>
<span class="header-section-number">12.8</span> Step 4: Generate gene ranking<a class="anchor" aria-label="anchor" href="#step-4-generate-gene-ranking"><i class="fas fa-link"></i></a>
</h2>
<p>We will now apply a gene-level statistic to the results table of differential expression analysis to generate a gene ranking which will serve as input to GSEAPreranked.</p>
<p>As we have performed differential expression analysis with
- voom/limma
- DESeq2
- edgeR,
we will also generate three gene rankings, one for each of these three methods.</p>
<p>The formula of the gene-level ranking metric is  
<span class="math display">\[ (-1) * \log_{10}(p\text{-value}) * \text{sign}(\text{log fold change})\]</span>.
Note that by “p-value”, we refer to the non-adjusted p-value.</p>
<div id="option-1-generate-gene-ranking-from-limmavoom-results" class="section level3" number="12.8.1">
<h3>
<span class="header-section-number">12.8.1</span> Option 1: Generate gene ranking from limma/voom results<a class="anchor" aria-label="anchor" href="#option-1-generate-gene-ranking-from-limmavoom-results"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb219"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue_limma</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_limma</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA</span> <span class="op">&lt;-</span> <span class="va">DE_results_limma</span><span class="op">[</span><span class="va">ind_nonNA_pvalue_limma</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_limma</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_limma</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><strong>Inspect the first 10 genes from the gene ranking</strong></p>
<div class="sourceCode" id="cb220"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_limma</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;    RPS4Y1    TMSB4Y     APAF1      LPXN     TRAM2   VIPAS39 </span></span>
<span><span class="co">#&gt; 54.997623 38.221013  3.555641  3.201891  3.165236  3.072060 </span></span>
<span><span class="co">#&gt;    MAP3K1    SLC9A6      PHKB     HCFC2 </span></span>
<span><span class="co">#&gt;  3.019964  2.925713  2.875909  2.816500</span></span></code></pre></div>
</div>
<div id="option-2-generate-gene-ranking-from-deseq2-results" class="section level3" number="12.8.2">
<h3>
<span class="header-section-number">12.8.2</span> Option 2: Generate gene ranking from DESeq2 results<a class="anchor" aria-label="anchor" href="#option-2-generate-gene-ranking-from-deseq2-results"><i class="fas fa-link"></i></a>
</h3>
<p><strong>1.Shrink estimated log fold change values</strong></p>
<p>Before generating the gene ranking from the DESeq2 results, we need to perform an additional shrinkage that is specific to DESeq2. The intuition behind the shrinkage of log fold change values is as follows: As mentioned in the paper, RNA-Seq data consists (in its rare form) of count data which is inherently heteroscedastic, i.e. the variance of the count data depends on the mean count of the count data. It is observable that ratios between counts are considerably noisier in the lower magnitudes of counts compared to higher magnitudes, i.e.the log fold changes between both conditions are higher if the overall magnitude of counts is lower, independent of the actual extent to which the gene is differentially expressed between both conditions. DESeq2 addresses this issue by shrinking the estimated log fold changes in the direction of 0. The magnitude of shrinkage is higher if the available information for a gene is lower (which may be because of a low magnitude of counts, a high dispersion or few degrees of freedom.) A more detailed description is provided in the DESeq2 paper by Love et al. (2014).<br></p>
<p>The shrinkage is performed on object dds, which is a result of the function <em>DESeq()</em>:</p>
<div class="sourceCode" id="cb221"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="st">"apeglm"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"apeglm"</span>, update <span class="op">=</span> <span class="cn">FALSE</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">DE_results_DESeq2_shrink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                                      coef <span class="op">=</span> <span class="st">"condition_treated_vs_untreated"</span>,</span>
<span>                                      type<span class="op">=</span><span class="st">"apeglm"</span><span class="op">)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li>
<strong>type</strong>: method to perform shrinkage
<ul>
<li>we opt for the default “apeglm” but you can choose from two alternative options as well</li>
</ul>
</li>
<li>
<strong>coef</strong>: indicate the coefficients to be shrunk
<ul>
<li>we can obtain the right argument from the following function call:</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb222"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">resultsNames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Intercept"                     </span></span>
<span><span class="co">#&gt; [2] "condition_treated_vs_untreated"</span></span></code></pre></div>
<p>This shows us that we can either shrink the intercept or the “condition_treated_vs_untreated”. Since we do not want to shrink the intercept but the log fold changes, we opt for the second option “condition_treated_vs_untreated”.</p>
<p>Finish up the results table:</p>
<div class="sourceCode" id="cb223"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># transform results table to data frame</span></span>
<span><span class="va">DE_results_DESeq2_shrink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_shrink</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the first 10 genes from the results table: </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_shrink</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;         baseMean log2FoldChange      lfcSE     pvalue</span></span>
<span><span class="co">#&gt; DPM1    62.75664    0.007288891 0.04890425 0.72749780</span></span>
<span><span class="co">#&gt; SCYL3   70.48320    0.005422660 0.04847084 0.79539707</span></span>
<span><span class="co">#&gt; FIRRM   13.00620    0.018798782 0.05470876 0.28259317</span></span>
<span><span class="co">#&gt; FGR     43.59415    0.008038882 0.05231455 0.55947577</span></span>
<span><span class="co">#&gt; FUCA2   37.98066   -0.001246862 0.04796503 0.95273293</span></span>
<span><span class="co">#&gt; NFYA   228.56697    0.012392873 0.04812269 0.59315007</span></span>
<span><span class="co">#&gt; LAS1L   11.53865    0.001891725 0.05141019 0.89303039</span></span>
<span><span class="co">#&gt; ENPP4  175.91248    0.035023088 0.06667768 0.06531079</span></span>
<span><span class="co">#&gt; BAD     66.39379   -0.153104170 0.13823651 0.01008003</span></span>
<span><span class="co">#&gt; HS3ST1  10.12757   -0.005389391 0.05306809 0.53250191</span></span>
<span><span class="co">#&gt;             padj</span></span>
<span><span class="co">#&gt; DPM1   0.9221933</span></span>
<span><span class="co">#&gt; SCYL3  0.9423095</span></span>
<span><span class="co">#&gt; FIRRM  0.7199487</span></span>
<span><span class="co">#&gt; FGR    0.8668104</span></span>
<span><span class="co">#&gt; FUCA2  0.9859714</span></span>
<span><span class="co">#&gt; NFYA   0.8787295</span></span>
<span><span class="co">#&gt; LAS1L  0.9706589</span></span>
<span><span class="co">#&gt; ENPP4  0.4826632</span></span>
<span><span class="co">#&gt; BAD    0.2998836</span></span>
<span><span class="co">#&gt; HS3ST1 0.8522833</span></span></code></pre></div>
<p>Note that the estimated log fold change values (column log2FoldChange) are smaller (in absolute terms!) compared to the regular DESeq2 results, while the (adjusted and non-adjusted) p-value remain unchanged.<br>
Also, we have not performed shrinkage in Chapter ‘Differential Expression Analysis’ as there, the goal was to get the list of differentially expressed genes which are detected solely based on the adjusted p-value.</p>
<p><strong>2. Generate the gene ranking</strong></p>
<div class="sourceCode" id="cb224"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e.</span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_DESeq2_shrink</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value</span></span>
<span><span class="va">DE_results_noNA</span> <span class="op">&lt;-</span> <span class="va">DE_results_DESeq2_shrink</span><span class="op">[</span><span class="va">ind_nonNA_pvalue</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene</span></span>
<span><span class="va">rankvec_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_DESeq2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order</span></span>
<span><span class="va">rankvec_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_DESeq2</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Inspect the first 10 genes from the gene ranking</p>
<div class="sourceCode" id="cb225"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_DESeq2</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;     RPS4Y1     TMSB4Y       CTSW      APAF1     SLC9A6 </span></span>
<span><span class="co">#&gt; 204.393649  63.913902   4.462277   3.748896   3.567843 </span></span>
<span><span class="co">#&gt;       LPXN      HCFC2     MAP3K1    VIPAS39      TRAM2 </span></span>
<span><span class="co">#&gt;   3.558986   3.341867   3.281130   3.272801   3.175153</span></span></code></pre></div>
</div>
<div id="option-3-generate-gene-ranking-from-edger-results" class="section level3" number="12.8.3">
<h3>
<span class="header-section-number">12.8.3</span> Option 3: Generate gene ranking from edgeR results<a class="anchor" aria-label="anchor" href="#option-3-generate-gene-ranking-from-edger-results"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb226"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Subset the gene expression data set to those genes that have a p-value (i.e. </span></span>
<span><span class="co"># which have been NOT been excluded from differential expression analysis)</span></span>
<span></span>
<span><span class="co"># indicate those genes WITH a p-value</span></span>
<span><span class="va">ind_nonNA_pvalue</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DE_results_edgeR</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset gene expression data set to those genes with a p-value </span></span>
<span><span class="va">DE_results_noNA</span> <span class="op">&lt;-</span> <span class="va">DE_results_edgeR</span><span class="op">[</span><span class="va">ind_nonNA_pvalue</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. create vector that contains the value of the gene-level ranking metric for each gene </span></span>
<span><span class="va">rankvec_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">$</span><span class="va">p_adj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. assign respective gene ID to each value in the vector </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rankvec_edgeR</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">DE_results_noNA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. sort the vector in descending order </span></span>
<span><span class="va">rankvec_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">rankvec_edgeR</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><strong>Inspect the first 10 genes from the gene ranking</strong></p>
<div class="sourceCode" id="cb227"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;      RPS4Y1      TMSB4Y        CTSW       APAF1        LPXN </span></span>
<span><span class="co">#&gt;         Inf 116.7241828   1.6608208   1.0328005   1.0016012 </span></span>
<span><span class="co">#&gt;      SLC9A6      MAP3K1       TRAM2        ATP8      MICAL2 </span></span>
<span><span class="co">#&gt;   0.9854321   0.8638818   0.8499991   0.8447814   0.8258597</span></span></code></pre></div>
<p>Note that gene RPS4Y1 has a ranking value of “Inf” (infinity). The reason behind this is that the gene has an adjusted p-value of <span class="math inline">\(0\)</span> in the edgeR results of differential expression analysis (as already observed above). We have to solve this issue, however, there is no common default approach. For the illustration purposes here, we replace the value Inf by the second biggest ranking value in the ranking.</p>
<div class="sourceCode" id="cb228"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># replace value Inf by second biggest value in the ranking </span></span>
<span><span class="va">rankvec_edgeR</span><span class="op">[</span><span class="va">rankvec_edgeR</span> <span class="op">==</span> <span class="cn">Inf</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">rankvec_edgeR</span><span class="op">[</span><span class="va">rankvec_edgeR</span> <span class="op">!=</span> <span class="cn">Inf</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect ranking again</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rankvec_edgeR</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;      RPS4Y1      TMSB4Y        CTSW       APAF1        LPXN </span></span>
<span><span class="co">#&gt; 116.7241828 116.7241828   1.6608208   1.0328005   1.0016012 </span></span>
<span><span class="co">#&gt;      SLC9A6      MAP3K1       TRAM2        ATP8      MICAL2 </span></span>
<span><span class="co">#&gt;   0.9854321   0.8638818   0.8499991   0.8447814   0.8258597</span></span></code></pre></div>
</div>
</div>
<div id="step-5-export-gene-ranking-to-text-file" class="section level2" number="12.9">
<h2>
<span class="header-section-number">12.9</span> Step 5: Export gene ranking to text file<a class="anchor" aria-label="anchor" href="#step-5-export-gene-ranking-to-text-file"><i class="fas fa-link"></i></a>
</h2>
<p>For the purpose of simplicity, we only export one of the gene rankings to a text file. Here, we export the ranking generated with limma.</p>
<div class="sourceCode" id="cb229"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this path indicates that we store the gene ranking as the text file </span></span>
<span><span class="co"># "gene_ranking.txt" in folder "Input_Objects_GSEAPreranked"</span></span>
<span><span class="va">path_conditions</span> <span class="op">&lt;-</span> <span class="st">"./data/Input_Objects_GSEAPreranked/gene_ranking.txt"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">rankvec_limma</span>,</span>
<span>            file <span class="op">=</span> <span class="va">path_conditions</span>,</span>
<span>            quote <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            row.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li><p><strong>x</strong>: object to be exported to a text file</p></li>
<li><p><strong>file</strong>: location (i.e. folder and file) in which the object is to be stored</p></li>
<li><p><strong>quote = FALSE</strong> ensures that none of the characters (in this case gene and sample identifiers) are surrounded by double quotes</p></li>
<li><p><strong>row.names = TRUE</strong> ensures that the gene IDs are included in the export</p></li>
<li><p><strong>col.names = TRUE</strong> ensures that no gene information is in the first row since the web-based tool ignores whatever is in the first row</p></li>
</ul>
</div>
<div id="step-6-further-preparation-in-excel" class="section level2" number="12.10">
<h2>
<span class="header-section-number">12.10</span> Step 6: Further preparation in Excel<a class="anchor" aria-label="anchor" href="#step-6-further-preparation-in-excel"><i class="fas fa-link"></i></a>
</h2>
<p>For this step, follow the instructions on the very bottom of the following link:<br>
(<a href="http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats" class="uri">http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats</a>)</p>
<p>Then, scroll to section “RNK: Ranked list file format (*.rnk)“.</p>

<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-orgHs_package" class="csl-entry">
Carlson, Marc. <em>Org.hs.eg.db: Genome Wide Annotation for Human</em>, 2022.
</div>
<div id="ref-cunningham2022ensembl" class="csl-entry">
Cunningham, Fiona, James E Allen, Jamie Allen, Jorge Alvarez-Jarreta, M Ridwan Amode, Irina M Armean, Olanrewaju Austine-Orimoloye, et al. <span>“Ensembl 2022.”</span> <em>Nucleic Acids Research</em> 50, no. D1 (2022): D988–95.
</div>
<div id="ref-tweeDEseq_package" class="csl-entry">
Gonzalez, Juan R, and Mikel Esnaola. <em>tweeDEseqCountData: RNA-Seq Count Data Employed in the Vignette of the tweeDEseq Package</em>, 2022. <a href="http://www.creal.cat/jrgonzalez/software.htm">http://www.creal.cat/jrgonzalez/software.htm</a>.
</div>
<div id="ref-huang2009bioinformatics" class="csl-entry">
Huang, Da Wei, Brad T Sherman, and Richard A Lempicki. <span>“Bioinformatics Enrichment Tools: Paths Toward the Comprehensive Functional Analysis of Large Gene Lists.”</span> <em>Nucleic Acids Research</em> 37, no. 1 (2009): 1–13.
</div>
<div id="ref-huang2009systematic" class="csl-entry">
———. <span>“<span class="nocase">Systematic and integrative analysis of large gene lists using DAVID bioinformatics resources</span>.”</span> <em>Nature Protocols</em> 4, no. 1 (2009): 44–57.
</div>
<div id="ref-huber2015orchestrating" class="csl-entry">
Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. <span>“Orchestrating High-Throughput Genomic Analysis with <span>Bioconductor</span>.”</span> <em>Nature Methods</em> 12, no. 2 (2015): 115–21.
</div>
<div id="ref-law2014voom" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. <span>“Voom: Precision Weights Unlock Linear Model Analysis Tools for <span class="nocase">RNA-seq</span> Read Counts.”</span> <em>Genome Biology</em> 15, no. 2 (2014): 1–17.
</div>
<div id="ref-love2014moderated" class="csl-entry">
Love, Michael I, Wolfgang Huber, and Simon Anders. <span>“<span class="nocase">Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2</span>.”</span> <em>Genome Biology</em> 15 (2014): 550.
</div>
<div id="ref-mootha2003pgc" class="csl-entry">
Mootha, Vamsi K, Cecilia M Lindgren, Karl-Fredrik Eriksson, et al. <span>“<span>PGC</span>-1<span class="math inline">\(\alpha\)</span>-Responsive Genes Involved in Oxidative Phosphorylation Are Coordinately Downregulated in Human Diabetes.”</span> <em>Nature Genetics</em> 34, no. 3 (2003): 267–73.
</div>
<div id="ref-pickrell2010understanding" class="csl-entry">
Pickrell, Joseph K, John C Marioni, Athma A Pai, Jacob F Degner, Barbara E Engelhardt, Everlyne Nkadori, Jean-Baptiste Veyrieras, Matthew Stephens, Yoav Gilad, and Jonathan K Pritchard. <span>“Understanding Mechanisms Underlying Human Gene Expression Variation with RNA Sequencing.”</span> <em>Nature</em> 464, no. 7289 (2010): 768–72.
</div>
<div id="ref-ritchie2015limma" class="csl-entry">
Ritchie, Matthew E, Belinda Phipson, DI Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. <span>“<span class="nocase">limma</span> Powers Differential Expression Analyses for <span>RNA-</span>Sequencing and Microarray Studies.”</span> <em>Nucleic Acids Research</em> 43, no. 7 (2015): e47–47.
</div>
<div id="ref-robinson2010edger" class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. <span>“<span class="nocase">edgeR: a Bioconductor package for differential expression analysis of digital gene expression data</span>.”</span> <em>Bioinformatics</em> 26, no. 1 (2010): 139–40.
</div>
<div id="ref-robinson2010scaling" class="csl-entry">
Robinson, Mark D, and Alicia Oshlack. <span>“A Scaling Normalization Method for Differential Expression Analysis of <span class="nocase">RNA-seq</span> Data.”</span> <em>Genome Biology</em> 11, no. 3 (2010): 1–9.
</div>
<div id="ref-sayers2022database" class="csl-entry">
Sayers, Eric W, Evan E Bolton, J Rodney Brister, Kathi Canese, Jessica Chan, Donald C Comeau, Ryan Connor, et al. <span>“Database Resources of the National Center for Biotechnology Information.”</span> <em>Nucleic Acids Research</em> 50, no. D1 (2022): D20–26.
</div>
<div id="ref-seal2023genenames" class="csl-entry">
Seal, Ruth L, Bryony Braschi, Kristian Gray, Tamsin EM Jones, Susan Tweedie, Liora Haim-Vilmovsky, and Elspeth A Bruford. <span>“Genenames. Org: The HGNC Resources in 2023.”</span> <em>Nucleic Acids Research</em> 51, no. D1 (2023): D1003–9.
</div>
<div id="ref-subramanian2005gene" class="csl-entry">
Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. <span>“Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.”</span> <em>Proceedings of the National Academy of Sciences</em> 102, no. 43 (2005): 15545–50.
</div>
<div id="ref-tarca2013comparison" class="csl-entry">
Tarca, Adi L, Gaurav Bhatti, and Roberto Romero. <span>“<span class="nocase">A Comparison of Gene Set Analysis Methods in Terms of Sensitivity, Prioritization and Specificity</span>.”</span> <em>PloS One</em> 8, no. 11 (2013): e79217.
</div>
<div id="ref-wu2021clusterprofiler" class="csl-entry">
Wu, Tianzhi, Erqiang Hu, Shuangbin Xu, Meijun Chen, Pingfan Guo, Zehan Dai, Tingze Feng, et al. <span>“clusterProfiler 4.0: A Universal Enrichment Tool for Interpreting Omics Data.”</span> <em>The Innovation</em> 2, no. 3 (2021): 100141.
</div>
<div id="ref-young2010gene" class="csl-entry">
Young, Matthew D, Matthew J Wakefield, Gordon K Smyth, and Alicia Oshlack. <span>“Gene Ontology Analysis for <span class="nocase">RNA-seq</span>: Accounting for Selection Bias.”</span> <em>Genome Biology</em> 11, no. 2 (2010): 1–12.
</div>
</div>
</div>
</div>




























  <div class="chapter-nav">
<div class="prev"><a href="padog.html"><span class="header-section-number">11</span> PADOG</a></div>
<div class="empty"></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#gseapreranked"><span class="header-section-number">12</span> GSEAPreranked</a></li>
<li><a class="nav-link" href="#libraries-11"><span class="header-section-number">12.1</span> Libraries</a></li>
<li><a class="nav-link" href="#load-data-10"><span class="header-section-number">12.2</span> Load data</a></li>
<li><a class="nav-link" href="#step-1-convert-ensembl-ids-to-hgnc-symbols"><span class="header-section-number">12.3</span> Step 1: Convert Ensembl IDs to HGNC symbols</a></li>
<li><a class="nav-link" href="#step-2-take-a-closer-look-at-the-duplicated-gene-ids-1"><span class="header-section-number">12.4</span> Step 2: Take a closer look at the duplicated gene IDs</a></li>
<li>
<a class="nav-link" href="#step-3-remove-the-duplicated-gene-ids"><span class="header-section-number">12.5</span> Step 3: Remove the duplicated gene IDs</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-keep-the-duplicate-with-the-lowest-subscript"><span class="header-section-number">12.5.1</span> Option 1: Keep the duplicate with the lowest subscript</a></li>
<li><a class="nav-link" href="#option-2-keep-the-rounded-mean-expression-value-of-all-duplicated-gene-ids-1"><span class="header-section-number">12.5.2</span> Option 2: Keep the (rounded) mean expression value of all duplicated gene IDs</a></li>
<li><a class="nav-link" href="#option-3-among-the-duplicates-keep-row-with-highest-overall-expression-values-i.e-highest-counts-across-all-samples"><span class="header-section-number">12.5.3</span> Option 3: Among the duplicates, keep row with highest overall expression values (i.e highest counts across all samples)</a></li>
</ul>
</li>
<li><a class="nav-link" href="#intermediate-step-choose-which-converted-gene-expression-data-sets-to-proceed-with"><span class="header-section-number">12.6</span> Intermediate step: Choose which converted gene expression data sets to proceed with</a></li>
<li>
<a class="nav-link" href="#step-3-differential-expression-analysis"><span class="header-section-number">12.7</span> Step 3: Differential expression analysis</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-differential-expression-analysis-using-limma-1"><span class="header-section-number">12.7.1</span> Option 1: Differential expression analysis using limma</a></li>
<li><a class="nav-link" href="#option-2-differential-expression-analysis-using-deseq2-1"><span class="header-section-number">12.7.2</span> Option 2: Differential expression analysis using DESeq2</a></li>
<li><a class="nav-link" href="#option-3-differential-expression-analysis-using-edger-1"><span class="header-section-number">12.7.3</span> Option 3: Differential expression analysis using edgeR</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#step-4-generate-gene-ranking"><span class="header-section-number">12.8</span> Step 4: Generate gene ranking</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-generate-gene-ranking-from-limmavoom-results"><span class="header-section-number">12.8.1</span> Option 1: Generate gene ranking from limma/voom results</a></li>
<li><a class="nav-link" href="#option-2-generate-gene-ranking-from-deseq2-results"><span class="header-section-number">12.8.2</span> Option 2: Generate gene ranking from DESeq2 results</a></li>
<li><a class="nav-link" href="#option-3-generate-gene-ranking-from-edger-results"><span class="header-section-number">12.8.3</span> Option 3: Generate gene ranking from edgeR results</a></li>
</ul>
</li>
<li><a class="nav-link" href="#step-5-export-gene-ranking-to-text-file"><span class="header-section-number">12.9</span> Step 5: Export gene ranking to text file</a></li>
<li><a class="nav-link" href="#step-6-further-preparation-in-excel"><span class="header-section-number">12.10</span> Step 6: Further preparation in Excel</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
