<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 8 GSEA (web-based tool) | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will export the transformed RNA-Seq measurements prepare and export sample conditions Note that some further preprocessing steps are required in Excel. For these further...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 8 GSEA (web-based tool) | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will export the transformed RNA-Seq measurements prepare and export sample conditions Note that some further preprocessing steps are required in Excel. For these further...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 8 GSEA (web-based tool) | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will export the transformed RNA-Seq measurements prepare and export sample conditions Note that some further preprocessing steps are required in Excel. For these further...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="active" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="gsea-web-based-tool" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> GSEA (web-based tool)<a class="anchor" aria-label="anchor" href="#gsea-web-based-tool"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will</p>
<ol style="list-style-type: decimal">
<li><p>export the transformed RNA-Seq measurements</p></li>
<li><p>prepare and export sample conditions</p></li>
</ol>
<p>Note that some further preprocessing steps are required in Excel. For these further preprocessing steps, click on the following link:<br><a href="http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats" class="uri">http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats</a>.</p>
<div id="libraries-7" class="section level2" number="8.1">
<h2>
<span class="header-section-number">8.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-7"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"tweeDEseqCountData"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load libraries</strong>:</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.creal.cat/jrgonzalez/software.htm">tweeDEseqCountData</a></span><span class="op">)</span> </span></code></pre></div>
<p>Description of the library:</p>
<ul>
<li>
<strong>tweeDEseqCountData</strong>: Here, we need the conditions of the samples of the pickrell data set because these are (in a processed form) required for the web-based tool GSEA.</li>
</ul>
</div>
<div id="load-data-6" class="section level2" number="8.2">
<h2>
<span class="header-section-number">8.2</span> Load data<a class="anchor" aria-label="anchor" href="#load-data-6"><i class="fas fa-link"></i></a>
</h2>
<p>For the web-based tool, we work with the pre-filtered and transformed gene expression measurements. Since GSEA accepts a wide variety of gene ID formats, we can input the gene IDs in the Ensembl ID format.<br>
Note that for the purpose of simplicity, we here work with the voom-transformed data. However, you can easily switch to the gene expression data set transformed using DESeq2’s varianceStabilizingTransformation.</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"./data/Results_RNASeq_Transformation/expression_data_voomtransformed_Ensembl.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or alternatively:</span></span>
<span><span class="co"># load("./data/Results_RNASeq_Transformation/expression_data_vsttransformed_Ensembl.Rdata")</span></span></code></pre></div>
<p>For an easier readability, we store the gene expression data set in a new object with a neutral name.</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_transformed</span> <span class="op">&lt;-</span> <span class="va">expression_data_voomtransformed_Ensembl</span></span></code></pre></div>
<p>As mentioned above, GSEA requires (a preprocessed version of) the sample conditions. In the case of the pickrell data set used for these illustrations, we obtain these using the following commands:</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load pickrell data </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pickrell</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># store sample conditions</span></span>
<span><span class="va">sample_conditions</span> <span class="op">&lt;-</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span></span></code></pre></div>
<div id="step-1-export-the-transformed-gene-expression-measurements" class="section level3" number="8.2.1">
<h3>
<span class="header-section-number">8.2.1</span> step 1: Export the (transformed) gene expression measurements<a class="anchor" aria-label="anchor" href="#step-1-export-the-transformed-gene-expression-measurements"><i class="fas fa-link"></i></a>
</h3>
<p>We export the transformed gene expression measurements to a .txt file and into the pre-specified folder “Input_Objects_GSEA_web”.</p>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Generate the path that indicates that we will store the transformed gene expression measurements in the object "expression_data_transformed.txt, which is located in the file "Input_Objects_GSEA_web"</span></span>
<span><span class="va">path_measurements</span>  <span class="op">&lt;-</span> <span class="st">"./data/Input_Objects_GSEA_web/expression_data_transformed.txt"</span></span>
<span></span>
<span><span class="co"># export the gene expression measurements </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">expression_data_transformed</span>,</span>
<span>            file <span class="op">=</span> <span class="va">path_measurements</span>,</span>
<span>            quote <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            row.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Arguments of function <em>write.table()</em>:</p>
<ul>
<li><p><strong>‘quote = FALSE’</strong>: ensures that none of the characters (in this case the gene and sample
identifiers, i.e. row and column names) are surrounded by double quotes</p></li>
<li><p><strong>‘row.names = TRUE’</strong>: ensures that the gene IDs are included in the export</p></li>
<li><p><strong>‘col.names = TRUE’</strong>: ensures that the samples IDs are included in the export</p></li>
</ul>
</div>
<div id="step-2-prepare-and-export-the-sample-conditions" class="section level3" number="8.2.2">
<h3>
<span class="header-section-number">8.2.2</span> step 2: Prepare and export the sample conditions<a class="anchor" aria-label="anchor" href="#step-2-prepare-and-export-the-sample-conditions"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Prepare the sample conditions</strong></p>
<p>GSEA accepts the sample conditions in the (binary) format with values 0 and 1. For the pickrell data set, we inspect the current values in which the sample conditions are stored:</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># inspect the raw sample conditions:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_conditions</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] male   male   female male   female male   female male  </span></span>
<span><span class="co">#&gt;  [9] female male  </span></span>
<span><span class="co">#&gt; Levels: female male</span></span>
<span></span>
<span><span class="co"># inspect the levels of the sample conditions:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">levels</a></span><span class="op">(</span><span class="va">sample_conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "female" "male"</span></span></code></pre></div>
<p>Currently, the sample conditions are coded as “female” and “male”. We now want to convert both levels to 0 and 1:</p>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (i) create vector to contain the sample conditions in the right format</span></span>
<span><span class="va">sample_conditions_prepared</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (ii) assign all "females", which is the first level of the factor, the value 0</span></span>
<span><span class="va">sample_conditions_prepared</span><span class="op">[</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">levels</a></span><span class="op">(</span><span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># (iii) assign all "males", which is the second level of the factor, the value 1</span></span>
<span><span class="va">sample_conditions_prepared</span><span class="op">[</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">levels</a></span><span class="op">(</span><span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></code></pre></div>
<p>Inspect the (first few) sample conditions:</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_conditions_prepared</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] 1 1 0 1 0 1 0 1 0 1</span></span></code></pre></div>
<p><strong>Export the sample conditions</strong>:</p>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the following path indicates that we store the prepared sample conditions in the object "sample_conditions_prepared.txt" in the file "Input_Objects_GSEA_web". </span></span>
<span><span class="va">path_conditions</span> <span class="op">&lt;-</span> <span class="st">"./data/Input_Objects_GSEA_web/sample_conditions_prepared.txt"</span></span>
<span></span>
<span><span class="co"># export </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample_conditions_prepared</span>,</span>
<span>            file <span class="op">=</span> <span class="va">path_conditions</span>,</span>
<span>            quote <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            row.names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Arguments of function <em>write.table()</em>:</p>
<ul>
<li><p><strong>‘quote = FALSE’</strong>: ensures that none of the characters (in this case the gene and sample
identifiers, i.e. row and column names) are surrounded by double quotes</p></li>
<li><p><strong>‘row.names = TRUE’</strong> ensures that the gene IDs are included in the export</p></li>
<li><p><strong>‘col.names = TRUE’</strong> ensures that the samples IDs are included in the export</p></li>
</ul>
</div>
<div id="step-3-further-preparation-of-input-object-in-excel" class="section level3" number="8.2.3">
<h3>
<span class="header-section-number">8.2.3</span> step 3: Further preparation of input object in Excel<a class="anchor" aria-label="anchor" href="#step-3-further-preparation-of-input-object-in-excel"><i class="fas fa-link"></i></a>
</h3>
<p>For instructions of the further preparation of the text file in Excel, open the following link:<br><a href="http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats" class="uri">http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats</a><br></p>
<p>Follow the instructions in section</p>
<ul>
<li><p><strong>‘GCT: Gene Cluster Text file format (.gct)’</strong>: for further preparation of expression_data_transformed</p></li>
<li><p><strong>‘CLS: Categorical (e.g tumor vs. normal) class file format (.cls)’</strong>: for further preparation of sample_conditions_prepared</p></li>
</ul>
</div>
</div>
</div>



  <div class="chapter-nav">
<div class="prev"><a href="david.html"><span class="header-section-number">7</span> DAVID</a></div>
<div class="next"><a href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#gsea-web-based-tool"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li><a class="nav-link" href="#libraries-7"><span class="header-section-number">8.1</span> Libraries</a></li>
<li>
<a class="nav-link" href="#load-data-6"><span class="header-section-number">8.2</span> Load data</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#step-1-export-the-transformed-gene-expression-measurements"><span class="header-section-number">8.2.1</span> step 1: Export the (transformed) gene expression measurements</a></li>
<li><a class="nav-link" href="#step-2-prepare-and-export-the-sample-conditions"><span class="header-section-number">8.2.2</span> step 2: Prepare and export the sample conditions</a></li>
<li><a class="nav-link" href="#step-3-further-preparation-of-input-object-in-excel"><span class="header-section-number">8.2.3</span> step 3: Further preparation of input object in Excel</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
