<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 11 PADOG | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will go through the process of running PADOG. Note that PADOG requires the transformed RNA-Seq data and the genes in the Entrez ID format.  11.1 Libraries All necessary packages...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 11 PADOG | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will go through the process of running PADOG. Note that PADOG requires the transformed RNA-Seq data and the genes in the Entrez ID format.  11.1 Libraries All necessary packages...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 11 PADOG | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will go through the process of running PADOG. Note that PADOG requires the transformed RNA-Seq data and the genes in the Entrez ID format.  11.1 Libraries All necessary packages...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="active" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="padog" class="section level1" number="11">
<h1>
<span class="header-section-number">11</span> PADOG<a class="anchor" aria-label="anchor" href="#padog"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will go through the process of running PADOG. Note that PADOG requires the transformed RNA-Seq data and the genes in the Entrez ID format.</p>
<div id="libraries-10" class="section level2" number="11.1">
<h2>
<span class="header-section-number">11.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries-10"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"PADOG"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"tweeDEseqCountData"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"KEGGREST"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load libraries</strong></p>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">PADOG</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.creal.cat/jrgonzalez/software.htm">tweeDEseqCountData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">KEGGREST</span><span class="op">)</span></span></code></pre></div>
<p>Summary of the packages:</p>
<ul>
<li><p><strong>PADOG</strong>: Provides the implementation of the method PADOG</p></li>
<li><p><strong>tweeDEseqCountData</strong>: In addition to the preprocessed Pickrell data set, which we load in the following step, PADOG requires the conditions of the samples</p></li>
<li><p><strong>KEGGREST</strong>: To obtain the character value for a variety of organisms the user is required to provide in the function <em>padog()</em>. While the human organism is indicated by default, it might become necessary to consult the KEGGREST package if working with a different organism</p></li>
</ul>
</div>
<div id="load-data-9" class="section level2" number="11.2">
<h2>
<span class="header-section-number">11.2</span> Load Data<a class="anchor" aria-label="anchor" href="#load-data-9"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we load the voom-transformed Pickrell data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/Results_RNASeq_Transformation/expression_data_voomtransformed_Entrez.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># alternatively: load the gene expression measurements that have been transformed using </span></span>
<span><span class="co"># load("data/expression_data_vsttransformed_Entrez.Rdata")</span></span>
<span></span>
<span><span class="co"># additionally, we load the pickrell data set so that we can access the sample conditions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pickrell</span><span class="op">)</span></span></code></pre></div>
<p>The sample conditions (i.e. phenotype labels) of the pickrell data set can be accessed using</p>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span> </span>
<span><span class="co">#&gt;  [1] male   male   female male   female male   female male  </span></span>
<span><span class="co">#&gt;  [9] female male   female male   female male   female male  </span></span>
<span><span class="co">#&gt; [17] female female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [25] female male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [33] female male   female female female female male   female</span></span>
<span><span class="co">#&gt; [41] male   male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [49] female female male   female male   male   female female</span></span>
<span><span class="co">#&gt; [57] male   female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [65] female female female male   female</span></span>
<span><span class="co">#&gt; Levels: female male</span></span></code></pre></div>
<p>We proceed with the voom-transformed pickrell data set and the corresponding phenotype labels.</p>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># gene expression measurements (transformed)</span></span>
<span><span class="co"># note: you can also proceed with the vst-transformed gene expression measurements </span></span>
<span><span class="va">expression_data_transformed</span> <span class="op">&lt;-</span> <span class="va">expression_data_voomtransformed_Entrez</span></span>
<span></span>
<span><span class="co"># sample conditions</span></span>
<span><span class="va">sample_conditions</span> <span class="op">&lt;-</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span></span></code></pre></div>
</div>
<div id="padog-1" class="section level2" number="11.3">
<h2>
<span class="header-section-number">11.3</span> PADOG<a class="anchor" aria-label="anchor" href="#padog-1"><i class="fas fa-link"></i></a>
</h2>
<div id="step-1-prepare-sample-conditions" class="section level3" number="11.3.1">
<h3>
<span class="header-section-number">11.3.1</span> Step 1: Prepare Sample Conditions<a class="anchor" aria-label="anchor" href="#step-1-prepare-sample-conditions"><i class="fas fa-link"></i></a>
</h3>
<p>First, we inspect the form of the initial (raw) sample conditions.</p>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># look at the class: </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">sample_conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "factor"</span></span>
<span><span class="co"># -&gt; the sample labels are already coded as factors</span></span>
<span></span>
<span><span class="co"># the current levels are:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">levels</a></span><span class="op">(</span><span class="va">sample_conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "female" "male"</span></span></code></pre></div>
<p>PADOG requires a character vector with the class labels of the samples. It can only contain “c” for control samples or “d” for disease samples.</p>
<div class="sourceCode" id="cb172"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare sample conditions</span></span>
<span><span class="co"># we want to convert </span></span>
<span><span class="co"># (i) "female" to "c"</span></span>
<span><span class="co"># (ii) "male" to "d"</span></span>
<span><span class="va">sample_conditions_prep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">sample_conditions</span>, </span>
<span>                                levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"female"</span>,<span class="st">"male"</span><span class="op">)</span>, </span>
<span>                                labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"c"</span>,<span class="st">"d"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Inspect the prepared sample conditions:</p>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_conditions_prep</span></span>
<span><span class="co">#&gt;  [1] d d c d c d c d c d c d c d c d c c d c d c c d c d c c</span></span>
<span><span class="co">#&gt; [29] d c c d c d c c c c d c d d c c d c c d c c d c d d c c</span></span>
<span><span class="co">#&gt; [57] d c d c d c c d c c c d c</span></span>
<span><span class="co">#&gt; Levels: c d</span></span></code></pre></div>
</div>
<div id="step-2-run-padog" class="section level3" number="11.3.2">
<h3>
<span class="header-section-number">11.3.2</span> Step 2: Run PADOG<a class="anchor" aria-label="anchor" href="#step-2-run-padog"><i class="fas fa-link"></i></a>
</h3>
<p><strong>It is recommended to set a seed to ensure exact reproducibility of the results if the code is run at multiple time points</strong></p>
<p>You can specify any integer number as the seed. It is VERY IMPORTANT to choose the seed arbitrarily and WITHOUT INSPECTING the results. The seed should NEVER be specified based on which value yields the most preferable results.</p>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run PADOG: </span></span>
<span><span class="va">PADOG_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PADOG/man/padog.html">padog</a></span><span class="op">(</span>esetm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">expression_data_transformed</span><span class="op">)</span>, </span>
<span>                        group <span class="op">=</span> <span class="va">sample_conditions_prep</span>, </span>
<span>                        dseed <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Function arguments:</p>
<ul>
<li>
<p><strong>esetm</strong>: matrix that contains the expression measurements</p>
<ul>
<li>note: since the expression data is initially stored in a data frame, we transform it to a matrix when running PADOG</li>
</ul>
</li>
<li><p><strong>group</strong>: sample conditions (has values “c” and “d”)</p></li>
<li><p><strong>dseed</strong> : seed for random number generation (used in the process of phenotype permutation)</p></li>
</ul>
<p>additional arguments (we do not use here):</p>
<ul>
<li><p><strong>paired</strong>: indicates whether the samples in both groups are paired</p></li>
<li><p><strong>block</strong>: if the samples are paired (i.e. argument paired = TRUE), then the paired samples must have the same block value</p></li>
<li>
<p><strong>gslist</strong>: gives instructions on how to cluster the genes into gene sets</p>
<ul>
<li><p><strong>‘gslist = “KEGGRESTpathway”’</strong> (default): gene sets correspond to KEGG pathways</p></li>
<li><p>alternative: provide a user-defined list of gene sets</p></li>
</ul>
</li>
<li>
<p><strong>organism</strong>: organism from which the gene expression measurements are taken</p>
<ul>
<li><p>for human, set organism = “hsa”</p></li>
<li><p>the required character value for other organisms can be extracted from the KEGGREST package:</p></li>
<li><p>obtain required organisms from column organism</p></li>
</ul>
</li>
<li>
<p><strong>annotation</strong>: required if gslist is set to “KEGGRESTpathway” and the rownames of esetm are probe IDs</p>
<ul>
<li><p>can be set to NULL of gslist is set to “KEGGRESTpathway” and the rownames of esetm are in the Entrez gene ID format</p></li>
<li><p>if rownames are other gene IDs, then sett annotation = NULL and make sure that the rownames are elements of gslist (and unique!)</p></li>
</ul>
</li>
<li>
<p><strong>gs.names</strong>: contains names of gene sets -&gt; character vector</p>
<ul>
<li>must have the same length as gslist</li>
</ul>
</li>
<li><p><strong>NI</strong>: number of phenotype permutations employed in the assessment of the significance of a given gene set</p></li>
</ul>
<p>We want to take a first look at the results table. Note, however, that that we still have to add the adjusted p-values (see next step), based on which differential enrichment is eventually assessed.</p>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">PADOG_results</span> , n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Name    ID Size meanAbsT0   padog0 PmeanAbsT Ppadog</span></span>
<span><span class="co">#&gt; 04380 &lt;NA&gt; 04380   66  2.622104 3.832812     0.005  0.002</span></span>
<span><span class="co">#&gt; 04810 &lt;NA&gt; 04810  111  2.326027 5.476170     0.004  0.004</span></span>
<span><span class="co">#&gt; 03010 &lt;NA&gt; 03010   89  5.378671 9.659220     0.014  0.010</span></span>
<span><span class="co">#&gt; 05213 &lt;NA&gt; 05213   23  1.528150 1.461810     0.027  0.010</span></span>
<span><span class="co">#&gt; 05171 &lt;NA&gt; 05171  118  4.788746 8.020661     0.013  0.013</span></span>
<span><span class="co">#&gt; 05417 &lt;NA&gt; 05417  112  2.094534 3.096663     0.015  0.017</span></span>
<span><span class="co">#&gt; 05162 &lt;NA&gt; 05162   73  2.089532 3.042040     0.010  0.020</span></span>
<span><span class="co">#&gt; 00760 &lt;NA&gt; 00760   17  1.871113 3.583774     0.026  0.023</span></span>
<span><span class="co">#&gt; 04064 &lt;NA&gt; 04064   59  1.931306 3.552921     0.019  0.025</span></span>
<span><span class="co">#&gt; 04010 &lt;NA&gt; 04010  169  1.987648 4.770146     0.018  0.028</span></span></code></pre></div>
</div>
<div id="step-3-adjust-for-multiple-testing" class="section level3" number="11.3.3">
<h3>
<span class="header-section-number">11.3.3</span> Step 3: Adjust for Multiple Testing<a class="anchor" aria-label="anchor" href="#step-3-adjust-for-multiple-testing"><i class="fas fa-link"></i></a>
</h3>
<p>PADOG does <strong>not</strong> perform multiple testing adjustment internally so that must be performed by the user. We here work with a function from the R package base which uses the method by Benjamini and Hochberg by default.</p>
<div class="sourceCode" id="cb176"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add adjusted p-value in column Ppadog_adjusted</span></span>
<span><span class="va">PADOG_results</span><span class="op">$</span><span class="va">Ppadog_adjusted</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">PADOG_results</span><span class="op">$</span><span class="va">Ppadog</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="step-4-interpretation-of-the-results-1" class="section level3" number="11.3.4">
<h3>
<span class="header-section-number">11.3.4</span> Step 4: Interpretation of the results<a class="anchor" aria-label="anchor" href="#step-4-interpretation-of-the-results-1"><i class="fas fa-link"></i></a>
</h3>
<p>We want to inspect (a part of) the results and interpret the columns provided in the results table.</p>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">PADOG_results</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Name    ID Size meanAbsT0   padog0 PmeanAbsT Ppadog</span></span>
<span><span class="co">#&gt; 04380 &lt;NA&gt; 04380   66  2.622104 3.832812     0.005  0.002</span></span>
<span><span class="co">#&gt; 04810 &lt;NA&gt; 04810  111  2.326027 5.476170     0.004  0.004</span></span>
<span><span class="co">#&gt; 03010 &lt;NA&gt; 03010   89  5.378671 9.659220     0.014  0.010</span></span>
<span><span class="co">#&gt; 05213 &lt;NA&gt; 05213   23  1.528150 1.461810     0.027  0.010</span></span>
<span><span class="co">#&gt; 05171 &lt;NA&gt; 05171  118  4.788746 8.020661     0.013  0.013</span></span>
<span><span class="co">#&gt; 05417 &lt;NA&gt; 05417  112  2.094534 3.096663     0.015  0.017</span></span>
<span><span class="co">#&gt; 05162 &lt;NA&gt; 05162   73  2.089532 3.042040     0.010  0.020</span></span>
<span><span class="co">#&gt; 00760 &lt;NA&gt; 00760   17  1.871113 3.583774     0.026  0.023</span></span>
<span><span class="co">#&gt; 04064 &lt;NA&gt; 04064   59  1.931306 3.552921     0.019  0.025</span></span>
<span><span class="co">#&gt; 04010 &lt;NA&gt; 04010  169  1.987648 4.770146     0.018  0.028</span></span>
<span><span class="co">#&gt;       Ppadog_adjusted</span></span>
<span><span class="co">#&gt; 04380           0.696</span></span>
<span><span class="co">#&gt; 04810           1.000</span></span>
<span><span class="co">#&gt; 03010           1.000</span></span>
<span><span class="co">#&gt; 05213           1.000</span></span>
<span><span class="co">#&gt; 05171           1.000</span></span>
<span><span class="co">#&gt; 05417           1.000</span></span>
<span><span class="co">#&gt; 05162           1.000</span></span>
<span><span class="co">#&gt; 00760           1.000</span></span>
<span><span class="co">#&gt; 04064           1.000</span></span>
<span><span class="co">#&gt; 04010           1.000</span></span></code></pre></div>
<p>Differential enrichment of a given gene set can now be assessed based on the adjusted p-value in column <strong>Ppadog_adjusted</strong>. For instance: detect all gene sets with Ppadog_adjusted &lt; 0.05 as differentially enriched.</p>
<p>Additional columns:</p>
<ul>
<li><p><strong>Name</strong>: Name of gene set</p></li>
<li><p><strong>ID:</strong> Identifier of gene set</p></li>
<li><p><strong>Size</strong>: number of genes in gene set</p></li>
<li><p><strong>meanAbsT0</strong>: Mean of absolute (moderated) t-statistic of all genes that are a member of the given gene set</p></li>
<li><p><strong>padog0</strong>: Mean of abolute weighted moderate t-statistic of all genes that are a member of the given gene set</p></li>
<li><p><strong>PmeanAbsT</strong>: significance of of meanAbsT0</p></li>
<li><p><strong>Ppadog</strong>: significance of padog0</p></li>
</ul>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></div>
<div class="next"><a href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#padog"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="nav-link" href="#libraries-10"><span class="header-section-number">11.1</span> Libraries</a></li>
<li><a class="nav-link" href="#load-data-9"><span class="header-section-number">11.2</span> Load Data</a></li>
<li>
<a class="nav-link" href="#padog-1"><span class="header-section-number">11.3</span> PADOG</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#step-1-prepare-sample-conditions"><span class="header-section-number">11.3.1</span> Step 1: Prepare Sample Conditions</a></li>
<li><a class="nav-link" href="#step-2-run-padog"><span class="header-section-number">11.3.2</span> Step 2: Run PADOG</a></li>
<li><a class="nav-link" href="#step-3-adjust-for-multiple-testing"><span class="header-section-number">11.3.3</span> Step 3: Adjust for Multiple Testing</a></li>
<li><a class="nav-link" href="#step-4-interpretation-of-the-results-1"><span class="header-section-number">11.3.4</span> Step 4: Interpretation of the results</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
