<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 1 Pre-filtering | A Practical Guide through Running Gene Set Analysis in R</title>
<meta name="author" content="Milena Wünsch and Pat Callahan">
<meta name="description" content="In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across all samples. We distinguish between these two approaches since the...">
<meta name="generator" content="bookdown 0.34 with bs4_book()">
<meta property="og:title" content="Chapter 1 Pre-filtering | A Practical Guide through Running Gene Set Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across all samples. We distinguish between these two approaches since the...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 1 Pre-filtering | A Practical Guide through Running Gene Set Analysis in R">
<meta name="twitter:description" content="In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across all samples. We distinguish between these two approaches since the...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">A Practical Guide through Running Gene Set Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About</a></li>
<li class="book-part">Common Processing Steps</li>
<li><a class="active" href="pre-filtering.html"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="" href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></li>
<li><a class="" href="differential-expression-analysis.html"><span class="header-section-number">3</span> Differential expression analysis</a></li>
<li><a class="" href="transformation-of-the-rna-seq-data.html"><span class="header-section-number">4</span> Transformation of the RNA-Seq data</a></li>
<li class="book-part">No Gene ID Conversion</li>
<li><a class="" href="goseq.html"><span class="header-section-number">5</span> GOSeq</a></li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-go.html"><span class="header-section-number">6</span> clusterProfiler’s GSEA (with gene set database GO)</a></li>
<li><a class="" href="david.html"><span class="header-section-number">7</span> DAVID</a></li>
<li><a class="" href="gsea-web-based-tool.html"><span class="header-section-number">8</span> GSEA (web-based tool)</a></li>
<li class="book-part">With Gene ID Conversion</li>
<li><a class="" href="clusterprofilers-gsea-with-gene-set-database-kegg.html"><span class="header-section-number">9</span> clusterProfiler’s GSEA (with gene set database KEGG)</a></li>
<li><a class="" href="clusterprofilers-ora.html"><span class="header-section-number">10</span> clusterProfiler’s ORA</a></li>
<li><a class="" href="padog.html"><span class="header-section-number">11</span> PADOG</a></li>
<li><a class="" href="gseapreranked.html"><span class="header-section-number">12</span> GSEAPreranked</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="pre-filtering" class="section level1" number="1">
<h1>
<span class="header-section-number">1</span> Pre-filtering<a class="anchor" aria-label="anchor" href="#pre-filtering"><i class="fas fa-link"></i></a>
</h1>
<p>In this script, we will learn about two options to exclude all genes that do not have a sufficient number of read counts across
all samples. We distinguish between these two approaches since the different methods for differential expression analysis (see Chapter ‘Differential Expression Analysis’) propose different methods for pre-filtering.</p>
<ul>
<li>
<strong>option 1</strong>: pre-filtering using a function provided by edgeR</li>
<li>
<strong>option 2</strong>: pre-filtering proposed by DESeq2</li>
</ul>
<div id="libraries" class="section level2" number="1.1">
<h2>
<span class="header-section-number">1.1</span> Libraries<a class="anchor" aria-label="anchor" href="#libraries"><i class="fas fa-link"></i></a>
</h2>
<p>All necessary packages are available on Bioconductor, and should be installed from there if not already available on your machine.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"tweeDEseqCountData"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"edgeR"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load Libraries</strong></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.creal.cat/jrgonzalez/software.htm">tweeDEseqCountData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/">edgeR</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>Description of the libraries:</strong></p>
<ul>
<li><p><strong>tweeDESeqCountData</strong>: from this library we obtain the gene expression data set we will use for our illustrations.</p></li>
<li><p><strong>edgeR</strong>: offers a function for pre-filtering we will use below.</p></li>
</ul>
</div>
<div id="preparation-of-rna-seq-data-set-used-for-illustration" class="section level2" number="1.2">
<h2>
<span class="header-section-number">1.2</span> Preparation of RNA-Seq data set used for illustration<a class="anchor" aria-label="anchor" href="#preparation-of-rna-seq-data-set-used-for-illustration"><i class="fas fa-link"></i></a>
</h2>
<p>For the purpose of simplicity and readability, we store the gene expression measurements and sample conditions from the Pickrell data sets in objects with neutral names, namely in ‘expression_data’and ’sample_conditions’, respectively.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load pickrell data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pickrell</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># access and store gene expression measurements</span></span>
<span><span class="va">expression_data</span> <span class="op">&lt;-</span> <span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">pickrell.eset</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># access and store sample conditions </span></span>
<span><span class="va">sample_conditions</span> <span class="op">&lt;-</span> <span class="va">pickrell.eset</span><span class="op">$</span><span class="va">gender</span> </span></code></pre></div>
<p>Take a look at a few entries of the gene expression data set:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># inspect the read counts of the first 5 genes in the first 5 samples</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                 NA18486 NA18498 NA18499 NA18501 NA18502</span></span>
<span><span class="co">#&gt; ENSG00000000003       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000005       0       0       0       0       0</span></span>
<span><span class="co">#&gt; ENSG00000000419      22     105      40      55      67</span></span>
<span><span class="co">#&gt; ENSG00000000457      22     100     107      53      72</span></span>
<span><span class="co">#&gt; ENSG00000000460       5      23      10      18      15</span></span></code></pre></div>
<p>Take a look at the sample conditions:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_conditions</span></span>
<span><span class="co">#&gt;  [1] male   male   female male   female male   female male  </span></span>
<span><span class="co">#&gt;  [9] female male   female male   female male   female male  </span></span>
<span><span class="co">#&gt; [17] female female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [25] female male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [33] female male   female female female female male   female</span></span>
<span><span class="co">#&gt; [41] male   male   female female male   female female male  </span></span>
<span><span class="co">#&gt; [49] female female male   female male   male   female female</span></span>
<span><span class="co">#&gt; [57] male   female male   female male   female female male  </span></span>
<span><span class="co">#&gt; [65] female female female male   female</span></span>
<span><span class="co">#&gt; Levels: female male</span></span></code></pre></div>
<p>Inspect the number of genes (rows) and the number of samples (columns) in the gene expression data set:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 52580    69</span></span></code></pre></div>
</div>
<div id="pre-filtering-1" class="section level2" number="1.3">
<h2>
<span class="header-section-number">1.3</span> Pre-Filtering<a class="anchor" aria-label="anchor" href="#pre-filtering-1"><i class="fas fa-link"></i></a>
</h2>
<div id="option-1-pre-filtering-using-edgers-builting-function" class="section level3" number="1.3.1">
<h3>
<span class="header-section-number">1.3.1</span> Option 1: Pre-Filtering using edgeR’s builting function<a class="anchor" aria-label="anchor" href="#option-1-pre-filtering-using-edgers-builting-function"><i class="fas fa-link"></i></a>
</h3>
<p>This approach works with the function <em>filterByExpr()</em> from the package edgeR and is the proposed method for pre-filtering for the methods for differential expression analysis edgeR and voom/limma. This approach operates on the cpm-transformed count data (cpm: counts-per-million) and excludes all genes that do NOT have a certain number of counts-per-million in a certain number of samples. Note that this approach accounts for differences in the library size between the different samples.</p>
<p><strong>step 1: Generate input object required by filterByExpr()</strong></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_filterByExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">expression_data</span>, </span>
<span>                                        group <span class="op">=</span> <span class="va">sample_conditions</span><span class="op">)</span></span></code></pre></div>
<p>Description of function:</p>
<ul>
<li>
<strong>DGEList()</strong>: object to contain RNA sequencing measurements and additional information</li>
</ul>
<p>Description of arguments:</p>
<ul>
<li>
<strong>counts</strong>: matrix of RNA-Seq data</li>
<li>
<strong>group</strong>: vector that contains the condition of each sample</li>
</ul>
<p><strong>step 2: Filter out lowly expressed genes</strong></p>
<p>The function <em>filterByExpr()</em> creates an indicator which on which genes do and which do not have a sufficient amount of read counts across all samples. Based on this indicator, the gene expression data set is then filtered.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (i) for each gene, indicate if it fulfils the requirements to be kept for the subsequent analysis </span></span>
<span><span class="va">indicator_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">expression_data_filterByExpr</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (ii) filter the gene expression data set such that only those genes are kept which fulfill the requirements</span></span>
<span><span class="va">expression_data_filterByExpr</span> <span class="op">&lt;-</span> <span class="va">expression_data_filterByExpr</span><span class="op">[</span><span class="va">indicator_keep</span>,, keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span></code></pre></div>
<p>Note that the index <strong>keep.lib.sizes = FALSE</strong> ensures that the library size of each sample is recalculated after pre-filtering.</p>
<p><strong>step 3: Obtain final pre-filtered gene expression data set</strong></p>
<p>At this point, we transform the gene expression measurements back to a data frame. The reason for this is that some subsequent steps (such as the conversion of gene IDs do not work with the DGEList format).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_data_filterByExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">expression_data_filterByExpr</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># inspect the number of genes and the number of samples in the final pre-filtered gene expression data set </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression_data_filterByExpr</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6246</span></span></code></pre></div>
<p>Observe that the number of genes has been reduced compared to the original (unfiltered) gene expression data set.</p>
</div>
<div id="option-2-simpler-pre-filtering-approach-proposed-by-deseq2" class="section level3" number="1.3.2">
<h3>
<span class="header-section-number">1.3.2</span> Option 2: Simpler Pre-filtering approach (proposed by DESeq2)<a class="anchor" aria-label="anchor" href="#option-2-simpler-pre-filtering-approach-proposed-by-deseq2"><i class="fas fa-link"></i></a>
</h3>
<p>A simpler approach for pre-filtering has been proposed by the method for differential expression analysis DESeq2. In this approach, only those genes are kept for further analysis that have a pre-specified number of counts <span class="math inline">\(X\)</span> (such as 10) across all samples. A higher value of <span class="math inline">\(X\)</span> thereby leads to more genes being removed.<br>
Note that DESeq2 proposes a stricter version of pre-filtering in which those genes are kept which have <span class="math inline">\(X\)</span> number of counts in at least <span class="math inline">\(Y\)</span> samples.<br>
Note that none of the these two “simpler” approaches to pre-filtering take differences in library size into account.</p>
<p><strong>step 1: Pre-Filtering</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># indicate which genes have at least 10 read counts across all samples:</span></span>
<span><span class="va">indicator_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span> </span>
<span></span>
<span><span class="co"># alternative (and more strict) indicator:</span></span>
<span><span class="co"># indicator_keep &lt;- rowSums( expression_data &gt;=10) &gt;= 10</span></span>
<span></span>
<span><span class="co"># subset gene expression data set accordingly </span></span>
<span><span class="va">expression_data_filterDESEq2</span> <span class="op">&lt;-</span> <span class="va">expression_data</span><span class="op">[</span><span class="va">indicator_keep</span>,<span class="op">]</span></span></code></pre></div>
<p><strong>step 2: Inspect final pre-filtered gene expression data set</strong></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">expression_data_filterDESEq2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 10151    69</span></span></code></pre></div>
<p>Note that the number of genes in the gene expression data set has decreased compared to the initial (unfiltered) gene expression data set.</p>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="index.html">About</a></div>
<div class="next"><a href="gene-id-conversion-and-removal-of-the-resulting-duplicated-gene-ids.html"><span class="header-section-number">2</span> Gene ID conversion and removal of the resulting duplicated gene IDs</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#pre-filtering"><span class="header-section-number">1</span> Pre-filtering</a></li>
<li><a class="nav-link" href="#libraries"><span class="header-section-number">1.1</span> Libraries</a></li>
<li><a class="nav-link" href="#preparation-of-rna-seq-data-set-used-for-illustration"><span class="header-section-number">1.2</span> Preparation of RNA-Seq data set used for illustration</a></li>
<li>
<a class="nav-link" href="#pre-filtering-1"><span class="header-section-number">1.3</span> Pre-Filtering</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-pre-filtering-using-edgers-builting-function"><span class="header-section-number">1.3.1</span> Option 1: Pre-Filtering using edgeR’s builting function</a></li>
<li><a class="nav-link" href="#option-2-simpler-pre-filtering-approach-proposed-by-deseq2"><span class="header-section-number">1.3.2</span> Option 2: Simpler Pre-filtering approach (proposed by DESeq2)</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>A Practical Guide through Running Gene Set Analysis in R</strong>" was written by Milena Wünsch and Pat Callahan. It was last built on 2023-07-27.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
